---
title: 'Liver - gene module pruning and preservation analysis, sc and bulk'
author: "Jon Thompson, Pers lab, rkm916 at ku dot dk"
date: "`r Sys.time()`" 
output: workflowr::wflow_html
editor_options:
  chunk_output_type: console
params: 
  date: !r substr(gsub("-","",as.character(Sys.Date())),3,1000)
  randomSeed: !r 12345
  pValThreshold: !r 0.05
# output:
#   html_notebook:
#     df_print: paged
#     number_sections: yes
#     toc: yes
#     toc_depth: 3
#     toc_float: yes
#   html_document:
#     df_print: paged
#     number_sections: yes
#     toc: yes
#     toc_depth: 3
#     toc_float: yes
---

previous: ran WGCNA on perslab, moylan and gerhard separately

# Setup

## Set options


```{r}
options(stringsAsFactors = F, 
        use="pairwise.complete.obs", 
        warn=1, 
        verbose=F,
        mc.cores=40 # for parallel computation
        ) 
```

```{r}
randomSeed = 12345
set.seed(randomSeed)

pAdjMethod = "BH"
pValThreshold = 0.05 #params$pValThreshold

flagDate = substr(gsub("-","",as.character(Sys.Date())),3,1000)
```

## Load packages

```{r}

library("data.table")
library("dplyr")
library("ggplot2")
library("Matrix")
library("parallel")
library("here")
library("WGCNA")
library("corrplot")
```

## source utility functions

```{r}
source(here("perslab-sc-library", "utility_functions.R"))
source(here("perslab-sc-library", "functions_sc.R"))
```

## Set constants

generic project constants 

```{r}
#prefixData <- "aizarani"
prefixData = "liver_perslab_int"
prefixRun = "seurat_7"
prefixRunWGCNA = "wgcna3"
prefixOut <- "SCTint"

dirWGCNA_outs = "/projects/jonatan/pub-perslab/18-liver-wgcna/"
dirWGCNA_plots = paste0(dirWGCNA_outs, "plots/" )
dirWGCNA_RObjects = paste0(dirWGCNA_outs, "RObjects/" )
dirWGCNA_tables = paste0(dirWGCNA_outs, "tables/" )

colModule = "module"
colCellClust = "cell_cluster"
```

we are interested in a subset of the "cluster_perslab_coarse" cell types

```{r}
vec_celltypesOfInterest <- c("T-cells-alpha-beta",
                             "NK-like-cells",
                             "Endothelial-cells",
                             "T-cells-gamma-delta", 
                             "Macrophages",  
                             "Cholangiocytes",        
                             "Hepatocytes",
                             "Hepatic-stellate-cells", 
                             "Dendritic-cells", 
                             "Hepatic-stem-cells")
```

## load data

load gene module cell embedding matrix

```{r}
path_embed <-  paste0(dirWGCNA_tables, prefixData, "_",prefixRunWGCNA,"_kIM_cellModEmbed.csv.gz")
dt_embed <- fread(path_embed)
dt_embed[0:2,0:5]
```

load WGCNA gene module tables

perslab 

```{r}
path_geneMod <- paste0(dirWGCNA_tables,"liver_perslab_int_", prefixRunWGCNA,"_geneMod.csv.gz")
dt_geneMod_perslab <-fread(file=path_geneMod)
dt_geneMod_perslab[0:3,]
#                data    run   cell_cluster     module    genes        pkMs
# 1: liver_perslab_int wgcna3 B-cells-mature lightblue3 KIAA1551 0.009426350
# 2: liver_perslab_int wgcna3 B-cells-mature lightblue3     ZEB1 0.008230427
# 3: liver_perslab_int wgcna3 B-cells-mature lightblue3    XYLT1 0.008185864
```

moylan 2013

```{r}
path_geneMod_moylan2013 <- paste0(dirWGCNA_tables,"liver_moylan2013_", "wgcna1","_geneMod.csv.gz")
dt_geneMod_moylan2013 <-fread(file=path_geneMod_moylan2013)
dt_geneMod_moylan2013[0:3,]

#                data    run cell_cluster               module   genes      pkMs
# 1: liver_moylan2013 wgcna1   moylan2013 lightgoldenrodyellow    ANO2 0.2370768 
# 2: liver_moylan2013 wgcna1   moylan2013 lightgoldenrodyellow CNTNAP5 0.2365604 
# 3: liver_moylan2013 wgcna1   moylan2013 lightgoldenrodyellow  TEDDM1 0.2353492 

```

gerhard 2013

```{r}
path_geneMod_gerhard2018 <- paste0(dirWGCNA_tables,"liver_gerhard2018_", "wgcna2","_geneMod.csv.gz")
dt_geneMod_gerhard2018 <-fread(file=path_geneMod_gerhard2018)
dt_geneMod_gerhard2018[0:3,]
#             data    run cell_cluster       module           genes       pkMs
# 1: liver_gerhard2018 wgcna1  gerhard2018 midnightblue ENSG00000171848 0.03919702
# 2: liver_gerhard2018 wgcna1  gerhard2018 midnightblue ENSG00000066279 0.03745749
# 3: liver_gerhard2018 wgcna1  gerhard2018 midnightblue ENSG00000131747 0.03606625
```

load perslab cell metadata 

```{r}
dt_meta <- fread(here("data", paste0(prefixData, "_", prefixRun, "_metadata_full.csv.gz")))
dt_meta[0:3,0:5]
```

load Moylan 2013 bulk data

```{r}
dt_datExpr_moylan <- fread(here("data", "moylan2013.norm.expr.qc.csv.gz"))
```

# Analysis / filtering 

first map dt_geneMod genes to ensembl / hgnc

```{r}
dt_geneMod_perslab$ensembl <- gene_map(dataIn = dt_geneMod$genes, 
                               df_mapping = fread("/projects/jonatan/data/gene_remap/gene_annotation_hsapiens.txt.gz"),
                               from = "hgnc_symbol", 
                               to = "ensembl_gene_id", 
                               na.rm = F)
```

```{r}
dt_geneMod_moylan2013$ensembl <- gene_map(dataIn = dt_geneMod_moylan2013$genes, 
                               df_mapping = fread("/projects/jonatan/data/gene_remap/gene_annotation_hsapiens.txt.gz"),
                               from = "hgnc_symbol", 
                               to = "ensembl_gene_id", 
                               na.rm = F)
```

```{r}
colnames(dt_geneMod_gerhard2018)[which(colnames(dt_geneMod_gerhard2018)=="genes")] <- "ensembl"
```

```{r}
dt_geneMod_gerhard2018$genes <- gene_map(dataIn = dt_geneMod_gerhard2018$ensembl, 
                               df_mapping = fread("/projects/jonatan/data/gene_remap/gene_annotation_hsapiens.txt.gz"),
                               from = "ensembl_gene_id", 
                               to = "hgnc_symbol", 
                               na.rm = F)
```

add new filter columns to dt_geneMod to retain cell clusters of interest

```{r}
dt_geneMod_perslab[,cell_cluster_filter:=ifelse(cell_cluster %in% ..vec_celltypesOfInterest,cell_cluster,NA),]
dt_geneMod_perslab[,module_filter:=ifelse(cell_cluster %in% ..vec_celltypesOfInterest,module,NA),]
```

```{r}
dt_geneMod_moylan2013[,cell_cluster_filter:=cell_cluster]
dt_geneMod_moylan2013[,module_filter:=module]
```

```{r}
dt_geneMod_gerhard2018[,cell_cluster_filter:=cell_cluster]
dt_geneMod_gerhard2018[,module_filter:=module]
```

### combine geneMod data.tables into a single dt and write to disk

```{r}
dt_geneMod_comb <- data.table::rbindlist(l = list(dt_geneMod_perslab,dt_geneMod_gerhard2018, dt_geneMod_moylan2013), use.names = T)
```

```{r}
fwrite(dt_geneMod_comb, file="/projects/jonatan/pub-perslab/18-liver-wgcna/tables/liver_sc_bulk_geneMod_comb.csv.gz")
```

prune bulk RNAseq modules overlapping with any single cell modules

```{r}
opt <- list(path_df_NWA ="/projects/jonatan/pub-perslab/18-liver-wgcna/tables/liver_sc_bulk_geneMod_comb.csv.gz",
            colModule = "module_filter",
            colCellClust = "cell_cluster_filter",
            colGeneNames = "ensembl",
            colGeneWeights = "pkMs",
            minPropIntersect = 0.7,
            minPropIntersect_iter_increase = 1.02,
            minWeightedCor = 0.7,
            cellClusters_keep ='c("T-cells-alpha-beta", "NK-like-cells", "Endothelial-cells", "T-cells-gamma-delta", "Macrophages", "Cholangiocytes", "Hepatocytes","Hepatic-stellate-cells", "Dendritic-cells", "Hepatic-stem-cells")',
            corMethod = "pearson",
            mergeOrPrune = "merge",
            maxIter = 25, 
            dirOut = "/projects/jonatan/pub-perslab/18-liver-wgcna/",
            prefixOut = "merge1", 
            doPlot = T,
            RAMGbMax = 250)
```

# time Rscript /nfsdata/projects/jonatan/tools/wgcna-src/geneNetworkScripts/gene_module_merge2.R  --path_df_NWA /projects/jonatan/pub-perslab/18-liver-wgcna/tables/liver_sc_bulk_geneMod_comb.csv.gz --colGeneNames ensembl --colModule module_filter --colCellClust cell_cluster_filter --colGeneWeights pkMs --cellClusters_keep 'c("T-cells-alpha-beta", "NK-like-cells", "Endothelial-cells", "T-cells-gamma-delta", "Macrophages", "Cholangiocytes", "Hepatocytes","Hepatic-stellate-cells", "Dendritic-cells", "Hepatic-stem-cells")' --minPropIntersect  0.7 --minWeightedCor  0.7 --minPropIntersect_iter_increase  1.01 --maxIter 25 --corMethod pearson --mergeOrPrune prune --dirOut /projects/jonatan/pub-perslab/18-liver-wgcna/ --prefixOut merge1 --doPlot  T --RAMGbMax 250

re-load new geneMod table after pruning overlapping modules

```{r}
dt_geneMod <-fread(file="/projects/jonatan/pub-perslab/18-liver-wgcna/tables/liver_sc_bulk_geneMod_comb_merged.csv.gz")
```

```{r}
colCellClust <- "cell_cluster_merged"
colModule <- "module_merged"

length(unique(dt_geneMod$module_filter)[nchar(unique(dt_geneMod$module_filter))>0])
# [1] 220
length(unique(dt_geneMod$module_merged)[nchar(unique(dt_geneMod$module_merged))>0])
# [1] 209
```

209 modules

<!-- ## preservation analysis - Moylan2013 -->

<!-- ```{r} -->
<!-- opt = list(path_dt_geneMod = "/projects/jonatan/pub-perslab/18-liver-wgcna/tables/liver_perslab_int_wgcna3_geneMod_merged.csv.gz",  colGeneNames = "genes", colGeneWeights = "pkMs", colMod = "module_merged", colCellClust = "cell_cluster_merged", vec_pathsDatExpr = "c('perslab' = '/projects/jonatan/pub-perslab/18-liver-fred/output/liver_perslab_int_seurat_7_datExpr_combat.csv.gz', 'moylan2013'='/projects/jonatan/pub-perslab/18-liver-fred/data/moylan2013.norm.expr.qc.csv.gz')", vec_pathsMetadata = "c('perslab'='/projects/jonatan/pub-perslab/18-liver-fred/data/liver_perslab_int_seurat_7_metadata_full.csv.gz', 'moylan2013'='/projects/jonatan/pub-perslab/18-liver-fred/data/moylan2013.metadata.csv')", vec_metadataIdentCols = "c('perslab'='cluster_perslab_coarse', 'moylan2013'='data')", list_list_vec_identLvlsMatch = 'list("Cholangiocytes"=list("moylan2013"="moylan2013"), "Dendritic-cells"=list("moylan2013"="moylan2013"), "Endothelial-cells"=list("moylan2013"="moylan2013"),    "Hepatic-stellate-cells"=list("moylan2013"="moylan2013"), "Hepatic-stem-cells"=list("moylan2013"="moylan2013"), "Hepatocytes"=list("moylan2013"="moylan2013"),"Macrophages"=list("moylan2013"="moylan2013"),"NK-like-cells"=list("moylan2013"="moylan2013"),    "T-cells-alpha-beta"=list("moylan2013"="moylan2013"),"T-cells-gamma-delta"=list("moylan2013"="moylan2013"))', minCellClusterSize = 10, minGeneClusterSize = 15, dirOut = "/projects/jonatan/pub-perslab/18-liver-wgcna/", prefixOut = "liver_perslab_int_seurat_7_wgcna3", dirTmp = "/scratch/rkm916/tmp-wgcna/", networkType = 'c("signed hybrid")', corFnc = "cor", RAMGbMax = 300, path_runLog = NULL) -->
<!-- ``` -->

## preservation analysis



### single cell modules preservation in Gerhard2018

```{r}
# tmux l1
opt = list(path_dt_geneMod = "/projects/jonatan/pub-perslab/18-liver-wgcna/tables/liver_sc_bulk_geneMod_comb_merged.csv.gz",  
           colGeneNames = "ensembl", 
           colGeneWeights = "pkMs", 
           colMod = "module_merged", 
           colCellClust = "cell_cluster_merged", 
           vec_pathsDatExpr = "c('perslab' = '/projects/jonatan/pub-perslab/18-liver-fred/output/liver_perslab_int_seurat_7_datExpr_combat_ensembl.csv.gz',
           'gerhard2018'='/projects/jonatan/pub-perslab/18-liver-fred/data/gerhard2018_norm.counts.csv.gz')", 
           vec_pathsMetadata = "c('perslab'='/projects/jonatan/pub-perslab/18-liver-fred/data/liver_perslab_int_seurat_7_metadata_full.csv.gz', 
           'gerhard2018'='/projects/jonatan/pub-perslab/18-liver-fred/data/gerhard2018_patient_metadata_qc.csv.gz')", 
           vec_metadataIdentCols = "c('perslab'='cluster_perslab_coarse', 'gerhard2018'='data')", 
           list_list_vec_identLvlsMatch = 'list("Cholangiocytes"=list("gerhard2018"="gerhard2018"), 
           "Dendritic-cells"=list("gerhard2018"="gerhard2018"), 
           "Endothelial-cells"=list("gerhard2018"="gerhard2018"),    
           "Hepatic-stellate-cells"=list("gerhard2018"="gerhard2018"), 
           "Hepatic-stem-cells"=list("gerhard2018"="gerhard2018"),
           "Hepatocytes"=list("gerhard2018"="gerhard2018"),
           "Macrophages"=list("gerhard2018"="gerhard2018"),
           "NK-like-cells"=list("gerhard2018"="gerhard2018"),     
           "T-cells-alpha-beta"=list("gerhard2018"="gerhard2018"),
           "T-cells-gamma-delta"=list("gerhard2018"="gerhard2018"))', 
           minCellClusterSize = 10, 
           minpropModGenesInTestLvl = 0.5,
           minPropModGenesNon0inTestLvl = 0.1,
           dirOut = "/projects/jonatan/pub-perslab/18-liver-wgcna/", 
           prefixOut = "liver_perslab_int_seurat_7_wgcna3", 
           dirTmp = "/scratch/rkm916/tmp-wgcna/", 
           networkType = 'c("signed hybrid")', 
           corFnc = "cor", 
           RAMGbMax = 300, 
           path_runLog = NULL)
```


### Gerhard2018 modules preservation in single cell 

```{r}

opt = list(path_dt_geneMod = "/projects/jonatan/pub-perslab/18-liver-wgcna/tables/liver_sc_bulk_geneMod_comb_merged.csv.gz",  
           colGeneNames = "ensembl", 
           colGeneWeights = "pkMs", 
           colMod = "module_merged", 
           colCellClust = "cell_cluster_merged", 
           vec_pathsDatExpr = "c('gerhard2018'='/projects/jonatan/pub-perslab/18-liver-fred/data/gerhard2018_norm.counts.csv.gz',
           'liver_perslab' = '/projects/jonatan/pub-perslab/18-liver-fred/output/liver_perslab_int_seurat_7_datExpr_combat_ensembl.csv.gz')", 
           vec_pathsMetadata = "c('gerhard2018'='/projects/jonatan/pub-perslab/18-liver-fred/data/gerhard2018_patient_metadata_qc.csv.gz',
           'liver_perslab'='/projects/jonatan/pub-perslab/18-liver-fred/data/liver_perslab_int_seurat_7_metadata_full.csv.gz')", 
           vec_metadataIdentCols = "c('gerhard2018'='data', 'liver_perslab'='cluster_perslab_coarse')", 
           list_list_vec_identLvlsMatch = 'list("gerhard2018"=list("liver_perslab"=c("Cholangiocytes","Dendritic-cells","Endothelial-cells","Hepatic-stellate-cells","Hepatic-stem-cells", "Hepatocytes","Macrophages","NK-like-cells","T-cells-alpha-beta","T-cells-gamma-delta")))',
           minCellClusterSize = 10, 
           minpropModGenesInTestLvl = 0.5,
           minPropModGenesNon0inTestLvl = 0.1,
           dirOut = "/projects/jonatan/pub-perslab/18-liver-wgcna/", 
           prefixOut = "gerhard2018_wgcna2", 
           dirTmp = "/scratch/rkm916/tmp-wgcna/", 
           networkType = 'c("signed hybrid")', 
           corFnc = "cor", 
           RAMGbMax = 300, 
           path_runLog = NULL)
```

```{r}

path_dt_modPres <- paste0(dirWGCNA_tables, prefixData, "_", prefixRun, "_", prefixRunWGCNA, "_df_preservationStats.csv")

dt_modPres <- fread(path_dt_modPres)
```

quickly inspect module preservation stats

```{r}
table(dt_modPres$Z_stat>2)
```

```{r}
table(dt_modPres$log.p.Bonfsummary.pres<(log10(pValThreshold)))
```

add column with -log10 BH corrected p-values

```{r}
(1 - pnorm(dt_modPres$Z_stat)) %>% p.adjust(p=., method = pAdjMethod) -> vec_p.BH.pres

vec_p.BH.pres[vec_p.BH.pres==0] <- min(vec_p.BH.pres[vec_p.BH.pres>0])

vec_min.log.p.BH.pres <- -log10(vec_p.BH.pres)

dt_modPres[,min.log.p.BH.pres:=vec_min.log.p.BH.pres]
```

```{r}
dt_geneMod$cell_cluster_pres <- dt_geneMod$cell_cluster_merged
dt_geneMod$module_pres <- dt_geneMod$module_merged

vec_mod = unique(dt_geneMod$module_pres[dt_geneMod$cell_cluster_pres %in% vec_celltypesOfInterest])
# 182 modules

for (module in vec_mod){
  vec_logical = dt_geneMod$module_pres == module
  if (module %in% dt_modPres$module) {
    if (dt_modPres$min.log.p.BH.pres[dt_modPres$module==module]<=-log10(pValThreshold)) {
      dt_geneMod$cell_cluster_pres[vec_logical] <- dt_geneMod$module_pres[vec_logical] <- NA_character_
    } 
  } else {print(paste0(module, " not in dt_modPres"))}
}

length(unique(dt_geneMod$module_pres))  
#[1] 55
```

that leaves 41 modules + 14 from T-cells-alpha-beta for which modulePreservation failed (so they pass automatically)

## correlate with previous sc modules (diagnostics)

```{r}
prefixRun_WGCNA_cf = "wgcna2"
```

```{r}
dt_kMs <- fread(paste0(dirWGCNA_tables, prefixData,  "_", prefixRunWGCNA, "_kMs_full_join.csv.gz"))

dt_kMs_cf <- fread(paste0(dirWGCNA_tables, prefixData,  "_", prefixRun_WGCNA_cf, "_kMs_full_join.csv.gz"))
```

```{r}
vec_logical_col_kMs = colnames(dt_kMs) %in% unique(dt_geneMod$module_pres)
```

```{r}
vec_geneIntersect <- intersect(dt_kMs$genes, dt_kMs_wgcna2$genes)
```

```{r}

x= dt_kMs[match(vec_geneIntersect, genes),..vec_logical_col_x]
y = dt_kMs_wgcna2[match(vec_geneIntersect, dt_kMs$genes),-1] 

mat_cor <- cor(x = x, y=y, method = "pearson", use = "pairwise.complete.obs")
```

```{r, fig.height=16, fig.width = 16}
col2 <- rev(colorRampPalette(c("#67001F", "#B2182B", "#D6604D","#F4A582","#FDDBC7", "#FFFFFF", "#D1E5F0", "#92C5DE","#4393C3", "#2166AC", "#053061"))(200))

pdf(file = paste0(dirWGCNA_plots, prefixData,"_", prefixRunWGCNA, "_cf_cor_wgcna2_", params$date,".pdf"),width= 20, height=24)
corrplot(corr = mat_cor, 
         method = "color", 
         type = "upper", 
         is.corr = T,
         col = col2,
         tl.cex=0.7
         #cex= 0.6
         )
# heatmap(x = mat_cor, 
#         scale="none", 
#         keep.dendro=F, 
#        col =colorRampPalette(brewer.pal(n=9, name="RdYlBu"))(100))[1:60]
dev.off()
```

# wrap up

```{r}
fwrite(dt_geneMod, file=path_geneMod, compress="gzip")
```


load pruned combined module table

```{r}
dt_geneMod_comb_pruned <- paste0(dirWGCNA_tables,"liver_sc_bulk_geneMod_comb_merged.csv.gz")
dt_geneMod_comb_pruned <-fread(file=dt_geneMod_comb_pruned)
dt_geneMod_comb_pruned[0:3,]
#                data    run   cell_cluster     module    genes        pkMs
# 1: liver_perslab_int wgcna3 B-cells-mature lightblue3 KIAA1551 0.009426350
# 2: liver_perslab_int wgcna3 B-cells-mature lightblue3     ZEB1 0.008230427
# 3: liver_perslab_int wgcna3 B-cells-mature lightblue3    XYLT1 0.008185864
```

were any bulk modules pruned? 

```{r}
any(is.na(dt_geneMod_comb_pruned[data %in% c("liver_moylan2013", "liver_gerhard2018"),module_merged]))
any(nchar(dt_geneMod_comb_pruned[data %in% c("liver_moylan2013", "liver_gerhard2018"),module_merged])==0)
```

no!

append bulk gene module tables to full sc gene module table (including all columns)

```{r}
dt_geneMod_bulk_pruned_full <- dt_geneMod_comb_pruned[data %in%c("liver_moylan2013", "liver_gerhard2018"),]

dt_geneMod_bulk_pruned_full[,cell_cluster_filter:=cell_cluster]
dt_geneMod_bulk_pruned_full[,module_filter:=module]
dt_geneMod_bulk_pruned_full[,module_merged:=module]
dt_geneMod_bulk_pruned_full[,cell_cluster_merged:=cell_cluster]
dt_geneMod_bulk_pruned_full[,cell_cluster_pres:=cell_cluster]
dt_geneMod_bulk_pruned_full[,module_pres:=module]

```

```{r}
dt_geneMod_comb_pruned_full <- data.table::rbindlist(list(dt_geneMod, dt_geneMod_bulk_pruned_full), use.names=T)
```

# wrap up

```{r}
path_geneMod_comb_sc_bulk <- paste0(dirWGCNA_tables,"liver_perslab_int_", prefixRunWGCNA,"_geneMod_comb_sc_bulk.csv.gz")
fwrite(dt_geneMod_comb_pruned_full, file = path_geneMod_comb_sc_bulk, compress="gzip")
```


