---
title: 'Liver - figures'
author: "Jon Thompson, Pers lab, rkm916 at ku dot dk"
date: "`r Sys.time()`" 
output: workflowr::wflow_html
editor_options:
  chunk_output_type: console
params: 
  date: !r substr(gsub("-","",as.character(Sys.Date())),3,1000)
  randomSeed: !r 12345
  pValThreshold: !r 0.05
# output:
#   html_notebook:
#     df_print: paged
#     number_sections: yes
#     toc: yes
#     toc_depth: 3
#     toc_float: yes
#   html_document:
#     df_print: paged
#     number_sections: yes
#     toc: yes
#     toc_depth: 3
#     toc_float: yes
---

# Setup

## Load packages

```{r}

# workaround until we fix the project library AND the setting of environment variables
.libPaths(new=c(.libPaths(),"/nfsdata/tools/R/envs/shared/4.0_20201209"))

library("here")
library("data.table")
library("Matrix")
library("parallel")

library("Seurat")#, lib.loc = "/nfsdata/tools/R/envs/rkm916/x86_64-redhat-linux-gnu-library/4.0/")

library("dplyr")
library("ggplot2")
library("ggrepel")
library("igraph")
library("oaqc")
library("tidyr")
library("tidygraph")
library("ggraph")
library("wesanderson")
library("colorspace")
library("patchwork")
library("ggsci")
library("ggpubr")
library("graphlayouts")
library("pheatmap")

library("eulerr")
library("lattice")
# https://github.com/yanwu2014/swne
library("NNLM")
library("swne")

# source utility functions
source(here("perslab-sc-library", "utility_functions.R"))
source(here("perslab-sc-library", "functions_sc.R"))

# Set options

options(stringsAsFactors = F, 
        use="pairwise.complete.obs", 
        warn=1, 
        verbose=F,
        boot.parallel="multicore", # see ?boot: each worker process inherits the environment of the current session, including the workspace and the loaded namespaces and attached packages (but not the random number seed: see below). 
        boot.ncpus = 20,
        mc.cores=40 # for parallel computation
        ) 

# data.table
setDTthreads(threads = NULL, 
             restore_after_fork = NULL)
getDTthreads()

#options(datatable.WhenJisSymbolThenCallingScope=TRUE)

# Set constants

randomSeed = 12345
set.seed(randomSeed)
npcs = 60
# maxit = 100 # for rlm, see users.stat.umn.edu/~sandy/courses/8053/handouts/robust.pdf
flagDate =substr(gsub("-","",as.character(Sys.Date())),3,1000)


# generic project constants 

prefixData = "liver_perslab_int"
prefixRun = "seurat_7"
dirWGCNA_outs = "/projects/jonatan/pub-perslab/18-liver-wgcna/"
dirWGCNA_plots = paste0(dirWGCNA_outs, "plots/" )
dirWGCNA_RObjects = paste0(dirWGCNA_outs, "RObjects/" )
dirWGCNA_tables = paste0(dirWGCNA_outs, "tables/" )

colGeneNames = "genes"
colGeneWeights = "pkMs"
```

```{r}
fontSize_plotTitle = 14

fontSize_label_lg = 9
fontSize_label_md = 7
fontSize_label_sm = 4


fontSize_axisTitle_lg = 24
fontSize_axisTitle_md = 20
fontSize_axisTitle_sm = 16

fontSize_axisText_xlg = fontSize_legend_xlg = 20
fontSize_axisText_lg = fontSize_legend_lg = 15
fontSize_axisText_md = fontSize_legend_md = 12
fontSize_axisText_sm = fontSize_legend_sm = 9
fontSize_axisText_xsm = fontSize_legend_xsm = 7


alpha_unicode = "\u03B1"
Beta_unicode = "\u03B2"
gamma_unicode = "\u03B3"
delta_unicode = "\u03B4"



vec_celltype_abbrev = c("Chol", 
                        "DC", 
                        "Endo", 
                        "HSC", 
                        "Stem", 
                        "Hep", 
                        "Mac", 
                        "NK", 
                        paste0(alpha_unicode, Beta_unicode, "T"),#bquote(alpha*Beta~"T"),#expression(paste(alpha, beta, "T")), 
                        paste0(gamma_unicode,delta_unicode,"T"),#bquote(gamma*delta~"T"),#expression(paste(gamma*delta,~"T")),
                        "Bcell",
                        "Plasma")


names(vec_celltype_abbrev) = c("Cholangiocytes","Dendritic-cells", "Endothelial-cells","Hepatic-stellate-cells","Hepatic-stem-cells","Hepatocytes","Macrophages","NK-like-cells",paste0("T-cells-",alpha_unicode, Beta_unicode),paste0("T-cells-",gamma_unicode,delta_unicode),"B-cells-mature","Plasma-cells")
```


## load data

```{r}
seuratObj <- readRDS(here("output","liver_perslab_int_seurat_7_SCTint_perslab_labels_seuratObj.RDS.gz"))
```

Load modules dataframes

```{r}
dt_geneMod_perslab <- fread("/projects/jonatan/pub-perslab/18-liver-wgcna/tables/liver_perslab_int_wgcna3_geneMod.csv.gz")
#head(dt_geneMod_perslab)
```

```{r}
dt_geneMod_gerhard <- fread("/projects/jonatan/pub-perslab/18-liver-wgcna/tables/liver_gerhard2018_wgcna3_geneMod.csv.gz")

```

```{r}
dt_geneMod_comb = rbindlist(list(dt_geneMod_perslab, dt_geneMod_gerhard), use.names = T)
```

load association results

<!-- ```{r} -->
<!-- list_dt_condition_coef = lapply(c("steatosis","lob_inflam", "fibrosis"), function(condition) { -->
<!--   fread(file = paste0(dirWGCNA_tables, prefixData, "_sc_bulk_modules_scaled_",condition,"_association_in_gerhard2018_200511.csv"))}) -->
<!-- ``` -->

```{r}
dt_condition_coef <- fread(paste0(dirWGCNA_tables, prefixData, "_mod_assoc_in_gerhard2018_BH_201217.csv"))
```

<!-- ```{r} -->
<!-- dt_mod_embed_gerhard =  -->
<!-- ``` -->

```{r}
# dt_markers <- openxlsx::read.xlsx(here("output","liver_perslab_align_seurat_7_SCTransformIntegrated_clustermarkers.xlsx"))
# head(dt_markers)
```

```{r}
Seurat::DefaultAssay(seuratObj) <-"SCT_combat"
```

## preprocessing

rename modules in seurat meta data

module-metamodule dict

```{r}
if (F) {
  dict = dt_geneMod_perslab[,.(module, module_renamed, module_meta)]
  
  dict = dict[!duplicated(module_renamed) & !is.na(module_renamed) & nchar(module_renamed)>0]
  
  colnames(seuratObj@meta.data)[match(dict$module, colnames(seuratObj@meta.data))] <-  dict$module_renamed  
}
```

subset out erythrocytes,
revise cell cluster numbering,
add unicode to cell cluster names,
rerun dimensional reduction

(do once, then save Seurat object to disk)

```{r}
if (F) {
    
  seuratObj_erythrocytes = subset(x=seuratObj, subset = cluster_perslab_coarse == "Erythrocytes" )
  
  saveRDS(seuratObj_erythrocytes, here("output","liver_perslab_int_seurat_7_SCTint_perslab_labels_seuratObj_erythrocytes.RDS.gz"), compress="gzip")
  
  seuratObj <- subset(x=seuratObj, subset = cluster_perslab_coarse != "Erythrocytes" )

  # rename clusters
  
  
  vec_cluster_perslab <- seuratObj$cluster_perslab
  
  vec_cluster_perslab_numbers_unique<- gsub("-|\\D*", "", vec_cluster_perslab) %>% as.numeric %>% unique
  
  vec_cluster_perslab_numbers_unique <- sort(vec_cluster_perslab_numbers_unique)
  
  names(vec_cluster_perslab_numbers_unique) <- order(vec_cluster_perslab_numbers_unique)-1
  
  vec_cluster_perslab_revised <- gsub("\\d", "", vec_cluster_perslab)
  
  vec_cluster_perslab_revised <- paste0( names(vec_cluster_perslab_numbers_unique)[match(gsub("-|\\D*", "", vec_cluster_perslab), vec_cluster_perslab_numbers_unique)],vec_cluster_perslab_revised)
  
  seuratObj$cluster_perslab <- vec_cluster_perslab_revised
  
  for (i in 1:4) {
    seuratObj$cluster_perslab =   gsub(pattern=c("alpha-", "beta", "gamma-", "delta")[i], 
    replacement=c(alpha_unicode, Beta_unicode, gamma_unicode, delta_unicode)[i],
    x=seuratObj$cluster_perslab)
  
  }
  
  for (i in 1:4) {
    seuratObj$cluster_perslab_coarse =   gsub(pattern=c("alpha-", "beta", "gamma-", "delta")[i], 
    replacement=c(alpha_unicode, Beta_unicode, gamma_unicode, delta_unicode)[i],
    x=seuratObj$cluster_perslab_coarse)
  
  }
  
  Idents(seuratObj) <- "cluster_perslab"
  
  # re-run PCA and UMAP
  
  DefaultAssay(seuratObj) = "integrated"
  seuratObj %>% Seurat::RunPCA(object=., npcs=npcs, seed.use=randomSeed) %>% RunUMAP(object=., reduction="pca", seed.use=randomSeed, dims=1:npcs) -> seuratObj
  
  saveRDS(seuratObj, here("output","liver_perslab_int_seurat_7_SCTint_perslab_labels_seuratObj.RDS.gz"), compress="gzip")

}
```

generate cell type colors

```{r}

if (F) {
  vec_cluster_perslab = unique(seuratObj$cluster_perslab)
  length(vec_cluster_perslab)
  # 19
  
  # http://www.sthda.com/english/wiki/colors-in-r
  # https://github.com/karthik/wesanderson
  vec_colors_cluster_perslab =c(
    wes_palette("Rushmore1")[c(1:3,5)],
    wes_palette("Royal1")[c(1,4)],
    wes_palette("Royal2")[c(1,4)],
    wes_palette("Zissou1")[c(1,3,4)],
    wes_palette("Darjeeling1")[c(5)],
    wes_palette("Darjeeling2")[c(2,4)], 
    #wes_palette("Chevalier1")[3], 
    wes_palette("FantasticFox1")[5],  
    wes_palette("GrandBudapest2")[c(1,2,4)],
    wes_palette("IsleofDogs1")[5]
    )#RColorBrewer::brewer.pal(name = "Paired", n =12)[3:12]
  
  # shuffle colors till they look good on UMAP
  set.seed(randomSeed+5)
  vec_colors_cluster_perslab = sample(x=vec_colors_cluster_perslab, size = length(vec_colors_cluster_perslab), replace=F)
  
  names(vec_colors_cluster_perslab) = vec_cluster_perslab#[c(1,2,4,5,6,7,8,9,12,13)]
  
  saveRDS(vec_colors_cluster_perslab, file =  paste0(dirWGCNA_plots,prefixData,"_", prefixRun,  "_vec_colors_cluster_perslab.RDS"))
}

vec_colors_cluster_perslab = readRDS(paste0(dirWGCNA_plots,prefixData,"_", prefixRun,  "_vec_colors_cluster_perslab.RDS"))
```

### cluster perslab coarse colors

```{r}
if (F){
  vec_cluster_perslab_coarse =unique(seuratObj$cluster_perslab_coarse)
  length(vec_cluster_perslab_coarse)
  # 13
  
  # http://www.sthda.com/english/wiki/colors-in-r
  # https://github.com/karthik/wesanderson
  
  vec_colors_cluster_perslab_coarse = sapply(vec_cluster_perslab_coarse, function(cluster_coarse) {
    vec_colors_cluster_perslab[grep(cluster_coarse, names(vec_colors_cluster_perslab))[1]]
  })
  
  names(vec_colors_cluster_perslab_coarse) = vec_cluster_perslab_coarse
  # vec_colors_cluster_perslab_coarse = c(wes_palette("Darjeeling1"), wes_palette("Darjeeling2")[c(1,2,4)],wes_palette("Chevalier1")[c(1,2,4)], wes_palette("FantasticFox1")[5], wes_palette("Moonrise3")[2]) #RColorBrewer::brewer.pal(name = "Paired", n =12)[3:12]
  # 
  # # shuffle colors till they look good on UMAP
  # set.seed(randomSeed+7)
  # 
  # names(vec_colors_cluster_perslab_coarse) = sample(x=vec_cluster_perslab_coarse, size = length(vec_cluster_perslab_coarse), replace=F)#[c(1,2,4,5,6,7,8,9,12,13)]
  saveRDS(vec_colors_cluster_perslab_coarse, file =  paste0(dirWGCNA_plots,prefixData,"_", prefixRun,  "_vec_colors_cluster_perslab_coarse.RDS"))
}

vec_colors_cluster_perslab_coarse = readRDS(paste0(dirWGCNA_plots,prefixData,"_", prefixRun,  "_vec_colors_cluster_perslab_coarse.RDS"))
```


```{r}
vec_colors_meta = wes_palette("Darjeeling1")

vec_metamod = unique(dt_geneMod_perslab$module_meta) %>% sort
vec_metamod = vec_metamod[nchar(vec_metamod)>0]

names(vec_colors_meta) = vec_metamod
```

### PLOTS 


## Fig. 1

## Fig. 1 A: single-cell workflow figure - bare-bones UMAP 

```{r}
Idents(seuratObj) <- "cluster_perslab_coarse"
```

```{r}
p <- UMAPPlot(seuratObj, 
             #reduction = "umap",
             #group.by="cluster_perslab_coarse", 
             seed=randomSeed,
             #label.box=TRUE,
             cols=vec_colors_cluster_perslab_coarse,
             label = F
             #label.size = 5,
             #repel=T
             )

p <- p+ Seurat::NoAxes() + Seurat::NoLegend()#theme(legend.position="none")
# theme(axis.line=element_blank(),
#       axis.text.x=element_blank(),
#       axis.text.y=element_blank(),
#       axis.ticks=element_blank(),
#       axis.title.x=element_blank(),
#       axis.title.y=element_blank(),
#       legend.position="none",
#       panel.background=element_blank(),
#       panel.border=element_blank(),
#       panel.grid.major=element_blank(),
#       panel.grid.minor=element_blank(),
#       plot.background=element_blank())
p

ggsave(plot= p, filename = here("plots", paste0(prefixData, "_", prefixRun, "_fig1a_umap_cluster_perslab_",flagDate,".pdf")), width = 12, height=12)
```


### Fig. 1 B. dataset overview (3 plots)

<!-- ```{r} -->
<!-- Idents(seuratObj) <- "cluster_perslab" -->
<!-- ``` -->

<!-- save colors -->

<!-- ```{r} -->
<!-- # code from: https://github.com/satijalab/seurat/issues/257 -->
<!-- pbuild <- ggplot2::ggplot_build(p) # Use ggplot_build to deconstruct the ggplot object -->
<!-- pdata <- pbuild$data[[1]] -->
<!-- vec_colors_full = pdata$colour -->
<!-- names(vec_colors_full) = seuratObj$cluster_perslab -->
<!-- vec_colors_cluster_perslab = unique(vec_colors_full) -->
<!-- names(vec_colors_cluster_perslab) = unique(names(vec_colors_full)) -->

<!-- saveRDS(vec_colors_cluster_perslab, file =  paste0(dirWGCNA_plots,prefixData,"_", prefixRun,  "_vec_colors_cluster_perslab.RDS")) -->
<!-- ``` -->


<!-- ```{r} -->
<!-- Idents(seuratObj) <- "cluster_perslab_coarse" -->
<!-- ``` -->

Fig 1B : UMAP (cluster_perslab_coarse) 

```{r}

p1B <- DimPlot(seuratObj, 
             reduction = "umap",
             group.by="cluster_perslab_coarse", 
             seed=randomSeed, 
             label = T,
             cols= vec_colors_cluster_perslab_coarse,
             label.size = fontSize_label_md,
             repel=T)

p1B <- p1B + theme(
  legend.position="none",
  axis.title.x = element_text(size=fontSize_axisTitle_md),
  axis.title.y = element_text(size=fontSize_axisTitle_md))

p1B

#ggsave(plot= p, filename = here("plots", paste0(prefixData, "_", prefixRun, "_fig1B_umap_cluster_perslab_coarse_",flagDate,".pdf")), width = 12, height=12)
```

<!-- ```{r} -->
<!-- # code from: https://github.com/satijalab/seurat/issues/257 -->
<!-- pbuild <- ggplot2::ggplot_build(p) # Use ggplot_build to deconstruct the ggplot object -->
<!-- pdata <- pbuild$data[[1]] -->
<!-- vec_colors_full_1 = pdata$colour -->
<!-- names(vec_colors_full_1) = seuratObj$cluster_perslab_coarse -->
<!-- vec_colors_cluster_perslab_coarse = unique(vec_colors_full_1) -->
<!-- names(vec_colors_cluster_perslab_coarse) = unique(names(vec_colors_full_1)) -->

<!-- saveRDS(vec_colors_cluster_perslab_coarse, file =  paste0(dirWGCNA_plots,prefixData,"_", prefixRun,  "_vec_colors_cluster_perslab_coarse.RDS")) -->
<!-- ``` -->


<!-- ```{r} -->
<!-- dt_plot = data.table( -->
<!--   cluster_perslab = Idents(seuratObj), -->
<!--   seuratObj@reductions$umap@cell.embeddings) -->
<!-- ``` -->

<!-- ```{r} -->
<!-- max_lim = 11 -->
<!-- ``` -->

<!-- ```{r} -->
<!-- dt_cluster_perslab_mean_coords = dt_plot[,lapply(.SD, mean), by=cluster_perslab, .SDcols=c("UMAP_1","UMAP_2")] -->
<!-- ``` -->

<!-- ```{r} -->
<!-- set.seed(randomSeed) -->

<!-- p <- ggplot(data=dt_plot, mapping=aes(UMAP_1, UMAP_2, color=cluster_perslab)) + -->

<!--       geom_point(size=0.15, alpha=0.7) +  -->

<!--       scale_x_continuous(limits = c(-max_lim, max_lim)) +  -->
<!--       scale_y_continuous(limits=c(-max_lim, max_lim)) + -->

<!--       #scale_color_manual(values=vec_colors_cluster_perslab) +  -->

<!--       ggrepel::geom_text_repel( -->
<!--         direction    = "both", -->
<!--         segment.color = "grey50", -->
<!--         point.padding=0.1, -->
<!--         #label.padding =0.1, -->
<!--         force=0.1, -->
<!--         segment.size = 0.2, -->
<!--         data=dt_cluster_perslab_mean_coords, -->
<!--         mapping = aes( -->
<!--             x=UMAP_1, -->
<!--             y=UMAP_2, -->
<!--             label=stringr::str_wrap(cluster_perslab)), -->
<!--         color="black", -->
<!--         size=4,  -->
<!--         ) +  -->
<!--   theme_minimal() + -->
<!--   # removes all legends  -->
<!--   # https://stackoverflow.com/questions/35618260/remove-legend-ggplot-2-2 -->
<!--   theme(legend.position="none") -->

<!-- p -->

<!-- ``` -->




<!-- ```{r} -->
<!-- vec_colors_cluster_perslab_full = pdata$colour  -->

<!-- names(vec_colors_cluster_perslab_full) <- levels(Idents(seuratObj))[match(pdata$group, order(levels(Idents(seuratObj))))] -->

<!-- vec_colors_cluster_perslab <- unique(vec_colors_cluster_perslab_full) -->
<!-- names(vec_colors_cluster_perslab) <- unique(names(vec_colors_cluster_perslab_full)) -->

<!-- vec_colors_cluster_perslab -->
<!-- ``` -->

Fig 1C: barplot showing distribution of clusters per patient

```{r, include=F}
dt = data.table("cluster" = seuratObj$cluster_perslab_coarse,
                "donor"= as.character(seuratObj$sample_ID))

#dt[,cluster:=factor(cluster, levels = 0:max(cluster), ordered = T)]
dt[,n_donor := paste0(donor," (",.N, " cells)"), by=donor]
vec_factorLevels <- dt$n_donor[gsub("\\ .*","",dt$n_donor) %>% as.numeric %>% order] %>% unique
dt[,n_donor := factor(n_donor, levels = vec_factorLevels, ordered=T),]
dt_sum <- dt[,.N, by=.(n_donor,cluster)]
```

```{r, echo=F, fig.width=12}

p1C <- ggplot(dt_sum,
            aes(x = n_donor, y=N, fill = factor(cluster))) + 
  
  geom_bar(
    position="fill", 
    stat="identity",
    width=0.6,
    #position=position_dodge()
    ) + 
  
  scale_y_continuous(labels = scales::percent) +
  scale_fill_manual(values=vec_colors_cluster_perslab_coarse) +
  
  theme(
    axis.title.x = element_text(size=fontSize_axisTitle_lg, vjust=0),
    axis.text.x = element_text(
      angle = 90, 
      size=fontSize_axisText_xlg,
      vjust=0.5
      ),
    axis.title.y = element_text(size=fontSize_axisTitle_lg),
    axis.text.y = element_text(size=fontSize_axisText_xlg),
    
    legend.title = element_blank(),
    legend.text = element_text(size = fontSize_legend_xlg),
    legend.background = element_blank(),
    legend.box.background = element_blank(),
    plot.background=element_blank(),  
    aspect.ratio = 1.2
        ) +
  
  labs(x="donor", y="proportion", fill = "cluster")

p1C
#saveMeta(savefnc= ggsave, plot= p, filename = here("plots", paste0(prefixData, "_", prefixRun, "_fig1C_barplot_donor_cluster_pct_",flagDate,".pdf")), width = 8, height=7)
```


### check sample representation for describing in manu
```{r}
dt_tmp = data.table("sample_ID" = seuratObj@meta.data$sample_ID, "cluster_perslab_coarse" = seuratObj@meta.data$cluster_perslab_coarse)
dt_tmp_1 = dt_tmp[,.N, by=.(sample_ID, cluster_perslab_coarse)]
nrow(dt_tmp_1)
# [1] 117
# expecting 120
```

```{r}
for (id in unique(dt_tmp$sample_ID)) {
  for (clust in unique(dt_tmp$cluster_perslab_coarse)) {
    condition = quote(dt_tmp$sample_ID==id & dt_tmp$cluster_perslab_coarse == clust) 
    if (nrow(dt_tmp[eval(condition)])==0) {
      print(paste0("cluster ", clust, " contains no cells from sample ", id))
    }
  }
}
```



<!-- get coarse cluster marker genes -->

<!-- ```{r} -->
<!-- clusters = seuratObj@meta.data$cluster_perslab_coarse %>% table %>% names  -->

<!-- Seurat::Idents(seuratObj) <-seuratObj@meta.data$cluster_perslab_coarse -->

<!-- list_iterable = list("X"=clusters) -->
<!-- fun = function(cluster) {tryCatch({ -->
<!--   FindMarkers(seuratObj ,   -->
<!--           #cells.1=colnames(seurat_obj)[Idents(seurat_obj)==cluster], -->
<!--           #cells.2=NULL, -->
<!--           ident.1 = cluster, -->
<!--           only.pos = T, -->
<!--           #ident.2 = clusters[clusters!=cluster], -->
<!--           test.use  ="wilcox", -->
<!--           max.cells.per.ident=1000, -->
<!--           random.seed=randomSeed, -->
<!--           #latent.vars = if (!is.null(merge_specify) | !is.null(merge_group_IDs)) "sample_ID" else NULL, -->
<!--           verbose = T) -->
<!--           },  -->
<!--           error = function(err) { -->
<!--             NA_character_ -->
<!--             warning(paste0("findmarkers failed for ", cluster, " with error ")) -->
<!--           })} -->

<!-- list_markers=NULL -->
<!-- list_markers <- lapply(FUN=fun, "X"=list_iterable[[1]]) -->

<!-- # add the gene and cluster as a column -->
<!-- list_markers <- mapply(function(df_markers, cluster) { -->
<!--   if (!all(sapply(df_markers, is.na))) { -->
<!--     cbind("gene" = rownames(df_markers), "cluster"=rep(cluster, nrow(df_markers)), df_markers) -->
<!--   } else { -->
<!--     NA_character_ -->
<!--   } -->
<!-- }, -->
<!-- df_markers=list_markers,  -->
<!-- cluster=clusters, SIMPLIFY=F) -->

<!-- list_markers <- list_markers[!sapply(list_markers, function(markers) all(is.na(markers)))] -->
<!-- df_markers <- Reduce(x=list_markers, f=rbind) -->
<!-- rownames(df_markers) <- NULL -->

<!-- saveMeta(savefnc=openxlsx::write.xlsx, x = df_markers, file=here("output", paste0(prefixData, "_", prefixRun, "coarse_clustermarkers_", flagDate,".xlsx"))) -->

<!-- ``` -->

<!-- ```{r} -->
<!-- dt_markers = setDT(df_markers) -->
<!-- ``` -->


Fig 1D : cluster marker genes violin plot

```{r}

vec_markers = c(
  "MS4A1", # B-cells-mature
  "TM4SF4", # Cholangiocytes
  "HMGB2", # Dendritic cells
  "CD36", # Endothelial-cells
  "IGFBP7", # "Hepatic-stellate-cells
  "AFF3",#, # Hepatic-stem-cells
  "APOA1", # Hepatocytes
  "AIF1", # Macrophages
  "KLRF1", # NK-like-cells
  "MZB1", # Plasma-cells
  "IL7R", # T-cells-alpha-beta
  "GNLY") # T-cells-gamma-delta

  #"HBA1", # Eruthrocytes 
  
  #)
  
names(vec_markers) <-  sort(unique(seuratObj$cluster_perslab_coarse))
```

```{r}
factor_cell_clust_group <- factor(sort(unique(seuratObj$cluster_perslab_coarse)), ordered = T)
```

<!-- ```{r} -->
<!-- vec_markers = unique(vec_markers) -->
<!-- ``` -->

<!-- ### dot plot -->

<!-- ```{r,fig.width=18, fig.height=7} -->

<!-- p<-DotPlot_timshel( -->
<!--   object=seuratObj, -->
<!--   assay="SCT_combat", -->
<!--   slot = "data", -->
<!--   do.scale=T, -->
<!--   genes.plot=rev(vec_markers),#rev(df$gene_name), -->
<!--   cols.use = c("blue","white", "red"),#brewer.pal(n=11,"Spectral"),#c("blue","red"), -->
<!--   col.min = -2.5, -->
<!--   col.max = 2.5, -->
<!--   dot.min = 0, -->
<!--   dot.scale = 6, -->
<!--   group.by="cluster_perslab_coarse", -->
<!--   group.order=factor_cell_clust_group, -->
<!--   plot.legend = T, -->
<!--   do.return = T, -->
<!--   x.lab.rot = T, -->
<!--   coord_flip = F, -->
<!--   df.marker.panel.to.plot=NULL#df -->
<!-- ) -->

<!-- p -->

<!-- saveMeta(savefnc= ggsave, plot= p, filename = here("plots", paste0(prefixData, "_", prefixRun, "_SCT_combat_clust_coarse_markergene_dotplot",flagDate,".pdf")), width = length(vec_markers)/2.5, height=length(unique(names(vec_markers)))/3.5) -->
<!-- ``` -->


<!-- ```{r} -->
<!-- factor_markers = factor(vec_markers, ordered=T) -->
<!-- ``` -->

```{r}
all(names(vec_markers) == sort(names(vec_colors_cluster_perslab_coarse)))
# [1] TRUE
vec_colors_markers <- vec_colors_cluster_perslab_coarse[order(names(vec_colors_cluster_perslab_coarse))]

names(vec_colors_markers) <-vec_markers

vec_colors_markers
```

reorder cells in seuratObj 

```{r}
Idents(seuratObj) <- "cluster_perslab_coarse"
levels(x = seuratObj) = sort(names(vec_markers), decreasing = T)
```

```{r}
#p2  <- VlnPlot(object=seuratObj,
list_plot <- VlnPlot(object=seuratObj,
                     assay="SCT_combat",
                     features = vec_markers,
                     pt.size = 0,# 0.05,
                     cols=vec_colors_cluster_perslab_coarse,#vec_colors_markers,
                     sort = F, 
                     #group.by = "cluster_perslab_coarse", 
                     same.y.lims = F,
                     slot="data",
                     log = F,
                     combine = F,
                     flip=F)#,
                     #ncol=length(vec_markers))

names(list_plot) <- vec_markers

list_plot_flip <- lapply(1:length(list_plot), function(i) {
  plot_tmp=list_plot[[i]]
  plot_tmp = plot_tmp +
  coord_flip() +
  theme(
    plot.title = element_text(face = "bold.italic", size=fontSize_plotTitle*2),
    axis.text.y=element_blank(),
    plot.margin = margin(b=1, unit="cm")) #margin around entire plot (unit with the sizes of the top, right, bottom, and left margins)

  if (i==1) {
  plot_tmp <- plot_tmp +
  theme(
    axis.text.y=element_text(
    face = "bold",
    hjust=1,
    vjust=0,
    size=fontSize_axisText_xlg,
    angle=30
    )
  )
  }
  # } else {
  #   plot_tmp <- plot_tmp + 
  #   theme(axis.text.y=element_text(
      # vjust=0.2,
      # size=10,
      # angle=30,
  #     margin = margin(l = 0, r = 5)
  #     )
  #     )
  # }
  
  plot_tmp <- plot_tmp + 
    theme(
      axis.title.y=element_blank(),
      axis.title.x=element_blank(), 
      axis.text.x = element_text(
        size= fontSize_axisText_xsm,
        angle=35),
      plot.title = element_text(size=fontSize_plotTitle),
      #aspect.ratio=10,
      legend.position="none")
})

p1D <- patchwork::wrap_plots(... = list_plot_flip, 
                            ncol = length(list_plot_flip)
                            )

# add cell cluster stats barplot to side of marker gene violinplot grid

# * cell number
# * median genes per cell
# * median RNA per cell 


dt_meta_sc = data.table("cell_id" =  rownames(seuratObj@meta.data),  seuratObj@meta.data)

dt_cell_clust_stats = dt_meta_sc[,list("cells"=.N, "median genes"=median(nFeature_RNA), "median RNA"=median(nCount_RNA)), by=cluster_perslab_coarse]

dt_cell_clust_stats_melt = melt(dt_cell_clust_stats, id.vars=c("cluster_perslab_coarse"))

dt_cell_clust_stats_melt = dt_cell_clust_stats_melt[cluster_perslab_coarse != "Erythrocytes"]

dt_cell_clust_stats_melt = dt_cell_clust_stats_melt[order(-cluster_perslab_coarse)]

dt_cell_clust_stats_melt$cluster_perslab_coarse = factor(dt_cell_clust_stats_melt$cluster_perslab_coarse, levels=unique(dt_cell_clust_stats_melt$cluster_perslab_coarse), ordered = T)
# dt_cell_clust_stats_melt_RNA = dt_cell_clust_stats_melt[variable %in% c("median genes", "median RNA") ]
# dt_cell_clust_stats_melt_cells = dt_cell_clust_stats_melt[variable == "n_cells"]

vec_bar_colors =c("grey10", "grey40", "grey70")# RColorBrewer::brewer.pal(name = "PuOr", n=3)
names(vec_bar_colors) = unique(dt_cell_clust_stats_melt$variable)

# dt_cell_clust_stats_melt$variable <- factor(dt_cell_clust_stats_melt$variable, levels = c("median genes", "median RNA", "cells"), ordered=T)


p_bar_cellcluster = ggplot(data=dt_cell_clust_stats_melt, aes(x=cluster_perslab_coarse)) +
  
  # geom_bar(stat = "identity", mapping= aes(y=value, col=variable), position=position_dodge())+
  # geom_bar(stat="identity", mapping=aes(y=n_cells),fill="red", position=position_dodge())+ 
  
  geom_bar(stat="identity", mapping=aes(y=value, fill=variable), position=position_dodge())+
  # geom_bar(stat="identity", position=position_dodge())+ 
    coord_flip() + 
    scale_fill_manual(values=vec_bar_colors, guide = guide_legend(reverse = TRUE)) +
    #  scale_y_continuous(
    #   # Features of the first axis
    #   name = "RNA",
    #   # Add a second axis and specify its features
    #   sec.axis = sec_axis( trans=~.*.1, name="cells")
    # ) +
  
    # scale_x_discrete() +  

  
  #scale_fill_discrete(breaks = "n cells", "median n genes","median_n_RNA")+
  
    theme(
      legend.title = element_blank(),
      legend.text = element_text(size=fontSize_legend_lg),
      legend.box.background = element_blank(),
      
      #legend.key.size =  unit(x = c(1,1), units="cm"),
      legend.position = c(1,1),
      axis.title.x = element_blank(),
     #axis.text.x = element_blank(),#element_text(angle=90),#element_blank(),
     axis.text.x = element_text(
        size= fontSize_axisText_sm,
        angle=35),
      axis.line.x = element_blank(),
     
     axis.title.y = element_blank(),
     axis.text.y = element_blank(),
     
     axis.ticks.y = element_blank(),
      
 
      #axis.ticks.x = element_blank(),
      panel.grid.major = element_blank(), 
      panel.grid.minor = element_blank(),
     plot.margin = margin(l=0, r = 2, b=1, unit="cm"),
     plot.background=element_blank(),  
     panel.background = element_blank()
     ) 
  p_bar_cellcluster 
  
  # ggsave(plot = p_var_pres_macp, filename = paste0(dirWGCNA_plots,prefixData,"_", prefixRun,  "_sc_mod_pres_in_macp_barplot_", params$date,".pdf"))#, height=, width =12)

p1D <- patchwork::wrap_plots(... = list(p1D, p_bar_cellcluster), 
                            ncol = 2
                            ) + 
  plot_layout(widths = c(12,2))

# p1D <- p1D + ggtitle("Normalized batch corrected counts")
# p1D <- p1D + theme(plot.title = element_text(size=fontSize_plotTitle)) 
p1D
# save as png 1500 * 700

# ggsave( plot= p_comb , filename = here("plots", paste0(prefixData, "_", prefixRun, "_Fig1D_markergene_vlnplot_",flagDate,".pdf")), width = length(vec_markers), height=length(unique(names(vec_markers)))/2.2)

```


<!-- ```{r} -->
<!-- p1BCD <- patchwork::wrap_plots(... = list(p1B, p1C, p1D),  -->
<!-- #                            ncol = 2 -->
<!--                             ) +  -->
<!--   plot_layout(widths = c(12,6)) -->

<!-- p1BCD -->
<!-- ``` -->



## Fig. 2 A. WGCNA method workflow figure

### i) coarse UMAP with highlighted cluster

```{r}
p2Ai <- DimPlot(seuratObj, 
             reduction = "umap",
             group.by="cluster_perslab_coarse", 
             seed=randomSeed, 
             label = F,
             #cols= vec_colors_cluster_perslab_coarse,
             cells.highlight =list(c(colnames(seuratObj)[seuratObj$cluster_perslab_coarse=="NK-like-cells"])),
             cols.highlight = c("#FFFF33"),

             #label.size = fontSize_label_md,
             #repel=T
             )
p2Ai <- p2Ai + Seurat::NoAxes() + Seurat::NoLegend()

p2Ai

here("plots", paste0(prefixData, "_", prefixRun, "_Fig2Ai_umap_highlight_",flagDate,".pdf"))


```

### ii) correlation matrix

```{r}

mat_rnorm = matrix(data = rnorm(n = 2500, mean = 3, sd=6), nrow = 50)

# introduce some covariance
for (i in 1:nrow(mat_rnorm)) {
  idx_col =sample(1:ncol(mat_rnorm), size = 15, replace = F)
  mat_rnorm[i, idx_col] <- mat_rnorm[i, idx_col] * rnorm(1,2,2)
}

mat_cor = cor(mat_rnorm)
mat_cor[row(mat_cor)==col(mat_cor)] <- 0
#dt_cor = data.table(mat_cor)
#dt_cor_melt = data.table::melt(dt_cor)
#dt_cor_melt = data.table("X"=1:nrow(dt_cor), "Y"=1:nrow(dt_cor), dt_cor_melt)

p_cor = pheatmap(mat_cor, 
                 cluster_rows = T, 
                 cluster_cols = T,
                 legend=F)#, treeheight_row=0, treeheight_col=0)

here("plots", paste0(prefixData, "_", prefixRun, "_Fig2Aii_corr_mat_",flagDate,".pdf"))


```

### Fig. 2 A) iii) and iv): networkplots

```{r}
n_clust = 5
vec_graph_cols = wes_palette("Darjeeling1",n_clust)
names(vec_graph_cols) = c(1:n_clust)
```

```{r}
g <- igraph::sample_islands(n_clust,40,0.6,20)
g <- igraph::simplify(g)
igraph::V(g)$grp <- as.character(rep(1:n_clust,each = 40))
igraph::V(g)$size = runif(n_clust*40, min = 0.1, max=0.5)
igraph::E(g)$col <- FALSE
#igraph::E(g)$col[bb$backbone] <- TRUE
```

```{r}
p_g1 = ggraph(g,layout = "stress")+
  geom_edge_link0(edge_colour = "grey",edge_width = 0.1, edge_alpha = 0.5)+
  geom_node_point(aes(fill = grp, colour = grp, size=size), shape = 21)+
  #scale_fill_brewer(palette = "RdYlGn")+
  scale_fill_manual(values=vec_graph_cols) +
  theme_graph()+
  theme(legend.position = "none")

ggsave( plot= p_g1 , filename = here("plots", paste0(prefixData, "_", prefixRun, "_Fig2Aiii_iv_vi_network_",flagDate,".pdf")))
```

<!-- set one cluster to white  -->

<!-- ```{r} -->
<!-- vec_graph_cols_changed= vec_graph_cols -->
<!-- vec_graph_cols_changed[3] <- "white" -->

<!-- p_g2 = ggraph(g,layout = "stress")+ -->
<!--   geom_edge_link0(edge_colour = "black",edge_width = 0.05, edge_alpha = 0.5)+ -->
<!--   geom_node_point(aes(fill = grp, colour = "white", size=size), shape = 21)+ -->
<!--   #scale_fill_brewer(palette = "RdYlGn")+ -->
<!--   scale_fill_manual(values=vec_graph_cols_changed) + -->
<!--   theme_graph()+ -->
<!--   theme(legend.position = "none") -->

<!-- ggsave( plot=p_g2 , filename = here("plots", paste0(prefixData, "_", prefixRun, "_Fig2A_network_filtered_",flagDate,".pdf"))) -->
<!-- ``` -->


<!-- ```{r} -->
<!-- #source: http://mr.schochastics.net/netVizR.html -->

<!-- n_clust = 5 -->

<!-- g <- igraph::sample_islands(n_clust,40,0.6,40) -->
<!-- g <- igraph::simplify(g) -->
<!-- igraph::V(g)$grp <- as.character(rep(1:n_clust,each = 40)) -->

<!-- #bb <- graphlayouts::layout_as_backbone(g,keep = 0.4) -->
<!-- igraph::E(g)$col <- FALSE -->
<!-- igraph::E(g)$col[bb$backbone] <- TRUE -->

<!-- ggraph::ggraph(g,layout = "manual",x = bb$xy[,1],y = bb$xy[,2])+ -->
<!--   geom_edge_link0(aes(edge_colour = col),edge_width = 0.1)+ -->
<!--   geom_node_point(aes(fill = grp),shape = 21)+ -->
<!--   scale_fill_brewer(palette = "Set1")+ -->
<!--   scale_edge_color_manual(values=c(rgb(0,0,0,0.3),rgb(0,0,0,1)))+ -->
<!--   theme_graph()+ -->
<!--   theme(legend.position = "none") -->
<!-- ``` -->

<!-- ```{r} -->
<!-- set_graph_style() -->

<!-- # visualize communities of nodes -->
<!-- dolphin %>%  -->
<!--   activate(nodes) %>% -->
<!--   mutate(community = as.factor(group_louvain())) %>%  -->
<!--   ggraph() +  -->
<!--   geom_edge_link() +  -->
<!--   geom_node_point(aes(colour = community), size = 5) -->
<!-- ``` -->

<!-- ```{r} -->
<!-- vec_mods_plot_nw = c("Endothelial-cells_2","Endothelial-cells_5", "NK-like-cells_1")#,"Cholangiocytes_5","Cholangiocytes_6") -->
<!-- size = 30 -->
<!-- vec_mods_plot_nw_cols = wes_palette("FantasticFox1", n=length(vec_mods_plot_nw)) -->
<!-- names(vec_mods_plot_nw_cols) = vec_mods_plot_nw -->
<!-- ``` -->

<!-- ```{r} -->
<!-- ### NETWORK PLOT ####  -->

<!-- # get network kIM values -->
<!-- list_vec_mod_genes <- lapply(vec_mods_plot_nw, function(eachmod) { -->
<!--   vec_out = dt_geneMod_perslab[module_renamed == eachmod,"genes", with=F][[1]][1:min(size,length(dt_geneMod_perslab[module_renamed == eachmod,"genes", with=F][[1]]))] -->
<!--   names(vec_out) = rep(eachmod, length(vec_out)) -->
<!--   vec_out  -->
<!--   })  -->

<!-- list_vec_mod_kIMs <- lapply(vec_mods_plot_nw, function(eachmod) { -->
<!--   vec_out = dt_geneMod_perslab[module_renamed == eachmod,"pkMs", with=F][[1]][1:min(size,length(dt_geneMod_perslab[module_renamed == eachmod,"pkMs", with=F][[1]]))] -->
<!--   names(vec_out) = rep(eachmod, length(vec_out)) -->
<!--   # normalize kIMs -->
<!--   vec_out/sum(vec_out) -->

<!--   })  -->

<!-- # any overlap between modules? -->
<!-- vec_all_nw_mod_genes = unlist(list_vec_mod_genes, FALSE) -->
<!-- vec_all_nw_mod_kIMs = unlist(list_vec_mod_kIMs, FALSE) -->
<!-- any(duplicated(vec_all_nw_mod_genes)) -->
<!-- # TRUE -->

<!-- vec_logical_unique = !duplicated(vec_all_nw_mod_genes) -->

<!-- vec_all_nw_mod_genes = vec_all_nw_mod_genes[vec_logical_unique] -->
<!-- vec_all_nw_mod_kIMs = vec_all_nw_mod_kIMs[vec_logical_unique] -->
<!-- # order -->
<!-- # vec_mod_genes = vec_mod_genes[order(vec_mod_kIMs, decreasing = T)] -->
<!-- # vec_mod_kIMs = sort(vec_mod_kIMs, decreasing = T) -->

<!-- # take hubgenes -->
<!-- # vec_mod_genes = vec_mod_genes[1:min(size, length(vec_mod_genes))] -->
<!-- # vec_mod_kIMs = vec_mod_kIMs[1:min(size, length(vec_mod_kIMs))] -->

<!-- # subset expression data celltypes -->
<!-- mat_datExpr = subset(x=seuratObj, idents=gsub("_\\d$", "", vec_mods_plot_nw)) %>% GetAssayData(., assay="SCT_combat") %>% as.matrix -->

<!-- # subset expression data module genes -->
<!-- mat_datExpr <- mat_datExpr[vec_all_nw_mod_genes,] -->
<!-- # Compute network adjacency -->
<!-- mat_adj <- WGCNA::adjacency(t(mat_datExpr),  -->
<!--                        power = 16,  -->
<!--                        corFnc = "cor",  -->
<!--                        corOptions = list(use = "p"), -->
<!--                        type = "signed") -->
<!-- # vec_hub_genes <- vec_mod_genes[order(vec_mod_kIMs, decreasing=T)] -->
<!-- # vec_hub_kIMs <- vec_mod_kIMs[order(vec_mod_kIMs, decreasing=T)] -->


<!-- df_hub.data <- cbind.data.frame(module = names(vec_all_nw_mod_genes), -->
<!--                            gene = vec_all_nw_mod_genes,  -->
<!--                            kIM = vec_all_nw_mod_kIMs) -->

<!-- # hub.data$logfc <- DEGs$log2FoldChange[match(hub.data$gene, rownames(DEGs))] -->
<!-- # hub.data$p.adj <- DEGs$padj[match(hub.data$gene, rownames(DEGs))] -->
<!-- mat_adj %>%  -->
<!-- graph.adjacency(mode = "undirected", weighted = T, diag = FALSE) %>%  -->
<!-- as_tbl_graph() %>% upgrade_graph() %>% activate(nodes) %>% -->
<!-- dplyr::mutate(module = df_hub.data$module) %>% -->
<!-- dplyr::mutate(kIM = df_hub.data$kIM) %>%  -->
<!-- #dplyr::mutate(abs.logfc = abs(hub.data$logfc)) %>% -->
<!-- #dplyr::mutate(p.adj = hub.data$p.adj) %>%  -->
<!-- #dplyr::mutate(color=factor(ifelse(p.adj < 0.05, yes = "S", no = "NS"),  -->
<!-- #                           levels = c("NS", "S"))) %>% -->
<!--   activate(edges) %>%  -->
<!--   #dplyr::filter(weight>.15) %>%  -->
<!--   activate(nodes) ->hub.plot -->
<!-- #filter(!node_is_isolated()) -> hub.plot -->

<!-- #hub.plot$mod_cols = vec_mods_plot_nw_cols[match(df_hub.data$module, names(vec_mods_plot_nw_cols))] -->
<!-- # Plot -->
<!-- # layouts: https://rdrr.io/cran/ggraph/man/layout_tbl_graph_igraph.html -->
<!-- module.plot <- ggraph(hub.plot,  -->
<!--                       #layout = "igraph", -->
<!--                       #algorithm = "drl", -->
<!--                       layout = "stress" -->
<!--                       ) +  -->
<!--   geom_edge_link(color="darkgrey", show.legend=F, aes(alpha=weight)) +  -->
<!--   geom_node_point(aes(size = kIM),  -->
<!--                   #fill =,  -->
<!--                   shape=21,  -->
<!--                   alpha=0.8) +  -->
<!--   #scale_fill_manual(values= vec_mods_plot_nw_cols) + -->
<!--   # scale_size(breaks = c(1,2,3),  -->
<!--   #            limits = c(0,4),  -->
<!--   #            range = c(0, 8),  -->
<!--   #            labels = c(" \U00B1 1", " \U00B1 2", " \U00B1 3") -->
<!--   #            ) + -->
<!--   #geom_node_text(aes(label = name), fontface="bold", size=5, repel = T, color = "black") + -->
<!--   # guides( -->
<!--   #   size = guide_legend( -->
<!--   #     override.aes = list( -->
<!--   #   size=c(2,4,6)),  -->
<!--   #   keywidth = 0.8,  -->
<!--   #   keyheight = 0.8, order = 1) -->
<!--   #   ) +  -->
<!--                              #title = expression(bold(paste("Log"[2],~ -->
<!--                                                             #"fold-change"))))) +  -->
<!--   theme_graph(base_family = 'Helvetica') +  -->
<!--   theme(legend.title.align=0.5, -->
<!--         legend.position = "top", -->
<!--         legend.margin = margin(0,0,0,0, unit="cm"), -->
<!--         legend.title = element_text(size=8, face="bold"), -->
<!--         legend.text = element_text(size=8, face="bold"), -->
<!--         axis.text =element_blank(), -->
<!--         axis.ticks = element_blank(),  -->
<!--         axis.line=element_blank(), -->
<!--         axis.title = element_blank(), -->
<!--         legend.spacing.x = unit(0, "cm"), -->
<!--         # margin: top, right, bottom, and left -->
<!--         plot.margin = unit(c(0, 0, 0, 0), "cm")) + -->
<!--   coord_cartesian(clip="off") -->

<!-- module.plot -->


<!-- ``` -->


### Fig. 2 A v boxplot of module activity in patients

patient metadata

```{r}
dt_metadata <- fread(here("data", "liver_gerhard2018_patient_metadata_qc.csv.gz"))

```

### embeddings

this includes sc and bulk modules

```{r}
# df_embed = fread("/projects/jonatan/pub-perslab/18-liver-wgcna/tables/liver_perslab_int_wgcna3_mod_embed_gerhard2018.csv.gz") %>% setDF
df_scaled_embed = fread("/projects/jonatan/pub-perslab/18-liver-wgcna/tables/liver_perslab_int_wgcna3_mat_scaled_embed_gerhard2018.csv.gz") %>% setDF

mat_scaled_embed = df_scaled_embed[,-1] %>% as.matrix
rownames(mat_scaled_embed) = df_scaled_embed$run_accession
mat_scaled_embed[0:4,0:4]
```

```{r}
vec_mods_box = c("Cholangiocytes_5","Dendritic-cells_1")

```

```{r}

dt_boxplot = data.table(dt_metadata[,.(DCL.Patient.ID,Diagnosis)], "module"=mat_scaled_embed[,vec_mods_box])

#colnames(dt_boxplot) <- gsub("\\.Cholangiocytes_5|\\.Dendritic-cells_1","",colnames(dt_boxplot))
# dt_plot_sc = data.table("sample_ID" = seuratObj$sample_ID, 
#                           "fibrosis_grade"=seuratObj$fibrosis_grade,
#                           "module_activity"=log2(seuratObj@meta.data$`Hepatic-stellate-cells_4`+1))
dt_boxplot <- dt_boxplot[grepl("Fibrosis 3|NORMAL", Diagnosis)]
dt_boxplot <- dt_boxplot[!grepl("Fibrosis 3/4", Diagnosis)]
dt_boxplot <- melt(dt_boxplot,id.vars = c("DCL.Patient.ID", "Diagnosis"))

dt_boxplot$variable = gsub("module\\.","",dt_boxplot$variable)
colnames(dt_boxplot)[c(3,4)] <- c("module","activity")
dt_boxplot$Diagnosis = factor(dt_boxplot$Diagnosis, levels = c("NORMAL","Fibrosis 3"), ordered=TRUE)
#dt_boxplot$Diagnosis = factor(dt_boxplot$Diagnosis, levels = c("NORMAL","Fibrosis 3", "Fibrosis 3/4", "Fibrosis 4"), ordered=TRUE)
#dt_plot_box_subsample = dt_plot_box[,lapply(.SD, function(x) {head(x,n=3)}), by=Diagnosis]

```

```{r}
library(colorspace)
```

```{r}
for (i in 1:length(vec_mods_box)) {
  mod_plot = vec_mods_box[i]
  dt_boxplot_sub = dt_boxplot[module==mod_plot,colnames(dt_boxplot),with=F,]
  col_plot = vec_graph_cols[i]
  vec_cols_plot <- c(lighten(col = col_plot, amount = 0.3),darken(col = col_plot, amount = 0.3))
  names(vec_cols_plot) = levels(dt_boxplot_sub$Diagnosis)
  p_box = ggplot(data=dt_boxplot_sub, mapping=aes(x=Diagnosis, y=activity, fill=Diagnosis)) +
    geom_boxplot()  +
    #facet_grid(rows = module) + 
    scale_fill_manual(values=vec_cols_plot) +
    geom_jitter(width=0.1,alpha=0.5) +
    theme(legend.position = NULL)+
    theme(
        axis.text = element_text(size = fontSize_axisText_md), 
        axis.line = element_blank(), 
        axis.ticks = element_blank(), 
        axis.title = element_text(size = fontSize_axisTitle_md),#, face = "bold"), 
        panel.background = element_blank(), 
        #legend.position = "none",
        #panel.grid.minor = element_blank(), 
        #panel.grid.major = element_line(colour = "grey80"), 
        plot.background = element_blank(), 
        panel.grid = element_blank(),
        #text = element_text(size = ),
        legend.title=element_blank(),
        legend.text = element_text(size=fontSize_legend_lg),
        aspect.ratio = 1.5) + 
    labs(y="module activity")
  
  ggsave(plot = p_box, filename = paste0(dirWGCNA_plots,prefixData,"_", prefixRun,  "_fig2Av_boxplot_",i, "_", flagDate,".pdf"))
}
```

diagnostic scatterplots. Not for publication:

```{r}
plot(dt_boxplot[,list(SEX, BMI_surg, Diagnosis, Age, module)])
```


### Fig. 2 A) vi: large networkplots for clustering

```{r}
n_clust = 6
vec_graph_cols = c(wes_palette("Darjeeling1",5),wes_palette("Darjeeling1",(n_clust-5)))
names(vec_graph_cols) = c(1:n_clust)
```

```{r}
g2 <- igraph::sample_islands(n_clust,40,0.6,20)
#g2 <- igraph::simplify(g)
igraph::V(g2)$grp <- as.character(rep(1:n_clust,each = 40))
igraph::V(g2)$size = runif(n_clust*40, min = 0.1, max=0.5)
igraph::E(g2)$col <- FALSE
igraph::E(g2)$col[bb$backbone] <- TRUE
```

```{r}
p_g2 = ggraph(g2,layout = "stress")+
  geom_edge_link0(edge_colour = "grey",edge_width = 0.1, edge_alpha = 0.5)+
  geom_node_point(aes(fill = grp, colour = grp, size=size), shape = 21)+
  #scale_fill_brewer(palette = "RdYlGn")+
  scale_fill_manual(values=vec_graph_cols) +
  theme_graph()+
  theme(legend.position = "none")

ggsave(plot = p_g2 , filename = here("plots", paste0(prefixData, "_", prefixRun, "_Fig2Aiii_vi_network_",flagDate,".pdf")))
```





<!--  dendrogram (of mac mods) -->

<!-- ```{r} -->
<!-- library("WGCNA") -->
<!-- library("Seurat") -->
<!-- ``` -->

<!-- ```{r} -->
<!-- datExpr = subset(x=seuratObj, subset = cluster_perslab_coarse == "Macrophages") %>% GetAssayData(., assay = "SCT_combat", slot = "data") %>% as.matrix %>% t -->
<!-- ``` -->

<!-- subset to keep variable genes -->

<!-- ```{r} -->
<!-- seuratObj <- FindVariableFeatures(seuratObj, nfeatures = 2000) -->
<!-- ``` -->

<!-- ```{r} -->
<!-- datExpr = datExpr[,VariableFeatures(seuratObj)] -->
<!-- ``` -->

<!-- ```{r} -->
<!-- dim(datExpr) -->
<!-- # [1] 3613 2000 -->
<!-- ``` -->

<!-- ```{r} -->
<!-- powers = c(c(1:10), seq(from = 12, to=20, by=2)) -->
<!-- ``` -->

<!-- ```{r} -->
<!-- sft = pickSoftThreshold(datExpr, powerVector = powers, verbose = 5) -->
<!-- ``` -->

<!-- plot softpower scale free fits -->

<!-- ```{r} -->
<!-- sizeGrWindow(9, 5) -->
<!-- par(mfrow = c(1,2)); -->
<!-- cex1 = 0.9; -->
<!-- plot(sft$fitIndices[,1], -sign(sft$fitIndices[,3])*sft$fitIndices[,2], -->
<!-- xlab="Soft Threshold (power)",ylab="Scale Free Topology Model Fit,signed R^2",type="n", -->
<!-- main = paste("Scale independence")); -->
<!-- text(sft$fitIndices[,1], -sign(sft$fitIndices[,3])*sft$fitIndices[,2], -->
<!-- labels=powers,cex=cex1,col="red"); -->
<!-- ``` -->

<!-- mean connectivity -->
<!-- ```{r} -->
<!-- plot(sft$fitIndices[,1], sft$fitIndices[,5], -->
<!-- xlab="Soft Threshold (power)",ylab="Mean Connectivity", type="n", -->
<!-- main = paste("Mean connectivity")) -->
<!-- text(sft$fitIndices[,1], sft$fitIndices[,5], labels=powers, cex=cex1,col="red") -->
<!-- ``` -->

<!-- construct network  -->

<!-- ```{r} -->
<!-- net = blockwiseModules(datExpr, power = 2, -->
<!--                        TOMType = "signed", -->
<!--                        minModuleSize = 20, -->
<!--                        reassignThreshold = 0, -->
<!--                        mergeCutHeight = 0.25, -->
<!--                        numericLabels = TRUE,  -->
<!--                        pamStage=FALSE, -->
<!--                        #pamRespectsDendro = FALSE, -->
<!--                        saveTOMs = FALSE, -->
<!--                        verbose = 3) -->
<!-- ``` -->

<!-- plot dendro -->

<!-- ```{r} -->

<!-- mergedColors = labels2colors(net$colors) -->
<!-- pdf(file = paste0(dirWGCNA_plots, prefixData, "_", prefixRun, "_wgcna_dendroplot_macs.pdf")) -->
<!-- # open a graphics window -->
<!-- #sizeGrWindow(12, 9) -->
<!-- # Convert labels to colors for plotting -->

<!-- # Plot the dendrogram and the module colors underneath -->
<!-- plotDendroAndColors(net$dendrograms[[1]], mergedColors[net$blockGenes[[1]]], -->
<!-- "Module colors",dendroLabels = FALSE, hang = 0.03,addGuide = TRUE, guideHang = 0.05) -->
<!-- dev.off() -->
<!-- ``` -->

<!-- ### iii. show module preservation and filtering -->

<!-- ```{r} -->
<!-- vec_mod_dummy = runif(25,1,2) -->
<!-- vec_mod_dummy_corrupted = sapply(vec_mod_dummy, function(x) x*runif(1,0.8,1.2)) -->
<!-- mat_pres_dummy = matrix(data= c(vec_mod_dummy,vec_mod_dummy_corrupted), ncol=2) -->
<!-- ``` -->

<!-- ```{r} -->
<!-- pdf(file = here("plots", paste0(prefixData, "_" , prefixRun, "_", prefixOut, "_pres_heatmap_dummy_",flagDate,".pdf")), width=2, height=10) -->

<!-- pheatmap(mat_pres_dummy,  -->
<!--          cluster_rows = F,  -->
<!--          cluster_cols = F,  -->
<!--          color =  colorRampPalette(brewer.pal(n=9, name="Oranges"))(20), -->
<!--          #width=2, -->
<!--          #height=12 -->
<!--          ) -->

<!-- dev.off() -->
<!-- ``` -->


## Check for module enrichment with GWAS hit genes

```{r}

df_GWAS = openxlsx::read.xlsx(here("data","All nominated genes from NAFLD GWAS.xlsx"))

vec_genes_NAFLD = df_GWAS$Nominated.Gene
```

clean up crappy data

```{r}
list_genes_NAFLD = as.list(vec_genes_NAFLD)
list_vec_genes_NAFLD = lapply(list_genes_NAFLD, function(eachstr) {
  vec_genes = strsplit(x=eachstr, split=";")[[1]]
})

vec_genes_NAFLD = unlist(list_vec_genes_NAFLD, use.names = F)
vec_genes_NAFLD = vec_genes_NAFLD[nchar(vec_genes_NAFLD)>1]
vec_genes_NAFLD = toupper(vec_genes_NAFLD)
```

```{r}
vec_mods_perslab = unique(dt_geneMod_perslab[["module_renamed"]]) 
vec_mods_perslab = vec_mods_perslab[nchar(vec_mods_perslab)>0]
any(is.na(vec_mods_perslab))
# FALSE
```


```{r}
vec_mods_gerhard = unique(dt_geneMod_gerhard[["module_renamed"]]) 
vec_mods_gerhard = vec_mods_gerhard[nchar(vec_mods_gerhard)>0]
any(is.na(vec_mods_gerhard))
# FALSE
```


```{r}

list_geneWeights_perslab = lapply(vec_mods_perslab, function(mod) {
  condition = quote(dt_geneMod_perslab[["module_renamed"]]==mod)
  vec_mod_geneWeights = dt_geneMod_perslab[eval(condition), pkMs,]
  names(vec_mod_geneWeights) = dt_geneMod_perslab[eval(condition), ..colGeneNames, ][[1]]
  return(vec_mod_geneWeights)
})

names(list_geneWeights_perslab) = vec_mods_perslab

```



```{r}

list_geneWeights_gerhard = lapply(vec_mods_gerhard, function(mod) {
  condition = quote(dt_geneMod_gerhard[["module_renamed"]]==mod)
  vec_mod_geneWeights = dt_geneMod_gerhard[eval(condition), pkMs,]
  names(vec_mod_geneWeights) = dt_geneMod_gerhard[eval(condition), ..colGeneNames, ][[1]]
  return(vec_mod_geneWeights)
})

names(list_geneWeights_gerhard) = vec_mods_gerhard
```

```{r}
mat_NAFLD_genes_in_mod_perslab = sapply(list_geneWeights_perslab, function(vec_geneWeights) {
   vec_geneWeights_rank = order(vec_geneWeights, decreasing = F)/length(vec_geneWeights)
   names(vec_geneWeights_rank) = names(vec_geneWeights)
    ifelse(vec_genes_NAFLD %in% names(vec_geneWeights_rank), vec_geneWeights_rank, 0)
  })

rownames(mat_NAFLD_genes_in_mod_perslab) <- vec_genes_NAFLD
```


```{r}
vec_mod_hypergeom_perslab_pval = sapply(list_geneWeights_perslab, function(vec_geneWeights) {
  phyper(q=sum(vec_genes_NAFLD %in% names(vec_geneWeights)),
         m=length(vec_genes_NAFLD),
         n=length(vec_genes_perslab),
         k=length(vec_geneWeights), lower.tail = F)
  #vec_geneWeights_rank = order(vec_geneWeights, decreasing = F)/length(vec_geneWeights)
  })
```


```{r}
vec_mod_hypergeom_gerhard_pval = sapply(list_geneWeights_gerhard, function(vec_geneWeights) {
  phyper(q=sum(vec_genes_NAFLD %in% names(vec_geneWeights)),
         m=length(vec_genes_NAFLD),
         n=length(vec_genes_gerhard),
         k=length(vec_geneWeights), lower.tail = F)
  #vec_geneWeights_rank = order(vec_geneWeights, decreasing = F)/length(vec_geneWeights)
  })
```

```{r}
vec_mod_hypergeom_pval_BH = p.adjust(p=c(vec_mod_hypergeom_perslab_pval,vec_mod_hypergeom_gerhard_pval), method = "BH")

vec_mod_hypergeom_pval_BH <- sort(vec_mod_hypergeom_pval_BH)

vec_mod_hypergeom_pval_BH
```

filter out modules without any GWAS genes

```{r}
mat_NAFLD_genes_in_mod_perslab = mat_NAFLD_genes_in_mod_perslab[,apply(mat_NAFLD_genes_in_mod_perslab, MARGIN=2, FUN=function(eachcol){any(eachcol>0)})]
```

and genes not present in any module
```{r}
mat_NAFLD_genes_in_mod_perslab = mat_NAFLD_genes_in_mod_perslab[apply(mat_NAFLD_genes_in_mod_perslab, MARGIN=1, FUN=function(eachrow){any(eachrow>0)}),]
```

rename modules

```{r}
vec_mod_number = gsub("[^0-9]","",colnames(mat_NAFLD_genes_in_mod_perslab))
colnames(mat_NAFLD_genes_in_mod_perslab) = paste0("M-",vec_celltype_abbrev[match(gsub("_\\d","",colnames(mat_NAFLD_genes_in_mod_perslab)),names(vec_celltype_abbrev))], "-", vec_mod_number)

```

```{r}
pdf(file=here("plots", paste0(prefixData, "_", prefixRun, "_GWAS_genes_in_perslab_mods_renamed_heatmap_",flagDate,".pdf")), width=2.5, height = 4)
p_heatmap_mat_NAFLD_genes_in_mod_perslab = pheatmap(mat_NAFLD_genes_in_mod_perslab, 
                 cluster_rows = F, 
                 cluster_cols = T,
                 fontsize_row = 10, #default 10
                 legend=T, 
                 treeheight_row=0, 
                 treeheight_col=0)

dev.off()
#here("plots", paste0(prefixData, "_", prefixRun, "_Fig2Aii_corr_mat_",flagDate,".pdf"))

```

```{r}
openxlsx::write.xlsx(data.table("gene" = rownames(mat_NAFLD_genes_in_mod_perslab),mat_NAFLD_genes_in_mod_perslab),  paste0(dirWGCNA_tables,prefixData, "_", prefixRun, "_GWAS_genes_in_perslab_mods_table_",flagDate,".xlsx"))
```

```{r}
mat_NAFLD_genes_in_mod_gerhard = sapply(list_geneWeights_gerhard, function(vec_geneWeights) {
   vec_geneWeights_rank = order(vec_geneWeights, decreasing = F)/length(vec_geneWeights)
   names(vec_geneWeights_rank) = names(vec_geneWeights)
    ifelse(vec_genes_NAFLD %in% names(vec_geneWeights_rank), vec_geneWeights_rank, 0)
  })

rownames(mat_NAFLD_genes_in_mod_gerhard) <- vec_genes_NAFLD
```

filter out modules without any GWAS genes

```{r}
mat_NAFLD_genes_in_mod_gerhard = mat_NAFLD_genes_in_mod_gerhard[,apply(mat_NAFLD_genes_in_mod_gerhard, MARGIN=2, FUN=function(eachcol){any(eachcol>0)})]
```
and genes not present in any module
```{r}
mat_NAFLD_genes_in_mod_gerhard = mat_NAFLD_genes_in_mod_gerhard[apply(mat_NAFLD_genes_in_mod_gerhard, MARGIN=1, FUN=function(eachrow){any(eachrow>0)}),]
```

rename modules
```{r}
colnames(mat_NAFLD_genes_in_mod_gerhard) = gsub("gerhard2018_","M-Ger-",colnames(mat_NAFLD_genes_in_mod_gerhard))
```
 

```{r}
pdf(file=here("plots", paste0(prefixData, "_", prefixRun, "_GWAS_genes_in_gerhard_mods_renamed_heatmap_",flagDate,".pdf")), width=2, height = 2.2)
p_heatmap_mat_NAFLD_genes_in_mod_gerhard = pheatmap(mat_NAFLD_genes_in_mod_gerhard, 
                 cluster_rows = F, 
                 cluster_cols = T,
                 fontsize_row = 10, #default 10
                 legend=T, 
                 treeheight_row=0, 
                 treeheight_col=0)

dev.off() 
#here("plots", paste0(prefixData, "_", prefixRun, "_Fig2Aii_corr_mat_",flagDate,".pdf"))

```


```{r}
openxlsx::write.xlsx(data.table("gene" = rownames(mat_NAFLD_genes_in_mod_gerhard),mat_NAFLD_genes_in_mod_gerhard),  paste0(dirWGCNA_tables,prefixData, "_", prefixRun, "_GWAS_genes_in_gerhard_mods_table_",flagDate,".xlsx"))
```

## Check module enrichment for DE proteins

```{r}
vec_prot_DE = c("HSPA1B",
"CDKN1B",
"MVK",
"COL18A1",
"MSN",
"GRHPR",
"HPGD",
"PRKAR1A",
"HSD17B2",
"GPD1",
"RDX",
"DAP",
"TP53I3",
"GLYCTK",
"OLA1",
"IDH2",
"CYC1",
"ACLY",
"APOL3",
"VAPA",
"CYP3A4",
"PLIN2",
"ARCN1",
"GCHFR",
"MAT1A",
"HSD17B11",
"AHCY",
"SFXN2",
"PLIN1",
"SEC24A",
"RBBP9",
"SLC25A42")
```

# get modules

```{r}
vec_mods_perslab = unique(dt_geneMod_perslab[["module_renamed"]]) 
vec_mods_perslab = vec_mods_perslab[nchar(vec_mods_perslab)>0]
any(is.na(vec_mods_perslab))
# FALSE
```

```{r}
vec_mods_gerhard = unique(dt_geneMod_gerhard[["module_renamed"]]) 
vec_mods_gerhard = vec_mods_gerhard[nchar(vec_mods_gerhard)>0]
any(is.na(vec_mods_gerhard))
# FALSE
```

```{r}
list_geneWeights_perslab = lapply(vec_mods_perslab, function(mod) {
  condition = quote(dt_geneMod_perslab[["module_renamed"]]==mod)
  vec_mod_geneWeights = dt_geneMod_perslab[eval(condition), pkMs,]
  names(vec_mod_geneWeights) = dt_geneMod_perslab[eval(condition), ..colGeneNames, ][[1]]
  return(vec_mod_geneWeights)
})
names(list_geneWeights_perslab) = vec_mods_perslab
```

```{r}
list_geneWeights_gerhard = lapply(vec_mods_gerhard, function(mod) {
  condition = quote(dt_geneMod_gerhard[["module_renamed"]]==mod)
  vec_mod_geneWeights = dt_geneMod_gerhard[eval(condition), pkMs,]
  names(vec_mod_geneWeights) = dt_geneMod_gerhard[eval(condition), ..colGeneNames, ][[1]]
  return(vec_mod_geneWeights)
})
names(list_geneWeights_gerhard) = vec_mods_gerhard
```

```{r}
list_geneWeights_comb = c(list_geneWeights_perslab, list_geneWeights_gerhard)
```

```{r}
mat_prot_DE_in_mod_comb = sapply(list_geneWeights_comb, function(vec_geneWeights) {
   vec_geneWeights_rank = order(vec_geneWeights, decreasing = F)/length(vec_geneWeights)
   names(vec_geneWeights_rank) = names(vec_geneWeights)
    ifelse(vec_prot_DE %in% names(vec_geneWeights_rank), vec_geneWeights_rank, 0)
  })
rownames(mat_prot_DE_in_mod_comb) <- vec_prot_DE
```

```{r}
vec_mod_hypergeom_comb_pval = sapply(list_geneWeights_comb, function(vec_geneWeights) {
  phyper(q=sum(vec_prot_DE %in% names(vec_geneWeights)),
         m=length(vec_prot_DE),
         n=length(union(vec_genes_perslab, vec_genes_gerhard)),
         k=length(vec_geneWeights), lower.tail = F)
  #vec_geneWeights_rank = order(vec_geneWeights, decreasing = F)/length(vec_geneWeights)
  })
```


```{r}
vec_mod_hypergeom_pval_BH = p.adjust(p=vec_mod_hypergeom_comb_pval, method = "BH")

vec_mod_hypergeom_pval_BH <- sort(vec_mod_hypergeom_pval_BH)

vec_mod_hypergeom_pval_BH
```

filter out modules without any DE proteins

```{r}
mat_prot_DE_in_mod_comb = mat_prot_DE_in_mod_comb[,apply(mat_prot_DE_in_mod_comb, MARGIN=2, FUN=function(eachcol){any(eachcol>0)})]

# and genes not present in any module
mat_prot_DE_in_mod_comb = mat_prot_DE_in_mod_comb[apply(mat_prot_DE_in_mod_comb, MARGIN=1, FUN=function(eachrow){any(eachrow>0)}),]
```

rename modules

```{r}
vec_mod_number = gsub("2018","",colnames(mat_prot_DE_in_mod_comb))
vec_mod_number = gsub("[^0-9]","",vec_mod_number)

colnames(mat_prot_DE_in_mod_comb) = paste0("M-", ifelse(grepl("gerhard",colnames(mat_prot_DE_in_mod_comb)), no=vec_celltype_abbrev[match(gsub("_\\d","",colnames(mat_prot_DE_in_mod_comb)),names(vec_celltype_abbrev))], yes = "Ger"), "-", vec_mod_number)
```

```{r}
pdf(file=here("plots", paste0(prefixData, "_", prefixRun, "_DE_prot_in_comb_mods_renamed_heatmap_",flagDate,".pdf")), width=2.5, height = 4)
p_heatmap_mat_prot_DE_in_mod_comb = pheatmap(mat_prot_DE_in_mod_comb, 
                 cluster_rows = F, 
                 cluster_cols = T,
                 fontsize_row = 10, #default 10
                 legend=T, 
                 treeheight_row=0, 
                 treeheight_col=0)

dev.off()
#here("plots", paste0(prefixData, "_", prefixRun, "_Fig2Aii_corr_mat_",flagDate,".pdf"))

```

```{r}
openxlsx::write.xlsx(data.table("gene" = rownames(mat_prot_DE_in_mod_perslab),mat_prot_DE_in_mod_perslab),  paste0(dirWGCNA_tables,prefixData, "_", prefixRun, "_GWAS_genes_in_perslab_mods_table_",flagDate,".xlsx"))
```


## Fig 2(B) GO barplot + network plot for modules of interest

```{r}
dt_resultsGO = fread(paste0(dirWGCNA_tables, prefixData, "_", prefixRun, "_", "201216","_GERR_GOenrichmentResults.csv"))
dt_resultsGO$source = "GO"

# KEGG
dt_resultsKEGG = fread(paste0(dirWGCNA_tables, prefixData, "_", prefixRun,  "_", "201216","_GERR_KEGGenrichmentResults.csv"))
dt_resultsKEGG$source = "KEGG"

# REACTOME
dt_resultsREACT = fread(paste0(dirWGCNA_tables, prefixData, "_", prefixRun,  "_","201216" ,"_GERR_REACTOMEenrichmentResults.csv"))
dt_resultsREACT$source = "REACTOME"
```


```{r}
Idents(seuratObj) = "cluster_perslab_coarse"
```

< for each module, manually feed GO ids and p-values for HSC4 into REVIGO with all default parameters >

```{r}
size = 40
#mods_plot_all_genes = c("Hepatocytes_1","Macrophages_1")
mods_plot =c(
  "Hepatic-stellate-cells_1", 
  "Endothelial-cells_1", 
  "Macrophages_1", 
  "Macrophages_3", 
  "Hepatocytes_1")#, "Hepatocytes_4")
#mods_plot =c("Macrophages_1", "Macrophages_3")#, "Hepatocytes_4")

```

```{r}
for (mod_plot in mods_plot) {
  # 
  # if (mod_plot %in% mods_plot_all_genes) {
  #   size <- sum(dt_geneMod_perslab$module_renamed==mod_plot)
  # } else {
  #     size = defaultsize
  #   }
  # 
  if (file.exists(paste0(dirWGCNA_tables, prefixData, "_", prefixRun, "_", "201216","_GERR_GOenrichmentResults_revigo_", mod_plot,".csv"))) {
    dt_resultsGO_revigo = fread(paste0(dirWGCNA_tables, prefixData, "_", prefixRun, "_", "201216","_GERR_GOenrichmentResults_revigo_", mod_plot,".csv"))

    dt_resultsGO_revigo$source = "GO"

    dt_resultsGO_revigo = dt_resultsGO_revigo[eliminated == 0, .(term =description,source)]
  
    dt_resultsGO_revigo$minlog10_p.value_fisher_BH = -log10(dt_resultsGO$p.value_fisher_BH[match(dt_resultsGO_revigo$term, dt_resultsGO$GO_term)])
    
      # reorder and take top five
  dt_resultsGO_revigo = dt_resultsGO_revigo[order(-minlog10_p.value_fisher_BH)][1:min(5,nrow(dt_resultsGO_revigo))]
  
  }
  
   condition=quote(dt_resultsKEGG$module==mod_plot)
    dt_resultsKEGG_plot = dt_resultsKEGG[eval(condition),.(term =KEGG_term, minlog10_p.value_fisher_BH=-log10(p.value_fisher_BH),source)]
  if (mod_plot %in% dt_resultsKEGG$module) {
    # reorder and take top five
  dt_resultsKEGG_plot = dt_resultsKEGG_plot[order(-minlog10_p.value_fisher_BH)][1:min(5,nrow(dt_resultsKEGG_plot))]
  }
  
  
  condition=quote(dt_resultsREACT$module==mod_plot)
  dt_resultsREACT_plot = dt_resultsREACT[eval(condition),.(term =REACT_name, minlog10_p.value_fisher_BH=-log10(p.value_fisher_BH),source)]
  if (mod_plot %in% dt_resultsREACT$module) {
  # reorder and take top five
  dt_resultsREACT_plot = dt_resultsREACT_plot[order(-minlog10_p.value_fisher_BH)][1:min(5,nrow(dt_resultsREACT_plot))]
  }
  
  list_GERR = list()
  if (file.exists(paste0(dirWGCNA_tables, prefixData, "_", prefixRun, "_", "201216","_GERR_GOenrichmentResults_revigo_", mod_plot,".csv"))) { 
    list_GERR[["GO_revigo"]] <- dt_resultsGO_revigo
  }
  
  list_GERR[["KEGG"]] <- dt_resultsKEGG_plot
  list_GERR [["REACTOME"]] <- dt_resultsREACT_plot
  dt_results_GERR_plot = rbindlist(list_GERR, use.names=T) # columns aren't in the same order in REVIGO, hence match by colnames
  #dt_results_GERR_plot = dt_results_GERR_plot[order(source,-minlog10_p.value_fisher_BH)]

  # filter out non-significant
  dt_results_GERR_plot = dt_results_GERR_plot[minlog10_p.value_fisher_BH>-log10(0.05)]
  
  if (nrow(dt_results_GERR_plot)>0){ 
    
    
     ### truncate terms if required
  dt_results_GERR_plot$truncate = sapply(dt_results_GERR_plot$term, function(x){
      nchar(x)>55
      })
  
  
  dt_results_GERR_plot$term[dt_results_GERR_plot$truncate] <- sapply(dt_results_GERR_plot$term[dt_results_GERR_plot$truncate], function(long_term) {
      strsplit_tmp = strsplit(x = long_term, split=" ")[[1]]
      term_out =  paste0(c(strsplit_tmp[1:min(6, length(strsplit_tmp))], "..."), collapse=" ")
     return(term_out)})
    
  dt_results_GERR_plot$term = sapply(dt_results_GERR_plot$term, function(term) {
    paste0(toupper(substr(term,1,1)), substr(term,2,nchar(term)))
    })
    
  dt_results_GERR_plot$term = factor(dt_results_GERR_plot$term, levels=rev(unique(dt_results_GERR_plot$term)), ordered = T)
  
  vec_colors =  wes_palette(name="Cavalcanti1", n=3)
  names(vec_colors) = c("GO","KEGG","REACTOME")

  # Plot
  p_GO <- ggplot(dt_results_GERR_plot, aes(x=term, y=minlog10_p.value_fisher_BH,  fill=source)) +
    geom_bar(stat="identity", width=0.8, size=0.5, position=position_dodge()) + # fill="#73BCC9", ) +
    scale_fill_manual(values=vec_colors) +
    #theme_pubr(legend="none") +
    theme(

      
      axis.title.x = element_text(size = fontSize_axisTitle_md),
      axis.text.x = element_text(size=fontSize_axisText_lg),
       
      axis.title.y = element_blank(),
      axis.text.y = element_text(size=fontSize_axisText_xlg, face="bold"),
      axis.ticks.y = element_blank(),
      
      legend.title = element_blank(),
      legend.text = element_text(size = fontSize_legend_xlg),
         
       panel.grid.major = element_blank(), 
       panel.grid.minor = element_blank(),
       panel.background = element_blank(), 

        plot.background = element_blank(), 
        panel.grid = element_blank()) +
    #scale_size(range = c(5,10)) +
    #ggsci::scale_fill_lancet() +
    coord_flip() +
    xlab(NULL) + ylab(expression(paste(bold(-log[10]),bold("("),
                                       bolditalic("P"),bold(")"))))
  p_GO
  
  ggsave(plot = p_GO, filename = paste0(dirWGCNA_plots,prefixData,"_", prefixRun,  "_fig2B_left_GO_barplot_", mod_plot, "_", params$date,".pdf"), width=8)
  
  
  
    }
 
  
  ### NETWORK PLOT #### 
  
  vec_genes_NAFLD_hep1 = c("APOM","AFM", "ALDOB", "VTN")
  vec_genes_NAFLD_mac1 = c("HLA−DRA","HLA−DQA2","HLA−DQB2")
  
  # Compute network kme values
  vec_mod_genes <- dt_geneMod_perslab[module_renamed == mod_plot,"genes", with=F][[1]]
  vec_mod_kIMs <- dt_geneMod_perslab[module_renamed == mod_plot, "pkMs", with=F][[1]]

  # order
  vec_mod_genes = vec_mod_genes[order(vec_mod_kIMs, decreasing = T)]
  vec_mod_kIMs = sort(vec_mod_kIMs, decreasing = T)
  
  # take hubgenes
  vec_logical = vec_mod_genes %in% c(vec_mod_genes[1:min(size, length(vec_mod_genes))],  vec_genes_NAFLD_hep1,vec_genes_NAFLD_mac1) 
  vec_mod_genes =vec_mod_genes[vec_logical]
  vec_mod_kIMs = vec_mod_kIMs[vec_logical]
  
  mat_datExpr = subset(x=seuratObj, idents=gsub("_\\d$", "", mod_plot)) %>% GetAssayData(., assay="SCT_combat") %>% as.matrix
  
  mat_datExpr <- mat_datExpr[vec_mod_genes,]
  
  # Compute network adjacency
  mat_adj <- WGCNA::adjacency(t(mat_datExpr), 
                         power = 1, 
                         corFnc = "cor", 
                         corOptions = list(use = "p"),
                         type = "signed hybrid")

  
  vec_hub_genes <- vec_mod_genes[order(vec_mod_kIMs, decreasing=T)]
  vec_hub_kIMs <- vec_mod_kIMs[order(vec_mod_kIMs, decreasing=T)]
  df_hub.data <- cbind.data.frame(module = rep(mod_plot, length(vec_hub_genes)),
                             gene = vec_hub_genes, kIM = vec_hub_kIMs)

  # hub.data$logfc <- DEGs$log2FoldChange[match(hub.data$gene, rownames(DEGs))]
  # hub.data$p.adj <- DEGs$padj[match(hub.data$gene, rownames(DEGs))]
  mat_adj %>% 
  graph.adjacency(mode = "undirected", weighted = T, diag = FALSE) %>% 
  tidygraph::as_tbl_graph() %>% upgrade_graph() %>% activate(nodes) %>%
  dplyr::mutate(module = df_hub.data$module) %>%
  dplyr::mutate(kIM = df_hub.data$kIM) %>% 
  #dplyr::mutate(abs.logfc = abs(hub.data$logfc)) %>%
  #dplyr::mutate(p.adj = hub.data$p.adj) %>% 
  #dplyr::mutate(color=factor(ifelse(p.adj < 0.05, yes = "S", no = "NS"), 
  #                           levels = c("NS", "S"))) %>%
    activate(edges) %>% 
    #dplyr::filter(weight>.15) %>% 
    activate(nodes) %>%
  filter(!node_is_isolated()) -> hub.plot
  
  
    # add NAFLD GWAS genes which are in module to genes to be plotted (labelled)
    #vec_mod_genes_label = unique(c(vec_mod_genes[1:defaultsize], vec_genes_NAFLD[vec_genes_NAFLD %in% vec_mod_genes]))

  #vec_mod_genes_label = unique(c(vec_mod_genes[1:30], vec_genes_NAFLD_hep1, vec_genes_NAFLD_mac1))
    #}
  #      mat_adj[rownames(mat_adj) %in% vec_hep_hubgenes_and_circ_markers,rownames(mat_adj) %in% vec_hep_hubgenes_and_circ_markers] %>% 
  # graph.adjacency(mode = "undirected", weighted = T, diag = FALSE) %>% 
  # tidygraph::as_tbl_graph() %>% upgrade_graph() %>% activate(nodes) %>%
  # dplyr::mutate(module = df_hub.data[df_hub.data$gene %in% vec_hep_hubgenes_and_circ_markers,]$module) %>%
  # dplyr::mutate(kIM = df_hub.data[df_hub.data$gene %in% vec_hep_hubgenes_and_circ_markers,]$kIM) %>% 
  # #dplyr::mutate(abs.logfc = abs(hub.data$logfc)) %>%
  # #dplyr::mutate(p.adj = hub.data$p.adj) %>% 
  # #dplyr::mutate(color=factor(ifelse(p.adj < 0.05, yes = "S", no = "NS"), 
  # #                           levels = c("NS", "S"))) %>%
  #   activate(edges) %>% 
  #   #dplyr::filter(weight>.15) %>% 
  #   activate(nodes) %>%
  # filter(!node_is_isolated()) -> hub.plot_alt
       
  #}
  
  
  # Plot
  # layouts: https://rdrr.io/cran/ggraph/man/layout_tbl_graph_igraph.html
  module.plot <- ggraph(hub.plot, 
                        layout = "igraph",
                        algorithm = "drl",
                        #layout = "stress"
                        ) + 
    geom_edge_link(color="grey70", show.legend=F, aes(alpha=weight)) + 
    geom_node_point(aes(size = kIM), fill = "#73BCC9", shape=21, alpha=0.8) + 
    # scale_size(breaks = c(1,2,3), 
    #            limits = c(0,4), 
    #            range = c(0, 8), 
    #            labels = c(" \U00B1 1", " \U00B1 2", " \U00B1 3")
    #            ) +
    geom_node_text(
      aes(label = as.character(name)),
      fontface="bold.italic",
      size=fontSize_label_lg,
      repel = T,
      # incredibly, there is a mismatch in how the hyphon "-" is encoded in the genes..
      color = ifelse(gsub("\\W","_",rownames(mat_adj)) %in% gsub("\\W","_",c(vec_genes_NAFLD_hep1,vec_genes_NAFLD_mac1)), "red", "black") ) +
      
    guides(
      size = guide_legend(override.aes = list(
      #size=c(2,4,6)), 
      ),
      keywidth = 0.8, 
      keyheight = 0.8, order = 1)) + 
                               #title = expression(bold(paste("Log"[2],~
                                                              #"fold-change"))))) + 
    theme_graph(base_family = 'Helvetica') + 
    theme(
      legend.title.align=0.5,
      legend.position = "top",
      legend.margin = margin(0,0,0,0, unit="cm"),
      legend.title = element_text(size=fontSize_legend_xlg, face="bold"),
      legend.text = element_text(size=fontSize_legend_lg, face="bold"),
      legend.spacing.x = unit(0, "cm"),
      
      axis.text =element_blank(),
      axis.ticks = element_blank(), 
      axis.line=element_blank(),
      axis.title = element_blank(),
      # margin: top, right, bottom, and left
      plot.margin = unit(c(0, 0, 0, 0), "cm"),
          
      panel.grid.major = element_blank(), 
      panel.grid.minor = element_blank(),
      panel.background = element_blank(), 
 
      plot.background = element_blank(), 
      panel.grid = element_blank()
      ) +
    
    coord_cartesian(clip="off")
  
  module.plot

  ggsave(module.plot,filename = paste0(dirWGCNA_plots,prefixData,"_", prefixRun, "_fig2B_right_network_", mod_plot, "_", params$date,".pdf"), width=13, height=9)#, width=8, height=round(nrow(dt_results_GERR_metaMod_plot_sub)/9.5,0)) 
}

```



## Fig. 2 (C)  barplot module activity and preservation in MacParland


```{r}
path_seuratObj_macparland <- here("data", "macparland_seurat_obj3.RDS")

seuratObj_macparland <- load_obj(path_seuratObj_macparland)
```

```{r}
# includes Dendritic cells
path_dt_modPres_perslab_vs_macparland <- paste0(dirWGCNA_tables, "liver_perslab_int_seurat_7_wgcna3_cluster_perslab_coarse_vs_macparland_201215", "_df_preservationStats.csv")
dt_modPres_perslab_vs_macparland <- fread(path_dt_modPres_perslab_vs_macparland)
```

exclude dendritic for now

```{r}
dt_modPres_perslab_vs_macparland <- dt_modPres_perslab_vs_macparland[ref_lvl != "Dendritic-cells"]
```

```{r}
dt_modPres_perslab_vs_macparland_sub <- dt_modPres_perslab_vs_macparland[module %in% dt_geneMod_perslab$module_filter[nchar(dt_geneMod_perslab$module_renamed)>0]]
```

```{r}
dt_mod_dict = dt_geneMod_perslab[!duplicated(module_filter)]
```

```{r}
dt_modPres_perslab_vs_macparland_sub$module <- dt_mod_dict$module_renamed[match(dt_modPres_perslab_vs_macparland_sub$module, dt_mod_dict$module_filter)]
```

```{r}
# some are NA because many genes weren't in the moylan dataset

(1 - pnorm(dt_modPres_perslab_vs_macparland_sub$Z_stat)) %>% p.adjust(p=., method = pAdjMethod) -> vec_p.BH.pres

vec_p.BH.pres[vec_p.BH.pres==0] <- min(vec_p.BH.pres[vec_p.BH.pres>0], na.rm=T)  # avoid zero p-values which become inf when taking log

vec_min.log.p.BH.pres <- -log10(vec_p.BH.pres)

dt_modPres_perslab_vs_macparland_sub[,min.log.p.BH.pres:=vec_min.log.p.BH.pres]

```


```{r}

for (i in 1:4) {
  
    dt_modPres_perslab_vs_macparland_sub$module  =   gsub(pattern=c("alpha-", "beta", "gamma-", "delta")[i], 
  replacement=c(alpha_unicode, Beta_unicode, gamma_unicode, delta_unicode)[i],
  x=dt_modPres_perslab_vs_macparland_sub$module )
    
    
  dt_modPres_perslab_vs_macparland_sub$ref_lvl  =   gsub(pattern=c("alpha-", "beta", "gamma-", "delta")[i], 
  replacement=c(alpha_unicode, Beta_unicode, gamma_unicode, delta_unicode)[i],
  x=dt_modPres_perslab_vs_macparland_sub$ref_lvl )

    
        dt_modPres_perslab_vs_macparland_sub$test_lvl  =   gsub(pattern=c("alpha-", "beta", "gamma-", "delta")[i], 
  replacement=c(alpha_unicode, Beta_unicode, gamma_unicode, delta_unicode)[i],
  x=dt_modPres_perslab_vs_macparland_sub$test_lvl)
        
}

dt_modPres_perslab_vs_macparland_sub[,module_number:=gsub("[^0-9]","",module)]
dt_modPres_perslab_vs_macparland_sub[,module_abbrev:=paste0("M-",vec_celltype_abbrev[match(ref_lvl,names(vec_celltype_abbrev))], "-", module_number)]

```

```{r}

p_var_pres_macp = ggplot(data=dt_modPres_perslab_vs_macparland_sub, mapping=aes(x=module_abbrev, y=min.log.p.BH.pres, fill=ref_lvl)) +
    geom_bar(stat="identity", position=position_identity()) +#
             # + #, 
             #group=dt_condition_coef_sub$celltype) + 
  
    #geom_errorbar(aes(ymin=estimate-std.error, ymax=estimate+std.error), width=.1) +
    # geom_errorbar(aes(ymin=confInt_lower_percentile, ymax=confInt_upper_percentile), width=.1) +
  
    #scale_fill_brewer(palette="Paired")+
    scale_fill_manual(values=vec_colors_cluster_perslab_coarse[match(unique(dt_modPres_perslab_vs_macparland_sub$ref_lvl), names(vec_colors_cluster_perslab_coarse))]) + # this is a kluge
    #scale_fill_manual(values=unique(df_data$celltype)) + # this is a kluge
    #geom_hline(yintercept=cutoff, linetype="dashed", color = "red") +
    scale_y_continuous(expression(-log[10](P[preservation])), expand = c(0, 0)) +
    scale_x_discrete("module") +  
  
    geom_hline(yintercept=-log10(0.05),linetype="dotted") +
      
    # geom_text(aes(label=module),
    #         #angle=90,
    #         #vjust=-1,
    #         hjust = 2,
    #         color="black",
    #         angle=90,
    #        position = position_dodge(), 
    #        size=2.5) +
    #labs(y = "linear association") +
  
    theme(
     axis.text.x = element_text(angle=90,hjust=1,vjust=0.5,size = fontSize_axisText_md),#element_blank(),
     axis.title.x = element_blank(),
      axis.line.x = element_blank(),
     axis.ticks.x = element_blank(),
     
     axis.title.y = element_text(size = fontSize_axisTitle_sm),
     axis.text.y = element_text(size = fontSize_axisText_md),
     legend.title = element_blank(),
     legend.text = element_text(size = fontSize_legend_md),
     
     panel.grid.major = element_blank(), 
      panel.grid.minor = element_blank(),
         panel.background = element_blank(), 
    #legend.position = "none",
    #panel.grid.minor = element_blank(), 
    #panel.grid.major = element_line(colour = "grey80"), 
    plot.background = element_blank(), 
    panel.grid = element_blank(),
     aspect.ratio = 0.28)
  
  p_var_pres_macp 
  
  #ggsave(plot = p_var_pres_macp, filename = paste0(dirWGCNA_plots,prefixData,"_", prefixRun,  "_fig2C_mod_pres_in_macp_barplot_", params$date,".pdf"))#, height=, width =12)
```


## Fig. 2(D) WGCNA module association dotplots

### prep data tables for plots

```{r}

vec_colors_cluster_perslab_coarse = readRDS(paste0(dirWGCNA_plots,prefixData,"_", prefixRun,  "_vec_colors_cluster_perslab_coarse.RDS"))

for (i in 1:4) {
  names(vec_colors_cluster_perslab_coarse) =   gsub(pattern=c("alpha-", "beta", "gamma-", "delta")[i], 
  replacement=c(alpha_unicode, Beta_unicode, gamma_unicode, delta_unicode)[i],
  x=names(vec_colors_cluster_perslab_coarse))

}
 

```


```{r}

# data.table::dcast(data = dt_condition_coef[condition %in% c("fibrosis","lob_inflam")], formula= "module + estimate_original + confInt_lower_percentile + confInt_upper_percentile + p.value_percentile_BH~ condition", value.var = c("estimate_original", "confInt_lower_percentile", "confInt_upper_percentile", "p.value_percentile_BH"))

dt_fibrosis = dt_condition_coef[condition == "fibrosis"]
colnames(dt_fibrosis)[c(2,4:ncol(dt_fibrosis))] = paste0("fibrosis_",colnames(dt_fibrosis)[c(2,4:ncol(dt_fibrosis))])

dt_lob_inflam = dt_condition_coef[condition == "lob_inflam"]
colnames(dt_lob_inflam)[c(2,4:ncol(dt_lob_inflam))] = paste0("lob_inflam_",colnames(dt_lob_inflam)[c(2,4:ncol(dt_lob_inflam))])

dt_point = data.table::merge.data.table(x=dt_fibrosis, y= dt_lob_inflam, by="module", all=T)

# keep only modules significantly associated with a phenotype
dt_point = dt_point[lob_inflam_p.value_percentile_BH < params$pValThreshold/2 | fibrosis_p.value_percentile_BH < params$pValThreshold/2]

dt_point[,significance:= ifelse(test = lob_inflam_p.value_percentile_BH < params$pValThreshold/2, yes=ifelse(test=fibrosis_p.value_percentile_BH < params$pValThreshold/2, yes = "both",no="lobular inflammation"), no="fibrosis")]

dt_point = dt_point[!grepl("gerhard", module)]

dt_point$celltype = gsub("_\\d+", "", dt_point$module)
dt_point$celltype = gsub("_","-",dt_point$celltype)
dt_point$celltype = factor(dt_point$celltype, ordered=T)

# greek characters
for (i in 1:4) {
  dt_point$celltype =   gsub(pattern=c("alpha-", "beta", "gamma-", "delta")[i], 
  replacement=c(alpha_unicode, Beta_unicode, gamma_unicode, delta_unicode)[i],
  x=dt_point$celltype)

}
 



#vec_colors_cluster_perslab_coarse_re = vec_colors_cluster_perslab_coarse

#names(vec_colors_cluster_perslab_coarse_re) = gsub("-","_",names(vec_colors_cluster_perslab_coarse_re))

#vec_colors_cluster_perslab_coarse_re = vec_colors_cluster_perslab_coarse_re[names(vec_colors_cluster_perslab_coarse_re) %in% as.character(dt_point$celltype)]

dt_point[,module_number:=gsub("[^0-9]","",module)]
dt_point[,module_abbrev:=paste0("M-", vec_celltype_abbrev[match(celltype,names(vec_celltype_abbrev))], "-",module_number)]

# dt_point[,module_abbrev:=ifelse(grepl("bquote",vec_celltype_abbrev[match(celltype,names(vec_celltype_abbrev))]),
#          paste0(substr(vec_celltype_abbrev[match(celltype,names(vec_celltype_abbrev))],
#                                        1,
#                                        nchar(vec_celltype_abbrev[match(celltype,names(vec_celltype_abbrev))])-3), module_number, substr(vec_celltype_abbrev[match(celltype,names(vec_celltype_abbrev))],nchar(vec_celltype_abbrev[match(celltype,names(vec_celltype_abbrev))])-3, nchar(vec_celltype_abbrev[match(celltype,names(vec_celltype_abbrev))]))),
#          paste(vec_celltype_abbrev[match(celltype,names(vec_celltype_abbrev))],module_number))
#          ]


# add metamodule column

 
dt_dict = dt_geneMod_perslab[!duplicated(module_renamed), .(module_renamed, module_meta)]
dt_point[,metamodule := dt_dict$module_meta[match(dt_point$module, gsub("-","_",dt_dict$module_renamed))]]
```

```{r}
max_lim_x = round(max(abs(c(dt_point$lob_inflam_estimate_original))), 0)
max_lim_y = round(max(abs(c(dt_point$fibrosis_estimate_original))), 0)
```

#### color by cell type

```{r}
set.seed(randomSeed)

p_dot_celltype = ggplot(data=dt_point, mapping=aes(lob_inflam_estimate_original, fibrosis_estimate_original, color=celltype, shape=significance)) + 
geom_point(size=6, alpha=0.9) + 
  
  scale_x_continuous(limits = c(-(max_lim_x+0.4), max_lim_x), breaks=seq(-1.5,1,0.5)) + 
  scale_y_continuous(limits=c(-(max_lim_y+0.2), max_lim_y), breaks=seq(-2.5,2,0.5)) +
  
  coord_fixed() +
  
  ggrepel::geom_text_repel(
  #geom_text(
      #nudge_x = -1.8-dt_point$lob_inflam#[lob_inflam_estimate_original>1 | fibrosis_estimate_original>1]$lob_inflam_estimate_original,  
      #nudge_y      = 0,
      #parse=T,
      direction    = "both",
      segment.color = "grey50",
      #angle        = 0,
      #hjust        = 0,
      point.padding=NA,
      segment.size = 0.2,
      data = dt_point,#[lob_inflam_estimate_original>1 | fibrosis_estimate_original>1],
      mapping = aes(
        #x=lob_inflam_estimate_original,
        #y= fibrosis_estimate_original,
        label=stringr::str_wrap(module_abbrev)),
        #color=dt_point[abs(lob_inflam_estimate_original)>0.8 | abs(fibrosis_estimate_original)>0.8]$color), 
      color="black",
      size=fontSize_label_sm, 
      #fontface="bold"
      ) +
  
   # 
   # ggrepel::geom_text_repel(
   #    nudge_x = 0.9-dt_point[lob_inflam_estimate_original<(-1) | fibrosis_estimate_original<(-1)]$lob_inflam_estimate_original,  
   #    nudge_y      = 0,
   #    direction    = "y",
   #    segment.color = "grey50",
   #    #angle        = 0,
   #    hjust        = 0,
   #    segment.size = 0.2,
   #    data = dt_point[lob_inflam_estimate_original<(-1) | fibrosis_estimate_original<(-1)],
   #    mapping = aes(
   #      x=lob_inflam_estimate_original,
   #      y= fibrosis_estimate_original,
   #      label=stringr::str_wrap(module)),
   #      #color=dt_point[abs(lob_inflam_estimate_original)>0.8 | abs(fibrosis_estimate_original)>0.8]$color), 
   #    color="black",
   #    size=2.5, 
   #    #fontface="bold"
   #    ) +
  
  #labs(x = "z-score", y="-log(p-value)", colour="z-score") +
  xlab(expression(paste(bold("\u03B2")," lobular inflammation")))+
  ylab(expression(paste(bold("\u03B2")," fibrosis")))+
  
  #labs(x = "lobular inflammation", y="fibrosis") +
  
  scale_color_manual(
    values=vec_colors_cluster_perslab_coarse) +#,
  
  guides(
    colour = guide_legend(order=1),
    shape = guide_legend(reverse=TRUE, order=2)) + 
  
  # set legend order and reverse the order of shape legend
  scale_shape_manual(values=c(15, 16, 17))+
    #labels=levels(dt_point$celltype)) + 
  geom_hline(yintercept = 0, size=0.1) +
  geom_vline(xintercept = 0, size=0.1) + 
  theme(axis.text = element_text(size = fontSize_axisText_sm), 
        axis.line = element_line(colour = "grey80"), 
        axis.ticks = element_line(colour = "grey80"), 
        axis.title = element_text(size = fontSize_axisTitle_sm),#, face = "bold"), 
        panel.background = element_blank(), 
        #legend.position = "none",
        #panel.grid.minor = element_blank(), 
        #panel.grid.major = element_line(colour = "grey80"), 
        plot.background = element_blank(), 
        panel.grid = element_blank(),
        text = element_text(size = fontSize_axisText_md),
        legend.title=element_text(size=fontSize_legend_lg),
        legend.text = element_text(size=fontSize_legend_md),
        #aspect.ratio = 1
        ) 
  p_dot_celltype
  
  
  #TODO
  # ggsave isn't rendering unicode - save from Rstudio canvas using export -> save as image -> set dimensions 800 * 800
# ggsave(plot = p, filename = paste0(dirWGCNA_plots,prefixData,"_", prefixRun,  "_fig2D_sc_mod_association_dotplot_celltype_", params$date,".pdf"), height=10, width =12)
```


# Fig. 3.

## Fig. 3. (A)? SWNE embeddings

```{r}
dt_gprofiler_KEGG_cast = dcast.data.table(dt_gprofiler,
                                     formula = term_id ~ query, 
                                     value.var = "p_value", 
                                     fill=NA_real_,
                                     drop=FALSE,
                                     subset=.(significant==TRUE,source="KEGG"))

mat_gprofiler_KEGG = as.matrix(dt_gprofiler_KEGG_cast[,.SD, .SDcols = colnames(dt_gprofiler_KEGG_cast)[2:ncol(dt_gprofiler_KEGG_cast)]])

rownames(mat_gprofiler_KEGG) = dt_gprofiler_KEGG_cast$term_id
# convert p-values to -log10
mat_gprofiler_KEGG[!is.na(mat_gprofiler_KEGG)] = -log10(mat_gprofiler_KEGG[!is.na(mat_gprofiler_KEGG)])
mat_gprofiler_KEGG[is.na(mat_gprofiler_KEGG)] = 0

```

```{r}
seu_KEGG = CreateSeuratObject(counts=mat_gprofiler_KEGG, assay = "RNA")
seu_KEGG <- seu_KEGG %>% FindVariableFeatures %>% ScaleData %>% RunPCA(object=., npcs = 15, seed.use=randomSeed, verbose=F)

vec_genesets_embed <- dt_gprofiler[,j=head(term_name, 3), by=query][["V1"]] #already sorted by p-value

```

```{r}
swne.embedding = RunSWNE(
  object=seu_KEGG,
  proj.method = "sammon",
  reduction.use = "pca",
  cells.use = NULL,
  dims.use = 1:15,
  genes.use = NULL,
  dist.metric = "cosine",
  distance.matrix = NULL,
  n.cores = getDTthreads(),
#k,
  k.range=1:15,
  #var.genes=seu_KEGG@assays$geneset_enrichment_KEGG@var.features,
  loss = "mse",
  genes.embed=vec_genesets_embed,
  hide.factors = T,
  n_pull = 3,
  ica.fast = T,
  alpha.exp = 1.25,
  snn.exp = 1,
  snn.k = 10,
  use.paga.pruning = T,
  paga.qval.cutoff = 0.001,
  reduction.name = "swne",
  reduction.key = "SWNE_",
  return.format = "embedding",
  npcs = 15,
  nu = 15,
  nv = 15
)
```

Something completely different: embed modules and genesets in low-dimensional gene space

get a vector of all the genes in sc experiment

```{r}
vec_genes_perslab <- readRDS(here("data","liver_perslab_int_genesBackground.RDS"))
```

```{r}
vec_genes_gerhard <- readRDS(here("data","liver_gerhard2018_genesBackground.RDS"))
```

# GERR results
```{r}
dt_KEGGenrichmentResults = fread(paste0(dirWGCNA_tables, prefixData, "_", prefixRun,  "_201216_GERR_KEGGenrichmentResults.csv"))
```


#### Download KEGG genesets

```{r}
library("gage")
data(kegg.gs)
length(kegg.gs)
# [1] 177
# Hm it says 186 on the msigdb page http://software.broadinstitute.org/gsea/msigdb/genesets.jsp?collection=CP:KEGG

list_vec_genesetsKEGG = kegg.gs
rm(kegg.gs)
names(list_vec_genesetsKEGG) = substr(names(list_vec_genesetsKEGG), start=1, stop=8)

```

```{r}
data(egSymb)

list_vec_genesetsKEGG <- lapply(list_vec_genesetsKEGG, function(vec_geneset) {
  vec_geneset <- eg2sym(eg=vec_geneset)
  vec_geneset[!is.na(vec_geneset)]
})
```

keep most enriched genesets for each module 


```{r}
vec_mod_geneset_plot = c(
  "T-cells-gamma-delta_1",
  "Macrophages_4",
  "Endothelial-cells_2",
  "T-cells-alpha-beta_1",
  "Cholangiocytes_2",
  "Dendritic-cells_2",
  "Plasma-cells_1",
  "Hepatic-stellate-cells_3",
  "B-cells-mature_1",
  "Hepatic-stem-cells_1"
)
```


```{r}
dt_KEGGenrichmentResults <- dt_KEGGenrichmentResults[order(module, p.value_fisher_BH)]


vec_genesets_embed <- dt_KEGGenrichmentResults[p.value_fisher_BH<0.05 & module %in% vec_mod_geneset_plot,
                                               j=head(KEGG_id, 20), 
                                               by=module][["V1"]] #already sorted by p-value

list_vec_genesetsKEGG  = list_vec_genesetsKEGG[names(list_vec_genesetsKEGG) %in% vec_genesets_embed]

```

```{r}

# prepare a weighted binary gene * module matrix 
dt_mod = dcast.data.table(data = dt_geneMod_perslab, 
                           formula = genes ~ module_renamed,
                           subset = .((nchar(module_lob_inflam_gerhard2018)>0 |
                                         nchar(module_fibrosis_gerhard2018)>0) &
                                        module_renamed %in% vec_mod_geneset_plot), 
                          value.var = "pkMs", 
                          #fun.aggregate = length, 
                          fill = 0)


```

```{r}
# normalize module activity to sum to 1 in each column
dt_mod_norm = data.table("genes"= dt_mod$genes, dt_mod[,lapply(.SD, function(eachcol){eachcol/sum(eachcol)}),.SDcols = colnames(dt_mod)[2:ncol(dt_mod)]])

```

put unicode in T cell module names

```{r}
for (i in 1:4) {
  colnames(dt_mod_norm) <- gsub(pattern=c("alpha-", "beta", "gamma-", "delta")[i], 
  replacement=c(alpha_unicode, Beta_unicode, gamma_unicode, delta_unicode)[i],
  x=colnames(dt_mod_norm))
      
}
```

prepare gene * KEGG geneset sparse matrix 

for a given geneset, gene weights are equal and sum to 1

```{r}
vec_KEGG_allgenes = unlist(list_vec_genesetsKEGG, use.names=F)

dt_KEGG = sapply(list_vec_genesetsKEGG, function(vec_geneset) as.numeric(vec_KEGG_allgenes %in% vec_geneset)/length(vec_geneset)) %>% data.table()

dt_KEGG = data.table("genes"=vec_KEGG_allgenes, dt_KEGG)
```

```{r}
dt_module_KEGG = data.table::merge.data.table(x=dt_mod_norm, y= dt_KEGG, by="genes", all=T)

mat_module_KEGG = as.matrix(dt_module_KEGG[,.SD,.SDcols=colnames(dt_module_KEGG)[2:ncol(dt_module_KEGG)]])
rownames(mat_module_KEGG) = dt_module_KEGG$genes
mat_module_KEGG[is.na(mat_module_KEGG)] <- 0
```


```{r}
set.seed(randomSeed)
mat_module_KEGG_svd = svd(x = mat_module_KEGG, nv = 30)
mat_module_KEGG_umap = umap(d = mat_module_KEGG_svd$v, preserve.seed  = T)
```

```{r}
dict_KEGG_term_id = dt_KEGGenrichmentResults[!duplicated(KEGG_id)]
```

prep data for plot

```{r}
dt_plot_umap = data.table("variable"= colnames(mat_module_KEGG),
                          "var_type" = ifelse(grepl("hsa",colnames(mat_module_KEGG)), "KEGG_term", "module"),
                          mat_module_KEGG_umap$layout)

colnames(dt_plot_umap)[3:4] <- c("UMAP_1", "UMAP_2")
```

abbreviate module names

```{r}
dt_plot_umap$celltype <-  gsub("_\\d$","",dt_plot_umap$variable)
```

```{r}
dt_plot_umap$variable <- ifelse(grepl("hsa", dt_plot_umap$variable), 
                                yes=dt_plot_umap$variable,
                                no=paste0("M-", vec_celltype_abbrev[match(dt_plot_umap$celltype, names(vec_celltype_abbrev))], "-", gsub("\\D*","",dt_plot_umap$variable)))
```

```{r}
dt_plot_umap$variable <- ifelse(grepl("hsa",colnames(mat_module_KEGG)), dict_KEGG_term_id$KEGG_term[match(colnames(mat_module_KEGG), dict_KEGG_term_id$KEGG_id)], dt_plot_umap$variable)
```

plot 

```{r}
set.seed(randomSeed)

p_point_umap_module_KEGG = ggplot(data=dt_plot_umap, 
                                  mapping=aes(x=UMAP_1, 
                                              y=UMAP_2,  
                                              shape=var_type)) + 
geom_point(#size=6, 
           alpha=0.9, 
           mapping=aes(size=var_type),
           colour = wes_palette("Royal1",4)[4]) + 
  
  # scale_x_continuous(limits = c(-(max_lim_x+0.4), max_lim_x), breaks=seq(-1.5,1,0.5)) + 
  # scale_y_continuous(limits=c(-(max_lim_y+0.2), max_lim_y), breaks=seq(-2.5,2,0.5)) +
  
  coord_fixed() +
  
  ggrepel::geom_text_repel(
  #geom_text(
      #nudge_x = -1.8-dt_point$lob_inflam#[lob_inflam_estimate_original>1 | fibrosis_estimate_original>1]$lob_inflam_estimate_original,  
      #nudge_y      = 0,
      #parse=T,
      direction    = "both",
      segment.color = "grey50",
      #angle        = 0,
      #hjust        = 0,
      #point.padding=NA,
      segment.size = 0.2,
      data = dt_plot_umap,#[lob_inflam_estimate_original>1 | fibrosis_estimate_original>1],
      mapping = aes(
        x=UMAP_1,
        y= UMAP_2,
        label=stringr::str_wrap(variable)),
        #color=dt_point[abs(lob_inflam_estimate_original)>0.8 | abs(fibrosis_estimate_original)>0.8]$color), 
      color="black",
      size=fontSize_label_sm, 
      #fontface="bold"
      ) +
  
  #labs(x = "lobular inflammation", y="fibrosis") +
  
  guides(
    colour = guide_legend(order=1),
    shape = guide_legend(reverse=TRUE, order=2)) + 
  
  # set legend order and reverse the order of shape legend
  scale_shape_manual(values=c(15, 16))+
    #labels=levels(dt_point$celltype)) + 
  #geom_hline(yintercept = 0, size=0.1) +
  #geom_vline(xintercept = 0, size=0.1) + 
  theme(axis.text = element_text(size = fontSize_axisText_sm), 
        axis.line = element_line(colour = "grey80"), 
        axis.ticks = element_line(colour = "grey80"), 
        axis.title = element_text(size = fontSize_axisTitle_md),#, face = "bold"), 
        panel.background = element_blank(), 
        #legend.position = "none",
        #panel.grid.minor = element_blank(), 
        #panel.grid.major = element_line(colour = "grey80"), 
        plot.background = element_blank(), 
        panel.grid = element_blank(),
        text = element_text(size = fontSize_axisText_md),
        legend.title=element_text(size=fontSize_legend_lg),
        legend.text = element_text(size=fontSize_legend_md),
        #aspect.ratio = 1
        ) 
  p_point_umap_module_KEGG 
  
  #TODO
  
# ggsave(plot = p_point_umap_module_KEGG , filename = paste0(dirWGCNA_plots,prefixData,"_", prefixRun,  "_fig3A_mod_geneset_umap_dotplot_", params$date,".pdf"), height=10, width =12)
```

no - layout doesn't reflect actual module enrichment for geneset. FFS

try a barplot for lob_inflam association, 
with a geneset * module matrix of geneset enrichment significance below


```{r}
condition = quote(dt_condition_coef$module %in% gsub("-","_",vec_mod_geneset_plot))

dt_condition_coef_sub = dt_condition_coef[eval(condition)]

dt_condition_coef_sub <- dt_condition_coef_sub[condition == "lob_inflam"]


```

```{r}
# greek characters
for (i in 1:4) {
    dt_condition_coef_sub$module = gsub(pattern=c("alpha_", "beta", "gamma_", "delta")[i], 
  replacement=c(alpha_unicode, Beta_unicode, gamma_unicode, delta_unicode)[i],
  x=dt_condition_coef_sub$module)
      
}
 
```

```{r}
dt_condition_coef_sub$celltype = gsub("_\\d","", dt_condition_coef_sub$module)

dt_condition_coef_sub$module_abbrev = paste0("M-",vec_celltype_abbrev[match(dt_condition_coef_sub$celltype, gsub("-","_",names(vec_celltype_abbrev)))], "-", gsub("\\D*","",dt_condition_coef_sub$module))
```

```{r}
vec_colors_cluster_perslab_coarse_sub = vec_colors_cluster_perslab_coarse
names(vec_colors_cluster_perslab_coarse_sub) = gsub("-","_", names(vec_colors_cluster_perslab_coarse))
```

```{r}
p_bar_lobinflam <- ggplot(data=dt_condition_coef_sub, aes(x=module_abbrev, 
                                               y=estimate_original,
                                               fill=celltype)) +
    geom_bar(stat="identity",  width=1, size=0.5) + # fill="#73BCC9", ) +
  
    geom_errorbar(aes(ymin=confInt_lower_percentile, ymax=confInt_upper_percentile), width=.1) +
    #geom_hline(yintercept=-log10(0.05),linetype="dotted") +
    
    ylab(expression(paste(bold("\u03B2")," lobular inflammation")))+
      #ylab("lobular inflammation")+
  
      scale_fill_manual(values=vec_colors_cluster_perslab_coarse_sub[match(dt_condition_coef_sub$celltype, names(vec_colors_cluster_perslab_coarse_sub))]) +
    # #theme_pubr(legend="none") +
      theme(
        axis.title.x = element_blank(),#element_text(size = fontSize_axisTitle_md),
        axis.text.x = element_blank(),#element_text(size=fontSize_axisText_lg),
        axis.title.y = element_text(size = fontSize_axisTitle_sm),
        axis.ticks.x = element_blank(),
      
        
        
        legend.position = "none",
        legend.title = element_blank(),
        legend.text = element_blank(),#element_text(size = fontSize_legend_xlg),
  
         panel.grid.major = element_blank(),
         panel.grid.minor = element_blank(),
         panel.background = element_blank(),
  
          plot.background = element_blank(),
          panel.grid = element_blank(),
          aspect.ratio = 0.5
      ) #+
    #scale_size(range = c(5,10)) +
    #ggsci::scale_fill_lancet() +
    #coord_flip() +
    #xlab(NULL) + ylab(expression(paste(bold(-log[10]),bold("("),
     #                                  bolditalic("P"),bold(")"))))

p_bar_lobinflam
```

hypergeometric GERR enrichment panel

```{r}
vec_KEGG_terms_plot = c("RNA transport", 
                        "Oxidative phosphorylation",
                        "Ribosome")
```

```{r}
dt_KEGGenrichmentResults_sub <- dt_KEGGenrichmentResults[module %in% vec_mod_geneset_plot & KEGG_term %in% vec_KEGG_terms_plot]
```

add any missing values

```{r}
vec_logical_mod_complete = dt_KEGGenrichmentResults_sub[,.N, by=module][,N]==length(vec_KEGG_terms_plot)
names(vec_logical_mod_complete) = dt_KEGGenrichmentResults_sub[,.N, by=module][,module]
if (!all(vec_logical_mod_complete)) {
  for (mod in names(vec_logical_mod_complete)[!vec_logical_mod_complete]) {
    for (KEGG_term in vec_KEGG_terms_plot) {
      # condition = quote(dt_KEGGenrichmentResults_sub$module == mod & dt_KEGGenrichmentResults_sub$KEGG_term == KEGG_term)
      if (!any(dt_KEGGenrichmentResults_sub$module == mod & dt_KEGGenrichmentResults_sub$KEGG_term == KEGG_term)) {
        dt_KEGGenrichmentResults_sub = rbindlist(list(dt_KEGGenrichmentResults_sub, data.table(
          "module"=mod, 
           "coef"=0,
           "p.value_fisher"=1,
          "p.value_fisher_BH"=1,
           "KEGG_id"= dt_KEGGenrichmentResults_sub[!duplicated(module), KEGG_id][match(KEGG_term, dt_KEGGenrichmentResults_sub[!duplicated(module), KEGG_term])],
           "KEGG_term"=KEGG_term)))
      }
    }
  }
} 
```

# recompute BH of fisher P-val using only KEGG

```{r}
pAdjust_n = length(list_vec_genesetsKEGG)*length(vec_mod_geneset_plot)
#pAdjust_n = 30
dt_KEGGenrichmentResults_sub[, p.value_fisher_BH:=p.adjust(p.value_fisher, method = pAdjMethod, n=pAdjust_n)]
```

<!-- ```{r} -->
<!-- mat_mod_geneset_hypergeom = sapply(vec_mod_geneset_plot, function(module){ -->
<!--   condition = quote(dt_geneMod_perslab$module_renamed == module) -->
<!--   vec_mod_genes = dt_geneMod_perslab$genes[eval(condition)] -->
<!--   sapply(list_vec_genesetsKEGG[], function(vec_geneset) { -->
<!--     phyper( -->
<!--       q = sum( vec_mod_genes %in% vec_geneset), -->
<!--       m = length(vec_geneset), -->
<!--       n = length(vec_genes_perslab)-length(vec_geneset), -->
<!--       k = length(vec_mod_genes), lower.tail = F) -->
<!--   }) -->
<!-- }) -->
<!-- ``` -->

<!-- ```{r} -->
<!-- mat_mod_geneset_hypergeom_padj = apply(mat_mod_geneset_hypergeom, MARGIN = 1, FUN = function(eachrow) { -->
<!--   p.adjust(p=eachrow,method = pAdjMethod, n = 30#nrow(mat_mod_geneset_hypergeom)*ncol(mat_mod_geneset_hypergeom) -->
<!--            ) -->
<!--   }) -->
<!-- ``` -->

<!-- replace KEGG ids with KEGG terms  -->

<!-- ```{r} -->
<!-- colnames(mat_mod_geneset_hypergeom_padj) = dt_KEGGenrichmentResults$KEGG_term[match(colnames(mat_mod_geneset_hypergeom_padj), dt_KEGGenrichmentResults$KEGG_id)] -->
<!-- ``` -->

<!-- ```{r} -->
<!-- mat_mod_geneset_hypergeom_padj = mat_mod_geneset_hypergeom_padj[,colnames(mat_mod_geneset_hypergeom_padj) %in% c("RNA transport", "Oxidative phosphorylation","Ribosome")] -->
<!-- ``` -->

```{r}
dt_KEGGenrichmentResults_sub[,asterisk := ifelse(p.value_fisher_BH<1e-4, "****", ifelse(p.value_fisher_BH<1e-3, "***", ifelse(p.value_fisher_BH<1e-2, "**", ifelse(p.value_fisher_BH<0.05, "*", "NS"))))]
```

<!-- ```{r} -->
<!-- mat_mod_geneset_hypergeom_padj_asterisk = apply(X=mat_mod_geneset_hypergeom_padj, MARGIN=2, FUN=function(eachcol){  -->
<!--   ifelse(eachcol<1e-4, "****", ifelse(eachcol<1e-3, "***", ifelse(eachcol<1e-2, "**", ifelse(eachcol<0.05, "*", "NS")))) -->
<!--   } )  -->
<!-- ``` -->

```{r}
for (i in 1:4) {
  dt_KEGGenrichmentResults_sub$module <- gsub(pattern=c("alpha-", "beta", "gamma-", "delta")[i], 
  replacement=c(alpha_unicode, Beta_unicode, gamma_unicode, delta_unicode)[i],
  x=dt_KEGGenrichmentResults_sub$module)
      
}
```

<!-- ```{r} -->
<!-- for (i in 1:4) { -->
<!--   rownames(mat_mod_geneset_hypergeom_padj_asterisk) <- gsub(pattern=c("alpha-", "beta", "gamma-", "delta")[i],  -->
<!--   replacement=c(alpha_unicode, Beta_unicode, gamma_unicode, delta_unicode)[i], -->
<!--   x=rownames(mat_mod_geneset_hypergeom_padj_asterisk)) -->

<!-- } -->
<!-- ``` -->


```{r}
# dt_plot_hyper = data.table("module"=rownames(mat_mod_geneset_hypergeom_padj_asterisk), mat_mod_geneset_hypergeom_padj_asterisk)
# 
dt_plot_hyper <- dt_KEGGenrichmentResults_sub

dt_plot_hyper$module_abbrev = paste0("M-",vec_celltype_abbrev[match(gsub("_\\d$", "",dt_plot_hyper$module), names(vec_celltype_abbrev))], "-", gsub("\\D*","",dt_plot_hyper$module))

dt_plot_hyper$module = NULL
#dt_plot_hyper_melt = data.table::melt.data.table(dt_plot_hyper, id.vars= "module_abbrev", variable.name = "KEGG_term")



#dt_plot_hyper_melt$KEGG_term = factor(dt_plot_hyper_melt$KEGG_term, levels=c("RNA transport", "Oxidative phosphorylation","Ribosome"), ordered=T)

dt_plot_hyper$KEGG_term = factor(dt_plot_hyper$KEGG_term, levels=vec_KEGG_terms_plot, ordered=T)

```

```{r}
p_geneset_enrich = ggplot(data = dt_plot_hyper, mapping = aes(x=module_abbrev, y= KEGG_term)) + 
  geom_text(
    data = dt_plot_hyper,
    mapping = aes(label=asterisk),
          size=fontSize_label_sm) + 
  theme(#axis.text = element_text(size = fontSize_axisText_sm), 
      aspect.ratio = 0.2,
        axis.text.x = element_text(angle=90,hjust=1,vjust=0.5,size = fontSize_axisText_lg),
        axis.text.y = element_text(size = fontSize_axisText_lg),#angle=90,hjust=1,vjust=0.5),
        axis.line = element_blank(),
        axis.ticks = element_blank(),
        axis.title = element_blank(),#element_text(size = fontSize_axisTitle_md),#, face = "bold"), 
        panel.background = element_blank(), 
        legend.position = "none",
        #panel.grid.minor = element_blank(), 
        #panel.grid.major = element_line(colour = "grey80"), 
        plot.background = element_blank(), 
        panel.grid = element_blank(),
        # text = element_text(size = fontSize_axisText_md),
        # legend.title=element_text(size=fontSize_legend_lg),
        # legend.text = element_text(size=fontSize_legend_md),
        #aspect.ratio = 1
        ) 
p_geneset_enrich
```

```{r}
library(patchwork)
```

```{r}
p_bar_lobinflam / p_geneset_enrich
```

# Fig. 4

## Fig. 4 (A): Bulk modules bar graph for inflammation and fibrosis (alt. a dot plot) for the significant bulk modules

```{r}

dt_fibrosis_gerhard = dt_condition_coef[condition == "fibrosis" & grepl("gerhard",module)]
colnames(dt_fibrosis_gerhard)[c(2,4:ncol(dt_fibrosis_gerhard))] = paste0("fibrosis_",colnames(dt_fibrosis_gerhard)[c(2,4:ncol(dt_fibrosis_gerhard))])

dt_lob_inflam_gerhard = dt_condition_coef[condition == "lob_inflam" & grepl("gerhard",module)]
colnames(dt_lob_inflam_gerhard)[c(2,4:ncol(dt_lob_inflam_gerhard))] = paste0("lob_inflam_",colnames(dt_lob_inflam_gerhard)[c(2,4:ncol(dt_lob_inflam_gerhard))])


dt_point_gerhard = data.table::merge.data.table(x=dt_fibrosis_gerhard, y= dt_lob_inflam_gerhard, by="module", all=T)

# keep only modules significantly associated with a phenotype
dt_point_gerhard = dt_point_gerhard[lob_inflam_p.value_percentile_BH < params$pValThreshold/2 | fibrosis_p.value_percentile_BH < params$pValThreshold/2]

dt_point_gerhard[,significance:= ifelse(test = lob_inflam_p.value_percentile_BH < params$pValThreshold/2, yes=ifelse(test=fibrosis_p.value_percentile_BH < params$pValThreshold/2, yes = "both",no="lobular inflammation"), no="fibrosis")]

dt_point_gerhard[,module_abbrev:=gsub("gerhard2018_","M-Ger-",module)]
```

```{r}
max_lim_x = round(max(abs(c(dt_point_gerhard$lob_inflam_estimate_original))), 0)
max_lim_y = round(max(abs(c(dt_point_gerhard$fibrosis_estimate_original))), 0)
```

```{r}
set.seed(randomSeed)

p_dot_gerhard = ggplot(data=dt_point_gerhard, mapping=aes(lob_inflam_estimate_original, fibrosis_estimate_original,  shape=significance)) + 
geom_point(size=6, alpha=0.9, colour = wes_palette("Royal1",4)[4]) + 
  
  scale_x_continuous(limits = c(-(max_lim_x+0.4), max_lim_x), breaks=seq(-1.5,1,0.5)) + 
  scale_y_continuous(limits=c(-(max_lim_y+0.2), max_lim_y), breaks=seq(-2.5,2,0.5)) +
  
  coord_fixed() +
  
  ggrepel::geom_text_repel(
  #geom_text(
      #nudge_x = -1.8-dt_point$lob_inflam#[lob_inflam_estimate_original>1 | fibrosis_estimate_original>1]$lob_inflam_estimate_original,  
      #nudge_y      = 0,
      #parse=T,
      direction    = "both",
      segment.color = "grey50",
      #angle        = 0,
      #hjust        = 0,
      #point.padding=NA,
      segment.size = 0.2,
      data = dt_point_gerhard,#[lob_inflam_estimate_original>1 | fibrosis_estimate_original>1],
      mapping = aes(
        #x=lob_inflam_estimate_original,
        #y= fibrosis_estimate_original,
        label=stringr::str_wrap(module_abbrev)),
        #color=dt_point[abs(lob_inflam_estimate_original)>0.8 | abs(fibrosis_estimate_original)>0.8]$color), 
      color="black",
      size=fontSize_label_sm, 
      #fontface="bold"
      ) +
  xlab(expression(paste(bold("\u03B2")," lobular inflammation")))+
  ylab(expression(paste(bold("\u03B2")," fibrosis")))+
  #labs(x = "lobular inflammation", y="fibrosis") +
  
  guides(
    colour = guide_legend(order=1),
    shape = guide_legend(reverse=TRUE, order=2)) + 
  
  # set legend order and reverse the order of shape legend
  scale_shape_manual(values=c(15, 16, 17))+
    #labels=levels(dt_point$celltype)) + 
  geom_hline(yintercept = 0, size=0.1) +
  geom_vline(xintercept = 0, size=0.1) + 
  theme(axis.text = element_text(size = fontSize_axisText_sm), 
        axis.line = element_line(colour = "grey80"), 
        axis.ticks = element_line(colour = "grey80"), 
        axis.title = element_text(size = fontSize_axisTitle_sm),#, face = "bold"), 
        panel.background = element_blank(), 
        #legend.position = "none",
        #panel.grid.minor = element_blank(), 
        #panel.grid.major = element_line(colour = "grey80"), 
        plot.background = element_blank(), 
        panel.grid = element_blank(),
        text = element_text(size = fontSize_axisText_md),
        legend.title=element_text(size=fontSize_legend_lg),
        legend.text = element_text(size=fontSize_legend_md),
        #aspect.ratio = 1
        ) 
# 
#   theme(axis.text = element_text(size = fontSize_axisText_sm), 
#         axis.line = element_line(colour = "grey80"), 
#         axis.ticks = element_line(colour = "grey80"), 
#         axis.title = element_text(size = fontSize_axisTitle_md),#, face = "bold"), 
#         panel.background = element_blank(), 
#         #legend.position = "none",
#         #panel.grid.minor = element_blank(), 
#         #panel.grid.major = element_line(colour = "grey80"), 
#         plot.background = element_blank(), 
#         panel.grid = element_blank(),
#         text = element_text(size = fontSize_axisText_md),
#         legend.title=element_text(size=fontSize_legend_lg),
#         legend.text = element_text(size=fontSize_legend_md),
#         #aspect.ratio = 1
#         ) 
  p_dot_gerhard
  
  # pdf fails: instead use export -> 800*800
  
  
  
# ggsave(plot = p_dot_gerhard, filename = paste0(dirWGCNA_plots,prefixData,"_", prefixRun,  "_fig6A_bulk_mod_association_dotplot_", params$date,".pdf"), height=10, width =12)
```

## Fig. 4 (B) M-Hep-4 violinplot

```{r}
p_hep4 <- VlnPlot(object=seuratObj,
                     assay="SCT_combat",
                     features = "Hepatocytes_4",
                     pt.size = 0,# 0.05,
                     cols=vec_colors_cluster_perslab_coarse,#vec_colors_markers,
                     sort = F, 
                     group.by = "cluster_perslab_coarse", 
                     #same.y.lims = F,
                     slot="data",
                     log = F,
                  )#,


p_hep4 <- p_hep4 + 
  
  ggtitle("M-Hep- activity") + 
  
  #guides()

  theme(legend.position="none", 
        axis.title.x = element_blank(),
      
      axis.title.y=element_blank(),
      axis.text.y=element_text(
      hjust=1,
      vjust=0,
      size=fontSize_axisText_lg,
      angle=30),

      axis.text.x = element_text(
        size= fontSize_axisText_xlg,
        angle=35),
      plot.title = element_text(size=fontSize_plotTitle*2),
      aspect.ratio=0.4,
      #legend.position="none"
  )


p_hep4

#ggsave(plot= p_hep4, filename = here("plots", paste0(prefixData, "_", prefixRun, "_fig4B_vlnplot_MHep4_",flagDate,".pdf")), width = 6, height=6)
```


## Fig. 4 (C) M-Hep-8 GO and network

left: M-Hep-8 gene set enrichment barplot (plotted alongside Fig. 2 (A))
right: M-Hep-8 network plot (plotted alongside Fig. 2 (A))

## Fig. 4 (D): PIGR featureplot

```{r}
DefaultAssay(seuratObj) = "SCT_combat"

p_PIGR = FeaturePlot(seuratObj, features="PIGR", reduction = "umap", slot = "data")

p_PIGR = p_PIGR +
  theme(
      plot.title = element_text(face = "italic", size=fontSize_plotTitle*2),
    
  )
ggsave(plot=p_PIGR, filename = here("plots", paste0(prefixData, "_", prefixRun, "_fig4D_featplot_PIGR_",flagDate,".pdf")), width = 6, height=6)
```


# Fig. 5



## Fig. 5 (B): macrophage cluster UMAP cutout 

```{r}
p <- UMAPPlot(seuratObj, 
             #reduction = "umap",
             #pt.size=.01,
             group.by="cluster_perslab", 
             seed=randomSeed,
             #label.box=TRUE,
             cols=vec_colors_cluster_perslab,
             label = T,
             label.size = fontSize_label_sm,
             repel=T
             )

p <- p + theme(
  legend.position="none",
  axis.title.x = element_text(size=fontSize_axisTitle_md),
  axis.title.y = element_text(size=fontSize_axisTitle_md))

p

#ggsave(plot= p, filename = here("plots", paste0(prefixData, "_", prefixRun, "_fig5b_cluster_perslab_mac",flagDate,".pdf")), width = 12, height=12)
```

## Fig. 5 (C): macrophage cluster Violinplots

prepare seurat subset object

```{r}
seuratObj_macs <- subset(seuratObj, subset=cluster_perslab_coarse == "Macrophages")
```

<!-- ```{r} -->
<!-- vec_colors_cluster_perslab = readRDS(paste0(dirWGCNA_plots,prefixData,"_", prefixRun,  "_vec_colors_cluster_perslab.RDS")) -->
<!-- ``` -->

```{r}
vec_colors_cluster_perslab_macs = vec_colors_cluster_perslab[names(vec_colors_cluster_perslab) %in% seuratObj_macs$cluster_perslab]
```

```{r}
#colnames(seuratObj_macs@meta.data)[colnames(seuratObj_macs@meta.data)=="deepskyblue"] <- "gerhard2018_7"
Idents(seuratObj_macs) <- "cluster_perslab"
levels(seuratObj_macs) <- rev(c("5-Macrophages", "6-Macrophages","11-Macrophages", "14-Macrophages"))
```

```{r}

vec_genesPlot = c(#"S100A8",
                  #"S100A9",
                  #"S100A12",
                  "CD9", # SAM
                  "TREM2",#SAM
                  "MARCO", #Kuppfer
                  #"CD163", #Kuppfer
                  #"CD5L",# Kuppfer
                  #"TIMD4", #Kuppfer
                  "MNDA",#monocyte
                  "LYZ",#inflammatory
                  "CSTA" #inflammatory
                  #"CD74",#inflammatory
                  #"VCAN" # profibrotic
                  )
                   # top genes in "honeydew3" macrophage module
```

```{r}
list_plot_mac <- VlnPlot(object=seuratObj_macs,
                     assay="SCT_combat",
                     features = vec_genesPlot,
                     pt.size = 0,# 0.05,
                     cols=vec_colors_cluster_perslab_macs,#vec_colors_markers,
                     sort = F, 
                     #group.by = "cluster_perslab_coarse", 
                     same.y.lims = F,
                     slot="data",
                     log = F,
                     combine = F,
                     flip=F)#,
                     #ncol=length(vec_markers))

names(list_plot_mac) <- vec_genesPlot

list_plot_mac_flip <- lapply(1:length(list_plot_mac), function(i) {
  plot_tmp=list_plot_mac[[i]]
  plot_tmp = plot_tmp +
  coord_flip() +
  theme(
    plot.title = element_text(size=fontSize_plotTitle, face="bold.italic"),
    axis.text.y=element_blank(),
    plot.margin = margin(b=1, unit="cm")) #margin around entire plot (unit with the sizes of the top, right, bottom, and left margins)

  if (i==1) {
  plot_tmp <- plot_tmp +
  theme(
    axis.text.y=element_text(
    hjust=1,
    vjust=0,
    size=fontSize_axisText_lg,
    angle=30
    )
  )
  }
  # } else {
  #   plot_tmp <- plot_tmp + 
  #   theme(axis.text.y=element_text(
      # vjust=0.2,
      # size=10,
      # angle=30,
  #     margin = margin(l = 0, r = 5)
  #     )
  #     )
  # }
  
  plot_tmp <- plot_tmp + 
    theme(
      aspect.ratio=2,
      axis.title.y=element_blank(),
      axis.title.x=element_blank(), 
      axis.text.x = element_text(
        size= fontSize_axisText_sm,
        angle=35),
      plot.title = element_text(size=fontSize_plotTitle),
      legend.position="none")
})

p5C <- patchwork::wrap_plots(... = list_plot_mac_flip, 
                            ncol = length(list_plot_mac_flip)
                            )

ggsave(plot= p5C , filename = here("plots", paste0(prefixData, "_", prefixRun, "_fig5c_mac_genes_vlnplot_",flagDate,".pdf")))#, width = 12, height=12)
```

## Fig. 5 (D)

```{r}
vec_modsPlot = c( "Macrophages_1","Macrophages_2","Macrophages_3","Macrophages_4") # 2 and 4 are associated with lob inflam
vec_modsPlot_abbrev = paste0("M-Mac-",1:length(vec_modsPlot))
```

```{r}
list_plot_mac_mods <- VlnPlot(object=seuratObj_macs,
                     assay="SCT_combat",
                     features = vec_modsPlot,
                     pt.size = 0,# 0.05,
                     cols=vec_colors_cluster_perslab_macs,#vec_colors_markers,
                     sort = F, 
                     #group.by = "cluster_perslab_coarse", 
                     same.y.lims = F,
                     slot="data",
                     log = F,
                     combine = F,
                     flip=F)#,

names(list_plot_mac_mods) <- vec_modsPlot

list_plot_mac_mods_flip <- lapply(1:length(list_plot_mac_mods), function(i) {
  plot_tmp=list_plot_mac_mods[[i]]
  plot_tmp = plot_tmp +
  coord_flip() +
  ggtitle(label=vec_modsPlot_abbrev[i]) +
  theme(
    plot.title = element_text(size=fontSize_plotTitle, face="bold"),
    axis.text.y=element_blank(),
    plot.margin = margin(b=1, unit="cm")) #margin around entire plot (unit with the sizes of the top, right, bottom, and left margins)

  if (i==1) {
  plot_tmp <- plot_tmp +
  theme(
    axis.text.y=element_text(
    hjust=1,
    vjust=0,
    size=fontSize_axisText_lg,
    angle=30
    )
  )
  }
  # } else {
  #   plot_tmp <- plot_tmp + 
  #   theme(axis.text.y=element_text(
      # vjust=0.2,
      # size=10,
      # angle=30,
  #     margin = margin(l = 0, r = 5)
  #     )
  #     )
  # }
  
  plot_tmp <- plot_tmp + 
    theme(
      aspect.ratio=2,
      axis.title.y=element_blank(),
      axis.title.x=element_blank(), 
      axis.text.x = element_text(
        size= fontSize_axisText_sm,
        angle=35),
      plot.title = element_text(size=fontSize_plotTitle),
      legend.position="none")
})

p5D <- patchwork::wrap_plots(... = list_plot_mac_mods_flip, 
                            ncol = length(list_plot_mac_mods_flip)
                            )

p5D

ggsave(plot= p5D, filename = here("plots", paste0(prefixData, "_", prefixRun, "_fig5c_mac_mods_vlnplot_",flagDate,".pdf")))#, width = 12, height=12)
```


<!-- ```{r} -->
<!-- list_p <- VlnPlot(object = seuratObj_macs, -->
<!--               assay= "SCT_combat", -->
<!--               pt.size=0.05, -->
<!--               ncol = length(vec_modsPlot ), -->
<!--               cols=vec_colors_cluster_perslab_macs,#vec_colors_markers, -->
<!--               features = vec_modsPlot,   -->
<!--               #group.by = "cluster_perslab" -->
<!--               combine = F) -->

<!-- list_p <- lapply(list_p, function(p) { -->
<!--   p + theme( -->
<!--     plot.title = element_text(size=15), -->
<!--     axis.title.x= element_blank(),  -->
<!--     legend.position="none", -->
<!--     plot.background=element_blank(),   -->
<!--     plot.margin = margin(l=1.5, unit="cm")) -->
<!-- }) -->

<!-- p1 <- patchwork::wrap_plots(... = list_p,  -->
<!--                             ncol = length(list_p) -->
<!--                             ) -->

<!-- p1 -->



<!-- ggsave(plot=p1, filename = here("plots", paste0(prefixData, "_", prefixRun, "_macs_modules_vlnplots_cluster_perslab_", flagDate, ".pdf"))) -->
<!-- ``` -->

## Fig. 5 (G) and (H) M-Mac-3 VlnPlots



```{r}
Idents(seuratObj) <- "cluster_perslab"
levels(seuratObj) <- unique(seuratObj$cluster_perslab)[order(gsub("\\d+-","", unique(seuratObj$cluster_perslab)), decreasing=F)]
```

### Fig. 5 (G) 

```{r}
DefaultAssay(seuratObj) <- "SCT_combat"
maxy = max(seuratObj$Macrophages_3)
```

```{r}
vlnplot_MMac3 <- VlnPlot(object=seuratObj,
                     assay="SCT_combat",
                     features = "Macrophages_3",
                     pt.size = 0,# 0.05,
                     cols=vec_colors_cluster_perslab,#vec_colors_markers,
                     sort = F, 
                     #group.by = "cluster_perslab_coarse", 
                     same.y.lims = F,
                     slot="data",
                     log = F,
                     combine = T,#F,
                     flip=F)#,
vlnplot_MMac3 <- vlnplot_MMac3 +
    ylim(0, round(maxy)) + 
      
  ggtitle(label = "M-Mac-3") + 
    theme(
      axis.text.y=element_text(hjust=1,vjust=0,size=fontSize_axisText_md,angle=35),
     # aspect.ratio=3,
      axis.title.y=element_blank(),
      axis.title.x=element_blank(), 
      axis.text.x = element_text(size= fontSize_axisText_lg,angle=35),
      #plot.title = element_text(size=fontSize_plotTitle),
        #aspect.ratio=2,
      legend.position="none",
    plot.title = element_text(size=fontSize_plotTitle*2, face="bold"),
    plot.margin = margin(l=3, unit="cm")
    )
vlnplot_MMac3
# ggsave(plot= vlnplot_MMac3, filename = here("plots", paste0(prefixData, "_", prefixRun, "_fig5G_MMac3_perslab_",flagDate,".pdf")))#, width = 12, height=12)
```


### Fig. 5 H

```{r}
Idents(seuratObj_macparland) <- "Cluster_annot"
levels(seuratObj_macparland) <- c(
  "Mature_B_cells",
  "Cholangiocytes",
  "Portal_endothelial_cells",
  "Periportal_LSECs",
  "Central_venous_LSECs",
  "Hepatic_stellate_cells",
  "Erythroid_cells",
  "Hep_1",
  "Hep_2",
  "Hep_3",
  "Hep_4",
  "Hep_5",
  "Hep_6",
  "Non_inflammatory_Macs",
  "Inflammatory_Macs",
  "NK_like_cells",
  "Plasma_cells",
  "Alpha_Beta_T_cells",
  "Gamma_delta_T_cells_1",
  "Gamma_delta_T_cells_2")

```

```{r}
vec_colors_cluster_macp = c(wes_palette("BottleRocket1", n=7), wes_palette("BottleRocket2", 5), wes_palette("Rushmore1",5), wes_palette("Royal1",3))

names(vec_colors_cluster_macp) = levels(seuratObj_macparland)
```

```{r}
vlnplot_MMac3_macparland <- VlnPlot(object=seuratObj_macparland,
                     assay="SCT",
                     features = "Macrophages_3",
                     pt.size = 0,# 0.05,
                  cols=vec_colors_cluster_macp,#vec_colors_markers,
                     sort = F, 
                     #group.by = "cluster_perslab_coarse", 
                     same.y.lims = F,
                     slot="data",
                     log = F,
                     combine = T,#F,
                     flip=F)#,
vlnplot_MMac3_macparland<- vlnplot_MMac3_macparland+
 ylim(0,round(maxy)) +
  ggtitle(label = "M-Mac-3") + 
    # scale_y_continuous(limits=c(-(max_lim_y+0.2), max_lim_y), breaks=seq(-2.5,2,0.5)) +
    theme(
      axis.text.y=element_text(hjust=1,vjust=0,size=fontSize_axisText_md,angle=35),
     # aspect.ratio=3,
      axis.title.y=element_blank(),
      axis.title.x=element_blank(), 
      axis.text.x = element_text(size= fontSize_axisText_lg,angle=35),
      #plot.title = element_text(size=fontSize_plotTitle),
        #aspect.ratio=2,
      legend.position="none",
    plot.title = element_text(size=fontSize_plotTitle*2, face="bold"),
    plot.margin = margin(l=3, unit="cm")
    )
vlnplot_MMac3_macparland
# ggsave(plot= vlnplot_MMac3, filename = here("plots", paste0(prefixData, "_", prefixRun, "_fig5G_MMac3_perslab_",flagDate,".pdf")))#, width = 12, height=12)
```

write out module embeddings to disk

```{r}
dt_mod_embed_macparland = data.table("cell_id" = rownames(seuratObj_macparland@meta.data), "Cluster_annot" =seuratObj_macparland@meta.data$Cluster_annot, seuratObj_macparland@meta.data[,colnames(seuratObj_macparland@meta.data) %in% gsub("-",".",dt_geneMod_perslab$module_renamed)])
```

```{r}
fwrite(dt_mod_embed_macparland, file="/projects/jonatan/pub-perslab/18-liver-wgcna/tables/liver_perslab_int_wgcna3_mat_scaled_embed_macparland.csv.gz", compress="gzip")
```

```{r}
openxlsx::write.xlsx(dt_mod_embed_macparland, file="/projects/jonatan/pub-perslab/18-liver-wgcna/tables/liver_perslab_int_wgcna3_mat_scaled_embed_macparland.csv.xlsx")
```


## Fig. 6 (?): violin plots of bulk module expression in coarse clusters

```{r}
if (F) {
  dict_mods_gerhard = dt_geneMod_gerhard[,.(module, module_renamed)]
    
  dict_mods_gerhard = dict_mods_gerhard[!duplicated(module_renamed) & !is.na(module_renamed) & nchar(module_renamed)>0]
    
  dict_mods_gerhard$module_abbrev = gsub("gerhard2018_", "M-Ger-",dict_mods_gerhard$module_renamed)
  
  for (old_modname in dict_mods_gerhard$module) {
    print(old_modname)
    if (old_modname %in% colnames(seuratObj@meta.data)) {
        colnames(seuratObj@meta.data)[grep(old_modname, colnames(seuratObj@meta.data))] <-  dict_mods_gerhard$module_abbrev[match(old_modname, dict_mods_gerhard$module)]
    }
  }

}
```


```{r}
Idents(seuratObj) = "cluster_perslab_coarse"
levels(seuratObj) = sort(names(vec_colors_cluster_perslab_coarse), decreasing = T)
```


```{r}
list_plot <- VlnPlot(object=seuratObj,
                     assay="SCT_combat",
                     features = gsub("-","\\.",dict_mods_gerhard$module_abbrev),
                     pt.size = 0,# 0.05,
                     cols=vec_colors_cluster_perslab_coarse,#vec_colors_markers,
                     sort = F, 
                     #group.by = "cluster_perslab_coarse", 
                     same.y.lims = F,
                     slot="data",
                     log = F,
                     combine = F,
                     flip=F)#,
                     #ncol=length(vec_markers))

names(list_plot) <- dict_mods_gerhard$module_abbrev

list_plot_flip <- lapply(1:length(list_plot), function(i) {
  plot_tmp=list_plot[[i]]
  plot_tmp = plot_tmp +
   ggtitle(label=dict_mods_gerhard$module_abbrev[i]) + 
    coord_flip() +
  theme(
    plot.title = element_text(face = "bold", size=fontSize_plotTitle*2),
    axis.text.y=element_blank(),
    plot.margin = margin(b=1, unit="cm")) #margin around entire plot (unit with the sizes of the top, right, bottom, and left margins)

  if (i==1) {
  plot_tmp <- plot_tmp +
  theme(
    axis.text.y=element_text(
    hjust=1,
    vjust=0,
    size=fontSize_axisText_lg,
    angle=30
    )
  )
  }
  # } else {
  #   plot_tmp <- plot_tmp + 
  #   theme(axis.text.y=element_text(
      # vjust=0.2,
      # size=10,
      # angle=30,
  #     margin = margin(l = 0, r = 5)
  #     )
  #     )
  # }
  
  plot_tmp <- plot_tmp + 
    theme(
      axis.title.y=element_blank(),
      axis.title.x=element_blank(), 
      axis.text.x = element_text(
        size= fontSize_axisText_xsm,
        angle=35),
      plot.title = element_text(size=fontSize_plotTitle),
      #aspect.ratio=10,
      legend.position="none")
})

p6B <- patchwork::wrap_plots(... = list_plot_flip, 
                            ncol = length(list_plot_flip)
                            )

```

Fig. 6. How similar are the bulk modules to single cell and metamodules?

```{r}
vec_mods_perslab <- dt_geneMod_perslab[,module_renamed] %>% unique 
vec_mods_perslab <- vec_mods_perslab[!is.na(vec_mods_perslab) & nchar(vec_mods_perslab)>0]

vec_mods_meta <- dt_geneMod_perslab[,module_meta] %>% unique 
vec_mods_meta <- vec_mods_meta[!is.na(vec_mods_meta) & nchar(vec_mods_meta)>0] %>% sort

vec_mods_gerhard <- dt_geneMod_gerhard[,module_renamed] %>% unique 

```


```{r}

list_genelists_perslab = lapply(vec_mods_perslab, function(mod) {
  condition = quote(dt_geneMod_perslab[["module_renamed"]]==mod)
  vec_mod_genes = dt_geneMod_perslab[eval(condition), ..colGeneNames, ][[1]] %>% unique
  return(vec_mod_genes)
})

names(list_genelists_perslab) = vec_mods_perslab
```

```{r}

list_geneWeights_perslab = lapply(vec_mods_perslab, function(mod) {
  condition = quote(dt_geneMod_perslab[["module_renamed"]]==mod)
  vec_mod_geneWeights= dt_geneMod_perslab[eval(condition), pkMs,]
  names(vec_mod_geneWeights) = dt_geneMod_perslab[eval(condition), ..colGeneNames, ][[1]]
  return(vec_mod_geneWeights)
})

names(list_geneWeights_perslab) = vec_mods_perslab
```


```{r}

list_genelists_meta = lapply(vec_mods_meta, function(mod) {
  condition = quote(dt_geneMod_perslab[["module_meta"]]==mod)
  vec_mod_genes = dt_geneMod_perslab[eval(condition), ..colGeneNames, ][[1]] %>% unique
  return(vec_mod_genes)
})

names(list_genelists_meta) = vec_mods_meta
```



```{r}

list_genelists_gerhard = lapply(vec_mods_gerhard, function(mod) {
  condition = quote(dt_geneMod_gerhard[["module_renamed"]]==mod)
  vec_mod_genes = dt_geneMod_gerhard[eval(condition), ..colGeneNames, ][[1]] %>% unique
  return(vec_mod_genes)
})

names(list_genelists_gerhard) = vec_mods_gerhard
```

the intersect matrix tells you how big a proportion of the col module's genes are also in the row module

```{r}

# outer loop is columns, inner is rows
fun = function(genelist) {
    sapply(list_genelists_perslab, function(genelistOther) {
      base::intersect(x=genelist, y=genelistOther) %>% length %>% '/'(length(genelist))
    }, simplify=T)
  }
  
mat_moduleGeneIntersect_gerhard_perslab <- sapply(FUN=fun,"X"=list_genelists_gerhard, simplify = T)

#mat_moduleGeneIntersect_gerhard_perslab[row(mat_moduleGeneIntersect_gerhard_perslab)==col(mat_moduleGeneIntersect_gerhard_perslab)] <- 0 
#<- mat_moduleGeneIntersect-diag(1, nrow=nrow(mat_moduleGeneIntersect))
```

```{r}
head(sort(as.numeric(mat_moduleGeneIntersect_gerhard_perslab), decreasing = T))

#     Min.  1st Qu.   Median     Mean  3rd Qu.     Max. 
#   0.00000 0.00000 0.00000 0.07862 0.02247 0.95833
```

```{r}

p_heatmap_moduleGeneIntersect_gerhard_perslab = pheatmap(mat_moduleGeneIntersect_gerhard_perslab, 
                 cluster_rows = T, 
                 cluster_cols = T,
                 legend=T)#, treeheight_row=0, treeheight_col=0)


#here("plots", paste0(prefixData, "_", prefixRun, "_Fig2Aii_corr_mat_",flagDate,".pdf"))

```

```{r}

# outer loop is columns, inner is rows
fun = function(genelist) {
    sapply(list_genelists_meta, function(genelistOther) {
      base::intersect(x=genelist, y=genelistOther) %>% length %>% '/'(length(genelist))
    }, simplify=T)
  }
  
mat_moduleGeneIntersect_gerhard_meta <- sapply(FUN=fun,"X"=list_genelists_gerhard, simplify = T)

#mat_moduleGeneIntersect_gerhard_meta[row(mat_moduleGeneIntersect_gerhard_meta)==col(mat_moduleGeneIntersect_gerhard_meta)] <- 0 
#<- mat_moduleGeneIntersect-diag(1, nrow=nrow(mat_moduleGeneIntersect))
```

```{r}
head(sort(as.numeric(mat_moduleGeneIntersect_gerhard_meta), decreasing = T))

#     Min.  1st Qu.   Median     Mean  3rd Qu.     Max. 
#   0.00000 0.00000 0.00000 0.07862 0.02247 0.95833
```

```{r}

p_heatmap_moduleGeneIntersect_gerhard_meta = pheatmap(mat_moduleGeneIntersect_gerhard_meta, 
                 cluster_rows = T, 
                 cluster_cols = T,
                 legend=T)#, treeheight_row=0, treeheight_col=0)


#here("plots", paste0(prefixData, "_", prefixRun, "_Fig2Aii_corr_mat_",flagDate,".pdf"))

```

So the overlap with single cell and meta modules is mainly between Gerhard 7 and metamodule 4.

Is there then an overlap with differentially expressed genes for cell clusters?

```{r}
DefaultAssay(seuratObj)  = "SCT_combat"
```


```{r}
list_df_markers_cluster_perslab = safeParallel(
  fun=function(clust) {
    FindMarkers(object = seuratObj, 
               slot="data",
               group.by = "cluster_perslab", 
               ident.1 = clust,
                only.pos = T,
                test.use  ="negbinom",
                max.cells.per.ident=1000,
                random.seed=randomSeed,
                verbose = T)
  }, 
  list_iterable = list("clust"=sort(unique(seuratObj$cluster_perslab))), 
  simplify=F, 
  n_cores = 19, 
  timeout = 120)

names(list_df_markers_cluster_perslab) = sort(unique(seuratObj$cluster_perslab))

```

```{r}
DefaultAssay(seuratObj) = "SCT_combat"
```

```{r}
list_df_markers_cluster_perslab_coarse = readRDS("/projects/jonatan/pub-perslab/18-liver-fred/output/liver_perslab_int_SCTint_combat_list_wilcox_all_markers_cluster_perslab_coarse.RDS.gz")
# list_df_markers_cluster_perslab_coarse = safeParallel(
#   fun=function(clust) {
#     FindMarkers(object = seuratObj, 
#                slot="counts",
#                group.by = "cluster_perslab_coarse", 
#                ident.1 = clust,
#                 only.pos = T,
#                 test.use  ="negbinom",
#                 max.cells.per.ident=1000,
#                 random.seed=randomSeed,
#                 verbose = T)
#   }, 
#   list_iterable = list("clust"=sort(unique(seuratObj$cluster_perslab_coarse))), 
#   simplify=F, 
#   n_cores = 19, 
#   timeout = 120)
# 
# names(list_df_markers_cluster_perslab_coarse) = sort(unique(seuratObj$cluster_perslab_coarse))
# 
# 

```


```{r}
n_max_markers_use = 100
```

```{r}
list_vec_topmarkers =   lapply(names(list_df_markers_cluster_perslab_coarse), function(clust_name) {
  vec_markers = rownames(list_df_markers_cluster_perslab[[clust_name]])[list_df_markers_cluster_perslab[[clust_name]]$p_val_adj < 0.05]
  vec_markers[1:min(n_max_markers_use, length(vec_markers))]
})

names(list_vec_topmarkers) = names(list_df_markers_cluster_perslab)
```

check overlaps gerhard marker genes

```{r}

# outer loop is columns, inner is rows
fun = function(genelist) {
    sapply(list_vec_topmarkers, function(vec_markers) {
      base::intersect(x=genelist, y=vec_markers) %>% length %>% '/'(length(genelist))
    }, simplify=T)
  }
  
mat_moduleGeneIntersect_gerhard_clustermarkers <- sapply(FUN=fun,"X"=list_genelists_gerhard, simplify = T)


#<- mat_moduleGeneIntersect-diag(1, nrow=nrow(mat_moduleGeneIntersect))
```

```{r}
head(sort(as.numeric(mat_moduleGeneIntersect_gerhard_clustermarkers), decreasing = T))

#     Min.  1st Qu.   Median     Mean  3rd Qu.     Max. 
#   0.00000 0.00000 0.00000 0.07862 0.02247 0.95833
```

```{r}

p_heatmap_moduleGeneIntersect_gerhard_clustermarkers = pheatmap(mat_moduleGeneIntersect_gerhard_clustermarkers, 
                 cluster_rows = T, 
                 cluster_cols = T,
                 legend=T)#, treeheight_row=0, treeheight_col=0)


#here("plots", paste0(prefixData, "_", prefixRun, "_Fig2Aii_corr_mat_",flagDate,".pdf"))

```

check overlaps sc modules with marker genes


```{r}

# outer loop is columns, inner is rows
fun = function(genelist) {
    sapply(list_vec_topmarkers, function(vec_markers) {
      base::intersect(x=genelist, y=vec_markers) %>% length %>% '/'(length(genelist))
    }, simplify=T)
  }
  
mat_moduleGeneIntersect_perslab_clustermarkers <- sapply(FUN=fun,"X"=list_genelists_perslab, simplify = T)

#mat_moduleGeneIntersect_perslab_clustermarkers[row(mat_moduleGeneIntersect_perslab_clustermarkers)==col(mat_moduleGeneIntersect_perslab_clustermarkers)] <- 0 
#<- mat_moduleGeneIntersect-diag(1, nrow=nrow(mat_moduleGeneIntersect))
```

```{r}
head(sort(as.numeric(mat_moduleGeneIntersect_perslab_clustermarkers), decreasing = T))

#     Min.  1st Qu.   Median     Mean  3rd Qu.     Max. 
#   0.00000 0.00000 0.00000 0.07862 0.02247 0.95833
```

```{r}

p_heatmap_moduleGeneIntersect_perslab_clustermarkers = pheatmap(mat_moduleGeneIntersect_perslab_clustermarkers, 
                 cluster_rows = T, 
                 cluster_cols = T,
                 legend=T)#, treeheight_row=0, treeheight_col=0)


#here("plots", paste0(prefixData, "_", prefixRun, "_Fig2Aii_corr_mat_",flagDate,".pdf"))

```

## plot which of the top marker genes are in the sc modules

```{r}
mat_gene_markers_in_mod = sapply(list_geneWeights_perslab, function(vec_geneWeights) {
   vec_geneWeights_rank = order(vec_geneWeights, decreasing=F)/length(vec_geneWeights)
   names(vec_geneWeights_rank) = names(vec_geneWeights)
    ifelse(list_vec_topmarkers$`10-Hepatocytes` %in% names(vec_geneWeights_rank), vec_geneWeights_rank, 0)
  })

rownames(mat_gene_markers_in_mod) <- list_vec_topmarkers$`10-Hepatocytes`
```

```{r}

p_heatmap_hep_clustermarkers_in_mod = pheatmap(mat_gene_markers_in_mod, 
                 cluster_rows = F, 
                 cluster_cols = T,
                 fontsize_row = 8, #default 10
                 legend=T,)#, treeheight_row=0, treeheight_col=0)


#here("plots", paste0(prefixData, "_", prefixRun, "_Fig2Aii_corr_mat_",flagDate,".pdf"))

```

## plot which of the top marker genes are in the sc modules

```{r}
mat_gene_markers_in_mod = sapply(list_geneWeights_perslab, function(vec_geneWeights) {
   vec_geneWeights_rank = order(vec_geneWeights, decreasing=F)/length(vec_geneWeights)
   names(vec_geneWeights_rank) = names(vec_geneWeights)
    ifelse(list_vec_topmarkers$`10-Hepatocytes` %in% names(vec_geneWeights_rank), vec_geneWeights_rank, 0)
  })

rownames(mat_gene_markers_in_mod) <- list_vec_topmarkers$`10-Hepatocytes`
```

```{r}

p_heatmap_hep_clustermarkers_in_mod = pheatmap(mat_gene_markers_in_mod, 
                 cluster_rows = F, 
                 cluster_cols = T,
                 fontsize_row = 6, #default 10
                 legend=T,)#, treeheight_row=0, treeheight_col=0)


#here("plots", paste0(prefixData, "_", prefixRun, "_Fig2Aii_corr_mat_",flagDate,".pdf"))

```

# Supplementary figures



## Suppl figure: cluster perslab (small clusters) UMAP

```{r}
p <- UMAPPlot(seuratObj_macparland, 
             #reduction = "umap",
             group.by="cluster_perslab", 
             seed=randomSeed,
             #label.box=TRUE,
             cols=vec_colors_cluster_perslab,
             label = T,
             label.size = fontSize_label_md,
             repel=T)

p <- p + theme(
  legend.position="none",
  axis.title.x = element_text(size=fontSize_axisTitle_md),
  axis.title.y = element_text(size=fontSize_axisTitle_md))

p

# saving unicode symbols to pdf doesn't work - copy png instead

#ggsave(plot= p, filename = here("plots", paste0(prefixData, "_", prefixRun, "_supplfig_umap_cluster_perslab_",flagDate,".pdf")), width = 12, height=12)
```


<!-- ```{r} -->
<!-- list_dt_condition_coef_sub = lapply(c("fibrosis","lob_inflam"), function(condition){ -->

<!--   condition_filter = if (condition=="fibrosis") quote(dt_condition_coef$condition=="fibrosis")  else quote(dt_condition_coef$condition=="lob_inflam") -->

<!--   dt_condition_coef_sub = dt_condition_coef[eval(condition_filter) & p.value_percentile_BH < params$pValThreshold] -->
<!--   # find matching cell type -->
<!--   vec_celltype <- dt_geneMod_comb[[paste0("cell_cluster_", condition, "_gerhard2018")]][sapply(dt_condition_coef_sub$module, function(eachmod) grep(eachmod, gsub("-","_",dt_geneMod_comb[[paste0("module_", condition, "_gerhard2018")]]))[1])]  -->

<!--   dt_condition_coef_sub[["celltype"]] = vec_celltype -->

<!--   # order rows -->
<!--   dt_condition_coef_sub = dt_condition_coef_sub[order(celltype,module)] -->
<!--   orderToUse1 = 1:length(vec_celltype) -->
<!--   # force row order with factors -->
<!--   dt_condition_coef_sub[["celltype"]] = factor(dt_condition_coef_sub[["celltype"]], ordered = T) -->
<!--   dt_condition_coef_sub[["celltype"]] = reorder(x=dt_condition_coef_sub[["celltype"]], X = orderToUse1, order=T) -->

<!--   dt_condition_coef_sub[["module"]] = factor(dt_condition_coef_sub[["module"]], ordered = T) #factor(gsub("perslab_","",dt_condition_coef_sub[["module"]]), ordered = T) -->
<!--   dt_condition_coef_sub[["module"]] = reorder(x=dt_condition_coef_sub[["module"]], X = orderToUse1, order=T) -->
<!--   # color bars by cell type -->
<!--   vec_color_celltype = vec_colors_cluster_perslab_coarse[vec_celltype]  %>% factor(.,ordered=T) -->

<!--   #freqs = table(dt_condition_coef_sub[["celltype"]]) -->

<!--   #vec_color_tmp = rep(x = vec_color_unique_tmp, times = freqs) %>% factor(.,ordered=T) -->
<!--   dt_condition_coef_sub[["color"]] = vec_color_celltype -->
<!--   return(dt_condition_coef_sub) -->
<!-- }) -->

<!-- names(list_dt_condition_coef_sub) = c("fibrosis","lob_inflam") -->
<!-- ``` -->

<!-- plot  -->

<!-- ```{r} -->
<!-- lapply(c("fibrosis","lob_inflam"), function(condition) { -->
<!--   dt_condition_coef_sub = list_dt_condition_coef_sub[[condition]] -->
<!--   p <-  -->
<!--     ggplot(data=dt_condition_coef_sub, aes(x=module, y=estimate_original, fill=celltype)) + -->
<!--     geom_bar(stat="identity", position=position_identity(), group=dt_condition_coef_sub$celltype) +  -->
<!--     #geom_errorbar(aes(ymin=estimate-std.error, ymax=estimate+std.error), width=.1) + -->
<!--     geom_errorbar(aes(ymin=confInt_lower_percentile, ymax=confInt_upper_percentile), width=.1) + -->
<!--     #scale_fill_brewer(palette="Paired")+ -->
<!--     scale_fill_manual(values=vec_colors_cluster_perslab_coarse[unique(dt_condition_coef_sub$celltype)]) + # this is a kluge -->
<!--     #scale_fill_manual(values=unique(df_data$celltype)) + # this is a kluge -->
<!--     #geom_hline(yintercept=cutoff, linetype="dashed", color = "red") + -->
<!--     scale_y_continuous("estimate_original", expand = c(0, 0)) + -->
<!--     scale_x_discrete("module") +   -->
<!--     # geom_text(aes(label=module), -->
<!--     #         #angle=90, -->
<!--     #         #vjust=-1, -->
<!--     #         hjust = 2, -->
<!--     #         color="black", -->
<!--     #         angle=90, -->
<!--     #        position = position_dodge(),  -->
<!--     #        size=2.5) + -->
<!--     theme(axis.text.x = element_text(angle=90,hjust=1),#element_blank(), -->
<!--         axis.line.x = element_blank(), -->
<!--         axis.ticks.x = element_blank(), -->
<!--         panel.grid.major = element_blank(),  -->
<!--         panel.grid.minor = element_blank()) -->

<!--   saveMeta(savefnc=ggsave,plot = p, -->
<!--            filename = paste0(dirWGCNA_plots,prefixData,"_", prefixRun,  "_sc_mod_", condition, "_assoc_gerhard_", params$date,".pdf"), height=10, width =15) -->
<!-- }) -->
<!-- ``` -->




  
###################################################################################################
############################################# LEFTOVERS ###########################################
###################################################################################################


~~# Fig.3: Metamodules~~

## Fig. 3 (A): dotplot coloured by metamodule

```{r}
set.seed(randomSeed)

p_dot_meta = ggplot(data=dt_point, mapping=aes(lob_inflam_estimate_original, fibrosis_estimate_original, color=metamodule, shape=significance)) + 
  geom_point(size=6, alpha=0.9) + 
  
  scale_x_continuous(limits = c(-(max_lim_x+0.4), max_lim_x), breaks=seq(-1.5,1,0.5)) + 
  scale_y_continuous(limits=c(-(max_lim_y+0.2), max_lim_y), breaks=seq(-2.5,2,0.5)) +
  
  coord_fixed() +
  
  ggrepel::geom_text_repel(
  #geom_text(
      #nudge_x = -1.8-dt_point$lob_inflam#[lob_inflam_estimate_original>1 | fibrosis_estimate_original>1]$lob_inflam_estimate_original,  
      #nudge_y      = 0,
      #parse=T,
      direction    = "both",
      segment.color = "grey50",
      #angle        = 0,
      #hjust        = 0,
      point.padding=NA,
      segment.size = 0.2,
      data = dt_point,#[lob_inflam_estimate_original>1 | fibrosis_estimate_original>1],
      mapping = aes(
        #x=lob_inflam_estimate_original,
        #y= fibrosis_estimate_original,
        label=stringr::str_wrap(module_abbrev)),
        #color=dt_point[abs(lob_inflam_estimate_original)>0.8 | abs(fibrosis_estimate_original)>0.8]$color), 
      color="black",
      size=fontSize_label_sm, 
      #fontface="bold"
      ) +
  
   # 
   # ggrepel::geom_text_repel(
   #    nudge_x = 0.9-dt_point[lob_inflam_estimate_original<(-1) | fibrosis_estimate_original<(-1)]$lob_inflam_estimate_original,  
   #    nudge_y      = 0,
   #    direction    = "y",
   #    segment.color = "grey50",
   #    #angle        = 0,
   #    hjust        = 0,
   #    segment.size = 0.2,
   #    data = dt_point[lob_inflam_estimate_original<(-1) | fibrosis_estimate_original<(-1)],
   #    mapping = aes(
   #      x=lob_inflam_estimate_original,
   #      y= fibrosis_estimate_original,
   #      label=stringr::str_wrap(module)),
   #      #color=dt_point[abs(lob_inflam_estimate_original)>0.8 | abs(fibrosis_estimate_original)>0.8]$color), 
   #    color="black",
   #    size=2.5, 
   #    #fontface="bold"
   #    ) +
  
  #labs(x = "z-score", y="-log(p-value)", colour="z-score") +
  #labs(x = "lobular inflammation", y="fibrosis") +
  xlab(expression(paste(bold("\u03B2")," lobular inflammation")))+
  ylab(expression(paste(bold("\u03B2")," fibrosis")))+
  
  scale_color_manual(
    values=vec_colors_meta) +#,
  
  scale_shape_manual(values=c(15, 16, 17))+
  
  # set legend order and reverse the order of shape legend
  guides(
    colour = guide_legend(order=1),
    shape = guide_legend(reverse=TRUE, order=2)) + 
  
    #labels=levels(dt_point$celltype)) + 
  geom_hline(yintercept = 0, size=0.1) +
  geom_vline(xintercept = 0, size=0.1) + 
 theme(axis.text = element_text(size = fontSize_axisText_sm), 
        axis.line = element_line(colour = "grey80"), 
        axis.ticks = element_line(colour = "grey80"), 
        axis.title = element_text(size = fontSize_axisTitle_sm),#, face = "bold"), 
        panel.background = element_blank(), 
        #legend.position = "none",
        #panel.grid.minor = element_blank(), 
        #panel.grid.major = element_line(colour = "grey80"), 
        plot.background = element_blank(), 
        panel.grid = element_blank(),
        text = element_text(size = fontSize_axisText_md),
        legend.title=element_text(size=fontSize_legend_lg),
        legend.text = element_text(size=fontSize_legend_md),
        #aspect.ratio = 1
        ) 
# 
  p_dot_meta
  
  # ggsave isn't rendering unicode - save from canvas as image
  # 1200 * 544
  
  # ggsave(plot = p, filename = paste0(dirWGCNA_plots,prefixData,"_", prefixRun,  "_fig2D_sc_mod_association_dotplot_metamod_", params$date,".pdf"), height=10, width =12)
```



## Fig 3 (B) left: metamodule correlation matrix

```{r}
dt_scaled_embed_metamod = fread(paste0(dirWGCNA_tables, "liver_perslab_int_wgcna3_modmeta_embed_gerhard2018.csv.gz"))


mat_scaled_embed_metamod = dt_scaled_embed_metamod[,-1] %>% as.matrix
rownames(mat_scaled_embed_metamod) = dt_scaled_embed_metamod$run_accession

colnames(mat_scaled_embed_metamod) <- paste0("MM-",1:5)
mat_scaled_embed_metamod[0:4,0:5]

```

```{r}
mat_cor_metamod = cor(mat_scaled_embed_metamod)

#mat_cor_metamod[row(mat_cor_metamod)==col(mat_cor_metamod)] <- 0
```

```{r}
col3 <- colorRampPalette(c("blue", "white", "red"))

pdf(file = paste0(dirWGCNA_plots, prefixData, "_", prefixRun, "_fig3B_corrplot_metamod_", flagDate,".pdf"), width = 6, height=4)

p <- corrplot::corrplot(mat_cor_metamod, 
                        method = "shade", 
                        col = col3(50),
                        type = "lower", 
                        diag = T, 
                        mar= c(2, 2, 3, 2), # c(bottom, left, top, right) 
                        order = "original", 
                        tl.cex=0.7,
                        cl.cex=0.6)

dev.off()
```


### try single cell correlation matrix

```{r}

mat_scaled_embed_metamod_sc = seuratObj@meta.data[,grep("metamodule", colnames(seuratObj@meta.data))]

colnames(mat_scaled_embed_metamod_sc) <- gsub("metamodule", "MM-", colnames(mat_scaled_embed_metamod))
mat_scaled_embed_metamod_sc[0:4,0:5]

```

```{r}
mat_cor_metamod_sc = cor(mat_scaled_embed_metamod_sc)

#mat_cor_metamod[row(mat_cor_metamod)==col(mat_cor_metamod)] <- 0
```

```{r}
col3 <- colorRampPalette(c("blue", "white", "red"))

pdf(file = paste0(dirWGCNA_plots, prefixData, "_", prefixRun, "_fig3B_corrplot_metamod_in_sc_", flagDate,".pdf"), width = 6, height=4)

p <- corrplot::corrplot(mat_cor_metamod_sc, 
                        method = "shade", 
                        col = col3(50),
                        type = "lower", 
                        diag = T, 
                        mar= c(2, 2, 3, 2), # c(bottom, left, top, right) 
                        order = "original", 
                        tl.cex=0.7,
                        cl.cex=0.6)

dev.off()
```

## Fig 3 (B) right: metamodule overlap Euler (Venn)  diagram 

```{r}

list_genes_meta = lapply(paste0("metamodule_", 1:5), function(mod_meta) {
    dt_geneMod_perslab[[colGeneNames]][dt_geneMod_perslab$module_meta==mod_meta]
  })


names(list_genes_meta) = paste0("metamodule_", 1:5)

# outer loop is columns, inner is rows
mat_overlap_meta = sapply(list_genes_meta, function(vec_genes) {
    sapply(list_genes_meta, function(vec_genes_other) {
      base::intersect(x=unique(vec_genes), y=unique(vec_genes_other)) %>% length %>% '/'(length(unique(vec_genes)))
    }, simplify=T)
  })

#mat_overlap_meta[row(mat_overlap_meta)==col(mat_overlap_meta)] <- 0
```

```{r}
max(mat_overlap_meta[mat_overlap_meta<1], na.rm = T)
# [1] 0.2
```

eulerr plot
source: https://cran.r-project.org/web/packages/eulerr/vignettes/introduction.html

```{r}
vec_genes_all = unlist(list_genes_meta, F) %>% unique

length(vec_genes_all)
# [1] 1699
```

```{r}

mat_euler <- lapply(list_genes_meta, function(x) vec_genes_all %in% x) %>% Reduce(f=rbind,x=.) %>% t
colnames(mat_euler) = paste0("MM-",1:5)
rownames(mat_euler) = vec_genes_all


```

```{r}
set.seed(randomSeed)
fit_euler <- euler(mat_euler)
```

check goodness of fit

```{r}

dotplot(resid(fit_euler), xlab = "",
        panel = function(...) {
          panel.abline(v = 0, lty = 2)
          panel.dotplot(...)
        })
```

```{r}
error_plot(fit_euler) 
# not too shabby
```

do final plot 

```{r}
pdf(file = paste0(dirWGCNA_plots, prefixData, "_", prefixRun, "_fig3B_euler_metamod_", flagDate,".pdf"), width = 6, height=6)

plot(fit_euler,
     quantities = TRUE,
     fill = vec_colors_meta,
     #lty = 1:3,
     labels = list(font = fontSize_label_lg))

dev.off()
```

## Fig 3. (C) i) GO barplots for meta modules 

```{r}
dt_resultsGO_metaMod = fread(paste0(dirWGCNA_tables, prefixData, "_", prefixRun, "_", "201118","_GERR_GOenrichmentResults_metaMod.csv"))
dt_resultsGO_metaMod$source = "GO"

# KEGG
dt_resultsKEGG_metaMod = fread(paste0(dirWGCNA_tables, prefixData, "_", prefixRun,  "_", "201118","_GERR_KEGGenrichmentResults_metaMod.csv"))
dt_resultsKEGG_metaMod$source = "KEGG"

# REACTOME
dt_resultsREACT_metaMod = fread(paste0(dirWGCNA_tables, prefixData, "_", prefixRun,  "_","201118" ,"_GERR_REACTOMEenrichmentResults_metaMod.csv"))
dt_resultsREACT_metaMod$source = "REACTOME"
```

< 
for each of the five metamodules in turn, feed GO ids and p-values into REVIGO with all default parameters 
>

```{r}

list_dt_resultsGO_metaMod_revigo = lapply(paste0("metaMod",c(1:3,5)), function(name) {
  dt_out =  fread(paste0(dirWGCNA_tables, prefixData, "_", prefixRun, "_", "201102","_GERR_GOenrichmentResults_revigo_", name, ".csv"))
 dt_out$metamodule = paste0(substr(x = tolower(name),1,7),"ule_", substr(name,8,8))
 return(dt_out)
                                          })

dt_resultsGO_metaMod_revigo = rbindlist(list_dt_resultsGO_metaMod_revigo)
dt_resultsGO_metaMod_revigo$source = "GO"
```


```{r}
dt_resultsGO_metaMod_revigo = dt_resultsGO_metaMod_revigo[eliminated == 0, .(metamodule,term =description,source)]

dt_resultsGO_metaMod_revigo$minlog10_p.value_fisher_BH = -log10(dt_resultsGO_metaMod$p.value_fisher_BH[match(dt_resultsGO_metaMod_revigo$term, dt_resultsGO_metaMod$GO_term)])

dt_resultsKEGG_metaMod_plot = dt_resultsKEGG_metaMod[,.(metamodule=module, term =KEGG_term, minlog10_p.value_fisher_BH=-log10(p.value_fisher_BH),source="KEGG")]

dt_resultsREACT_metaMod_plot = dt_resultsREACT_metaMod[,.(metamodule=module, term =REACT_name, minlog10_p.value_fisher_BH=-log10(p.value_fisher_BH),source="REACTOME")]

```

```{r}
dt_results_GERR_metaMod_plot = rbindlist(list(dt_resultsGO_metaMod_revigo, dt_resultsKEGG_metaMod_plot, dt_resultsREACT_metaMod_plot), use.names=T) # columns aren't in the same order in REVIGO, hence match by colnames

dt_results_GERR_metaMod_plot = dt_results_GERR_metaMod_plot[order(metamodule, source,-minlog10_p.value_fisher_BH)]

dt_results_GERR_metaMod_plot = dt_results_GERR_metaMod_plot[minlog10_p.value_fisher_BH>-log10(0.05)]

```


```{r}

list_p_GO = lapply(unique(dt_results_GERR_metaMod_plot$metamodule), function(metamod) {
    
   # subset table by metamodule
    dt_results_GERR_metaMod_plot_sub = dt_results_GERR_metaMod_plot[metamodule==metamod, with=F, j=colnames(dt_results_GERR_metaMod_plot)]
    
    # keep top five hits per source
    dt_results_GERR_metaMod_plot_sub <- dt_results_GERR_metaMod_plot_sub[,lapply(.SD,function(x) head(x,5)),by=source]
    
    dt_results_GERR_metaMod_plot_sub$truncate = sapply(dt_results_GERR_metaMod_plot_sub$term, function(x){
      nchar(x)>55
      })
    
    dt_results_GERR_metaMod_plot_sub$term[dt_results_GERR_metaMod_plot_sub$truncate] <- sapply(dt_results_GERR_metaMod_plot_sub$term[dt_results_GERR_metaMod_plot_sub$truncate], function(long_term) {
      
      strsplit_tmp = strsplit(x = long_term, split=" ")[[1]]
      shortened_term = paste0(c(strsplit_tmp[1:min(6, length(strsplit_tmp))], "..."), collapse=" ")
     return(shortened_term)})
    

    # dt_results_GERR_metaMod_plot_sub$truncate = sapply(dt_results_GERR_metaMod_plot_sub$term, function(x){
    #   nchar(x)>55
    #   })
    # 
    # dt_results_GERR_metaMod_plot_sub$term[dt_results_GERR_metaMod_plot_sub$truncate] <- sapply(dt_results_GERR_metaMod_plot_sub$term[dt_results_GERR_metaMod_plot_sub$truncate], function(long_term) {
    #   
    #   strsplit_tmp = strsplit(x = long_term, split=" ")[[1]]
    #   shortened_term = paste0(c(strsplit_tmp[1:min(4, length(strsplit_tmp))], "..."), collapse=" ")
    #  return(shortened_term)})
    
    #stringr::str_trunc(dt_results_GERR_metaMod_plot_sub$term, width=50,side="right")
    

    
    # truncate term if necessary 
    
    # remove any duplicates 
    dt_results_GERR_metaMod_plot_sub=dt_results_GERR_metaMod_plot_sub[!duplicated(term)]
    
    # fix term order 
    dt_results_GERR_metaMod_plot_sub$term = factor(dt_results_GERR_metaMod_plot_sub$term, levels=rev(unique(dt_results_GERR_metaMod_plot_sub$term)), ordered = T)
    
    vec_colors_geneset_db =  wes_palette(name="Cavalcanti1", n=3)
    names(vec_colors_geneset_db) = c("GO","KEGG","REACTOME")
  
    
# Plot
  p_GO <- ggplot(dt_results_GERR_metaMod_plot_sub, aes(x=term, y=minlog10_p.value_fisher_BH,  fill=source)) +
    geom_bar(stat="identity", colour="black", width=1, size=0.5) + # fill="#73BCC9", ) +
      scale_fill_manual(values=vec_colors_geneset_db[names(vec_colors_geneset_db) %in% unique(dt_results_GERR_metaMod_plot_sub$source)]) +
    #theme_pubr(legend="none") +
      theme(
      axis.title.x = element_text(size = fontSize_axisTitle_md),
      axis.text.x = element_text(size=fontSize_axisText_lg),
       
      axis.title.y = element_blank(),
      axis.text.y = element_text(size=fontSize_axisText_xlg, face="bold"),
      #axis.ticks.y = element_blank(),
      
      legend.title = element_blank(),
      legend.text = element_text(size = fontSize_legend_xlg),
         
       panel.grid.major = element_blank(), 
       panel.grid.minor = element_blank(),
       panel.background = element_blank(), 

      plot.background = element_blank(), 
      panel.grid = element_blank(),
      aspect.ratio = 2.5) +
    #scale_size(range = c(5,10)) +
    #ggsci::scale_fill_lancet() +
    coord_flip() +
    xlab(NULL) + ylab(expression(paste(bold(-log[10]),bold("("),
                                       bolditalic("P"),bold(")"))))
    #scale_size(range = c(5,10)) +
    #scale)
    # ggsci::scale_fill_lancet() +

    
    
  
  p_GO
  
  ggsave(p_GO,filename = paste0(dirWGCNA_plots,prefixData,"_", prefixRun,  "_fig3CDEFG_GO_barplot_", metamod, "_", params$date,".pdf"))#height=round(nrow(dt_results_GERR_metaMod_plot_sub)/9.5,0)) 
    
  
  
})
```

## Fig 3. (C) ii) violin plots of metamodule activity in sc 


```{r}
Idents(seuratObj) = "cluster_perslab_coarse"
levels(seuratObj) = sort(names(vec_colors_cluster_perslab_coarse), decreasing = T)
```

```{r}


list_p_vln_meta_sc = lapply(unique(dt_results_GERR_metaMod_plot$metamodule), function(metamod) {
  
  p_vln <- VlnPlot(object=seuratObj,
                     assay="SCT_combat",
                     features = metamod,
                     pt.size = 0,#0.05,# 0.05,
                     cols=vec_colors_cluster_perslab_coarse,#vec_colors_markers,
                     sort = F, 
                     #group.by = "cluster_perslab_coarse", 
                     #same.y.lims = F,
                     slot="data",
                     log = F,
                     #flip = T
                  )#,

  
  p_vln <- p_vln + 
    ggtitle(gsub("metamodule_", "MM-",metamod)) +
    coord_flip() + 
    theme(
      legend.position="none", 
      #plot.title = element_text(face = "italic"),
      #axis.text.y=element_blank(),
      #plot.margin = margin(b=1, unit="cm")) #margin around entire plot (unit with the sizes of the top, right, bottom, and left margins)
      axis.title.y=element_blank(),
      axis.text.y=element_text(
      hjust=1,
      vjust=0,
      size=fontSize_axisText_lg,
      angle=30),
      
     
      axis.title.x=element_blank(), 
      axis.text.x = element_text(
        size= fontSize_axisText_xsm,
        angle=35),
      plot.title = element_text(size=fontSize_plotTitle),
      aspect.ratio=2.5,
      #legend.position="none"
      
  )
  
  p_vln
  
  # ug, can't save these as pdf due to unicode .. need to do it manually
  # ggsave(plot= p_vln, filename = here("plots", paste0(prefixData, "_", prefixRun, "_Fig3C_vlnplot_", metamod, "_", flagDate,".pdf")), width = 6, height=6)
})
# p_vln_meta_sc <- patchwork::wrap_plots(... = list_p_vln_meta_sc, 
#                             nrow = length(list_p_vln_meta_sc)
#                             )

```



## metamodule network plot - single plot

ABANDONED - LAYOUTS DON'T REFLECT STRUCTURE WELL ENOUGH...

```{r}
 
# Compute network adjacency
mat_adj <- WGCNA::adjacency(mat_scaled_embed_filter, 
                       power = 1, 
                       corFnc = "cor", 
                       corOptions = list(use = "p"),
                       type = "signed hybrid")



# hub.data$logfc <- DEGs$log2FoldChange[match(hub.data$gene, rownames(DEGs))]
# hub.data$p.adj <- DEGs$padj[match(hub.data$gene, rownames(DEGs))]
mat_adj %>% 
graph.adjacency(mode = "undirected", weighted = T, diag = FALSE) %>% 
tidygraph::as_tbl_graph() %>% upgrade_graph() %>% activate(nodes) %>%
dplyr::mutate(metamodule = df_hub.data$metamodule) %>%
dplyr::mutate(module = df_hub.data$module) %>% 
dplyr::mutate(size = df_hub.data$size) %>%
dplyr::mutate(colour = df_hub.data$colour) %>%
#
#dplyr::mutate(abs.logfc = abs(hub.data$logfc)) %>%
#dplyr::mutate(p.adj = hub.data$p.adj) %>% 
#dplyr::mutate(color=factor(ifelse(p.adj < 0.05, yes = "S", no = "NS"), 
#                           levels = c("NS", "S"))) %>%
  activate(edges) %>% 
  #dplyr::filter(weight>.15) %>% 
  activate(nodes) -># %>% filter(!node_is_isolated()) 
  hub.plot



# Plot
# layouts: https://rdrr.io/cran/ggraph/man/layout_tbl_graph_igraph.html
module.plot <- ggraph(hub.plot, 
                      #layout = "igraph",
                      #algorithm = "drl",
                      layout = "stress"
                      ) + 
  geom_edge_link(color="grey50", show.legend=F, aes(alpha=weight)) + 
  geom_node_point(aes(size = size, fill=colour),  shape=21, alpha=0.8) + 
  #scale_fill_manual(values = vec_colors_meta) +
  # scale_size(breaks = c(1,2,3), 
  #            limits = c(0,4), 
  #            range = c(0, 8), 
  #            labels = c(" \U00B1 1", " \U00B1 2", " \U00B1 3")
  #            ) +
  geom_node_text(aes(label = module), fontface="bold", size=5, repel = T, color = "black") +
  # guides(
  #   size = guide_legend(override.aes = list(
  #   size=c(2,4,6)), 
  #   keywidth = 0.8, 
  #   keyheight = 0.8, order = 1)) + 
                             #title = expression(bold(paste("Log"[2],~
                                                            #"fold-change"))))) + 
  theme_graph(base_family = 'Helvetica') + 
  theme(legend.title.align=0.5,
        legend.position = "top",
        legend.margin = margin(0,0,0,0, unit="cm"),
        legend.title = element_text(size=8, face="bold"),
        legend.text = element_text(size=8, face="bold"),
        axis.text =element_blank(),
        axis.ticks = element_blank(), 
        axis.line=element_blank(),
        axis.title = element_blank(),
        legend.spacing.x = unit(0, "cm"),
        # margin: top, right, bottom, and left
        plot.margin = unit(c(0, 0, 0, 0), "cm")) +
  coord_cartesian(clip="off")

module.plot

ggsave(module.plot,filename = paste0(dirWGCNA_plots,prefixData,"_", prefixRun,  "_", mod_plot, "_networkplot_", params$date,".pdf"))
```


individual meta-module network plots

```{r}

list_p_metamod_nw <- lapply(vec_metamodules, function(metamodule) {
  
  
  # Compute network adjacency
  mat_adj <- WGCNA::adjacency(mat_scaled_embed_filter[,colnames(mat_scaled_embed_filter) %in% list_vec_metamod_modules[[metamodule]]], 
                         power = 1, 
                         corFnc = "cor", 
                         corOptions = list(use = "p"),
                         type = "signed hybrid")
  
  df_hub.data_sub = df_hub.data[df_hub.data$metamodule==metamodule,]
  
  # hub.data$logfc <- DEGs$log2FoldChange[match(hub.data$gene, rownames(DEGs))]
  # hub.data$p.adj <- DEGs$padj[match(hub.data$gene, rownames(DEGs))]
  mat_adj %>% 
  graph.adjacency(mode = "undirected", weighted = T, diag = FALSE) %>% 
  tidygraph::as_tbl_graph() %>% upgrade_graph() %>% activate(nodes) %>%
  #dplyr::mutate(metamodule = df_hub.data_sub$metamodule) %>%
  dplyr::mutate(module = df_hub.data_sub$module) %>% 
  dplyr::mutate(size = df_hub.data_sub$size) %>%
  dplyr::mutate(colour = df_hub.data_sub$colour) %>%
  #
  #dplyr::mutate(abs.logfc = abs(hub.data$logfc)) %>%
  #dplyr::mutate(p.adj = hub.data$p.adj) %>% 
  #dplyr::mutate(color=factor(ifelse(p.adj < 0.05, yes = "S", no = "NS"), 
  #                           levels = c("NS", "S"))) %>%
    activate(edges) %>% 
    #dplyr::filter(weight>.15) %>% 
    activate(nodes) -># %>% filter(!node_is_isolated()) 
    hub.plot
  
  
  
  # Plot
  # layouts: https://rdrr.io/cran/ggraph/man/layout_tbl_graph_igraph.html
  module.plot <- ggraph(hub.plot, 
                        layout = "igraph",
                        algorithm = "drl",
                        #layout = "stress"
                        ) + 
    geom_edge_link(color="darkgrey", show.legend=F, aes(alpha=weight)) + 
    geom_node_point(aes(size = size, fill=colour),  shape=21, alpha=0.8) + 
    #scale_fill_manual(values = vec_colors_meta) +
    # scale_size(breaks = c(1,2,3), 
    #            limits = c(0,4), 
    #            range = c(0, 8), 
    #            labels = c(" \U00B1 1", " \U00B1 2", " \U00B1 3")
    #            ) +
    geom_node_text(aes(label = module), fontface="bold", size=5, repel = T, color = "black") +
    # guides(
    #   size = guide_legend(override.aes = list(
    #   size=c(2,4,6)), 
    #   keywidth = 0.8, 
    #   keyheight = 0.8, order = 1)) + 
                               #title = expression(bold(paste("Log"[2],~
                                                              #"fold-change"))))) + 
    theme_graph(base_family = 'Helvetica') + 
    theme(legend.title.align=0.5,
          legend.position = "top",
          legend.margin = margin(0,0,0,0, unit="cm"),
          legend.title = element_text(size=8, face="bold"),
          legend.text = element_text(size=8, face="bold"),
          axis.text =element_blank(),
          axis.ticks = element_blank(), 
          axis.line=element_blank(),
          axis.title = element_blank(),
          legend.spacing.x = unit(0, "cm"),
          # margin: top, right, bottom, and left
          plot.margin = unit(c(0, 0, 0, 0), "cm")) +
    coord_cartesian(clip="off")
  
  module.plot
})

ggsave(module.plot,filename = paste0(dirWGCNA_plots,prefixData,"_", prefixRun,  "_", mod_plot, "_networkplot_", params$date,".pdf"))
```


## Fig. 5 (A) fibrosis severity coefficient bar plot 
REMOVED

```{r}
dt_fibrosis_severity_bar =  dt_condition_coef[condition == "fibrosis_severity" & p.value_percentile_BH < params$pValThreshold/2]
dt_fibrosis_severity_bar = dt_fibrosis_severity_bar[!grepl("gerhard", module)]
dt_fibrosis_severity_bar$celltype = factor(gsub("_\\d+", "", dt_fibrosis_severity_bar$module) %>% gsub("_","-",.), ordered=T)
dt_fibrosis_severity_bar[,module_meta := dt_dict$module_meta[match(dt_fibrosis_severity_bar$module, dt_dict$module_renamed)]]
```

```{r}
# greek characters
for (i in 1:4) {
  dt_fibrosis_severity_bar$celltype =   gsub(pattern=c("alpha-", "beta", "gamma-", "delta")[i], 
  replacement=c(alpha_unicode, Beta_unicode, gamma_unicode, delta_unicode)[i],
  x=dt_fibrosis_severity_bar$celltype)

    dt_fibrosis_severity_bar$module = gsub(pattern=c("alpha_", "beta", "gamma_", "delta")[i], 
  replacement=c(alpha_unicode, Beta_unicode, gamma_unicode, delta_unicode)[i],
  x=dt_fibrosis_severity_bar$module)
      
}
 
```

abbreviated module names

```{r}
dt_fibrosis_severity_bar$module_abbrev <- paste0("M-", vec_celltype_abbrev[match(gsub("_\\d$","",dt_fibrosis_severity_bar$module), gsub("-","_",names(vec_celltype_abbrev)))], "-", gsub("\\D*","",dt_fibrosis_severity_bar$module))
```

```{r}
max_lim_bar = round(max(abs(dt_fibrosis_severity_bar$estimate_original)),0)+0.1
```

```{r}
p_bar_fib_severity <- 
    ggplot(data=dt_fibrosis_severity_bar, aes(x=module_abbrev, y=estimate_original, fill=celltype)) +
  
    geom_bar(stat="identity", 
             position=position_identity()) + #, 
             #group=dt_condition_coef_sub$celltype) + 
  
    #geom_errorbar(aes(ymin=estimate-std.error, ymax=estimate+std.error), width=.1) +
    geom_errorbar(aes(ymin=confInt_lower_percentile, ymax=confInt_upper_percentile), width=.1) +
  
    #scale_fill_brewer(palette="Paired")+
    scale_fill_manual(values=vec_colors_cluster_perslab_coarse[match(unique(dt_fibrosis_severity_bar$celltype), names(vec_colors_cluster_perslab_coarse))]) + # this is a kluge
    #scale_fill_manual(values=unique(df_data$celltype)) + # this is a kluge
    #geom_hline(yintercept=cutoff, linetype="dashed", color = "red") +
    scale_y_continuous("Fibrosis severity", expand = c(0, 0)) +
    scale_x_discrete("module") +  
    # geom_text(aes(label=module),
    #         #angle=90,
    #         #vjust=-1,
    #         hjust = 2,
    #         color="black",
    #         angle=90,
    #        position = position_dodge(), 
    #        size=2.5) +
    #labs(y = "linear association") +
  expand_limits(y=c(-2,0.5)) +
    theme(
     axis.text.x = element_text(angle=90,hjust=1, size=fontSize_axisText_xlg),#element_blank(),
     axis.title.x = element_blank(),
      axis.line.x = element_blank(),
      axis.ticks.x = element_blank(),
      
      axis.title.y = element_text(size = fontSize_axisTitle_lg, hjust = 1, vjust =1),
      plot.background = element_blank(), 
      panel.grid = element_blank(),
     
     legend.position = "none", 
     legend.background = element_blank(),
     legend.box.background =element_blank(),
     #  legend.text = element_text(size=fontSize_legend_lg),
     # legend.title = element_blank(),
     
     aspect.ratio = 0.5)
   

  p_bar_fib_severity 
  
  ggsave(plot = p_bar_fib_severity, filename = paste0(dirWGCNA_plots,prefixData,"_", prefixRun,  "_fig5A_fibr_severity_barplot_", params$date,".pdf"))
```

 
## Fig. 5 (G) and (H) Calgranulin gene VlnPlots


```{r}
vec_calgranulin_genes <- c("S100A8","S100A9","S100A12") # top genes in "honeydew3" macrophage module
```

```{r}
Idents(seuratObj) <- "cluster_perslab"
levels(seuratObj) <- unique(seuratObj$cluster_perslab)[order(gsub("\\d+-","", unique(seuratObj$cluster_perslab)), decreasing=T)]
```

### Fig. 5 (G) 

```{r}
list_vlnplot_calgranulin <- VlnPlot(object=seuratObj,
                     assay="SCT_combat",
                     features = vec_calgranulin_genes,
                     pt.size = 0,# 0.05,
                     cols=vec_colors_cluster_perslab,#vec_colors_markers,
                     sort = F, 
                     #group.by = "cluster_perslab_coarse", 
                     same.y.lims = F,
                     slot="data",
                     log = F,
                     combine = F,
                     flip=F)#,

names(list_vlnplot_calgranulin) <- vec_calgranulin_genes

list_vlnplot_calgranulin_flip <- lapply(1:length(list_vlnplot_calgranulin), function(i) {
  plot_tmp=list_vlnplot_calgranulin[[i]]
  plot_tmp = plot_tmp +
  coord_flip() +
  ggtitle(label=vec_calgranulin_genes[i]) +
  theme(
    plot.title = element_text(size=fontSize_plotTitle, face="bold.italic"),
    axis.text.y=element_blank(),
    plot.margin = margin(b=1, unit="cm")) #margin around entire plot (unit with the sizes of the top, right, bottom, and left margins)

  if (i==1) {
  plot_tmp <- plot_tmp +
  theme(
    axis.text.y=element_text(
    hjust=1,
    vjust=0,
    size=fontSize_axisText_md,
    angle=30
    )
  )
  }
  # } else {
  #   plot_tmp <- plot_tmp + 
  #   theme(axis.text.y=element_text(
      # vjust=0.2,
      # size=10,
      # angle=30,
  #     margin = margin(l = 0, r = 5)
  #     )
  #     )
  # }
  
  plot_tmp <- plot_tmp + 
    theme(
      aspect.ratio=3,
      axis.title.y=element_blank(),
      axis.title.x=element_blank(), 
      axis.text.x = element_text(
        size= fontSize_axisText_sm,
        angle=35),
      plot.title = element_text(size=fontSize_plotTitle),
      #aspect.ratio=2,
      legend.position="none")
})

p5G <- patchwork::wrap_plots(... = list_vlnplot_calgranulin_flip, 
                            ncol = length(list_vlnplot_calgranulin_flip)
                            )

p5G

#ggsave(plot= p5G, filename = here("plots", paste0(prefixData, "_", prefixRun, "_fig5G_calgranulin_vlnplot_perslab_",flagDate,".pdf")))#, width = 12, height=12)
```


### Fig. 5 H

```{r}
Idents(seuratObj_macparland) <- "Cluster_annot"
levels(seuratObj_macparland) <- sort(unique(seuratObj_macparland$Cluster_annot), decreasing = T)
```

```{r}
list_vlnplot_calgranulin_macparland <- VlnPlot(object=seuratObj_macparland,
                     assay="SCT",
                     features = vec_calgranulin_genes,
                     pt.size = 0,# 0.05,
                     #cols=vec_colors_cluster_perslab,#vec_colors_markers,
                     sort = F, 
                     #group.by = "cluster_perslab_coarse", 
                     same.y.lims = F,
                     slot="data",
                     log = F,
                     combine = F,
                     flip=F)#,

names(list_vlnplot_calgranulin_macparland) <- vec_calgranulin_genes

list_vlnplot_calgranulin_macparland_flip <- lapply(1:length(list_vlnplot_calgranulin_macparland), function(i) {
  plot_tmp=list_vlnplot_calgranulin_macparland[[i]]
  plot_tmp = plot_tmp +
  coord_flip() +
  ggtitle(label=vec_calgranulin_genes[i]) +
  theme(
    plot.title = element_text(size=fontSize_plotTitle, face="bold.italic"),
    axis.text.y=element_blank(),
    plot.margin = margin(b=1, unit="cm")) #margin around entire plot (unit with the sizes of the top, right, bottom, and left margins)

  if (i==1) {
  plot_tmp <- plot_tmp +
  theme(
    axis.text.y=element_text(
    hjust=1,
    vjust=0,
    size=fontSize_axisText_md,
    angle=30
    )
  )
  }
  # } else {
  #   plot_tmp <- plot_tmp + 
  #   theme(axis.text.y=element_text(
      # vjust=0.2,
      # size=10,
      # angle=30,
  #     margin = margin(l = 0, r = 5)
  #     )
  #     )
  # }
  
  plot_tmp <- plot_tmp + 
    theme(
      aspect.ratio=3,
      axis.title.y=element_blank(),
      axis.title.x=element_blank(), 
      axis.text.x = element_text(
        size= fontSize_axisText_sm,
        angle=35),
      plot.title = element_text(size=fontSize_plotTitle),
      #aspect.ratio=2,
      legend.position="none")
})

p5H <- patchwork::wrap_plots(... = list_vlnplot_calgranulin_macparland_flip, 
                            ncol = length(list_vlnplot_calgranulin_macparland_flip)
                            )

p5H

#ggsave(plot= p5G, filename = here("plots", paste0(prefixData, "_", prefixRun, "_fig5G_calgranulin_vlnplot_perslab_",flagDate,".pdf")))#, width = 12, height=12)
```

 
## plot association betas with confidence interval

<!-- add confidence interval -->

<!-- ```{r} -->
<!-- list_dt_condition_coef = lapply(list_dt_condition_coef, function(dt_condition_coef){ -->
<!--   dt_condition_coef = dt_condition_coef[grep("perslab",module)] -->
<!--   dt_condition_coef[,lower:=qnorm(p=0.05/2, mean = estimate, sd=std.error)] -->
<!--   dt_condition_coef[,upper:=qnorm(p=1-0.05/2, mean = estimate, sd=std.error)] -->
<!-- }) -->
<!-- ``` -->

====================================================================

## featureplot of module embeddings


```{r}

#p = FeaturePlot(object = seuratObj, features = "")
```






```{r}
Idents(seuratObj) <- seuratObj$cluster_perslab_coarse

seuratObj <- subset(x=seuratObj,
                    idents=vec_celltypesOfInterest)
```

### load gene module cell embedding matrix

we will need to compute embeddings of all modules - including bulk - in sc

<!-- ```{r} -->
<!-- path_dt_embed_gerhard <- paste0(dirWGCNA_tables, "moylan2013","_","wgcna1","_modules_sc_cellModEmbed.csv") -->
<!-- dt_embed_moylan2013 <- fread(path_dt_embed_moylan2013) -->
<!-- dt_embed_moylan2013[0:4,0:5] -->

<!-- ``` -->

cbind embeddings

```{r}
colnames(dt_embed_moylan2013)[2:ncol(dt_embed_moylan2013)] <- paste0("moylan2013__",colnames(dt_embed_moylan2013)[2:ncol(dt_embed_moylan2013)])
```

```{r}
dt_embed <- cbind(dt_embed_sc,dt_embed_moylan2013)
```

filter embeddings to celltypes of interest GOT TO HERE

```{r}

vec_celltypesOfInterest %>% gsub("-","_",.) %>% lapply(., function(eachName) grepl(eachName,colnames(dt_embed_sc))) %>% Reduce(f = '|',x=.) %>% which -> vec_colsOfInterest

```

```{r}
dt_embed_sc_sub <- dt_embed_sc[,c(1:3, ..vec_colsOfInterest),]

colnames(dt_embed_sc_sub)
```

```{r}
dim(dt_embed_sc_sub)
#[1] 21007   130
```

127 modules

<!-- ```{r} -->
<!-- dt_embed_sc_meta <- fread(here("output", paste0(prefixData, "_" , prefixRun, "_", prefixOut, "_cellModEmbed_meta.csv"))) -->
<!-- ``` -->







```{r}
path_dt_metadataCombined <- here("data", "metadataCombined.csv")
dt_metadataCombined<-fread(file=path_dt_metadataCombined)

head(dt_metadataCombined)

#   ID      Age Sex Ethnicity      BMI Smoking Alcohol_use nas_sum fibrosis_grade lobular_inflammation_grade
# 1: 235L 44.55068   0    Danish 35.15850       1           1       6              2                          2
# 2: 242L 53.32055   1    Danish 37.48722       0           2       5              1                          2
# 3: 245L 43.53973   1    Danish 44.48491       1           1       5              2                          2
# 4: 247L 51.12603   0    Danish 52.65934       1           1       5              2                          2
# 5: 249L 64.40822   0    Danish 44.47846       1           1       4              2                          3
# 6: 252L 47.02192   1    Danish 44.85055       1           1       5              2                          2
```

```{r}
#load("/projects/jonatan/pub-perslab/18-liver-wgcna/RObjects/liver_perslab_int_wgcna1_final_session_image.RData.gz")
```

load data table from metadata association analysis 

```{r}
path_dt_metadataAssoc <- here("output", paste0(prefixData, "_" , prefixRun, "_moylan2013_wgcna_assoc_results.csv"))
dt_metadataAssoc <- fread(path_dt_metadataAssoc)
dt_metadataAssoc
```

# Analysis and plotting

Generate a lookup table for module - cell type 

```{r}
dt_geneModDict <- dt_geneMod[!duplicated(dt_geneMod[[colModule]]),]
```

```{r}
dt_geneModDict[[colModule]] %>% na.omit %>% length
# [1] 110
```


## correlation matrix 

Very high correlations between gene modules could indicate confounding

Even if serious confounding is not suspected, it may make sense to try to merge very overlapping and correlated modules across cell types, which the current cell-type-wise WGCNA workflow does not do.

## correlation matrix 

```{r}
mat_modCor <- cor(x=dt_embed_sc_sub[,-c("data", "cell_cluster", "cell_id")],
                               use="pairwise.complete.obs",
                               method="pearson")

```

```{r}

pdf(file = here("plots", paste0(prefixData, "_" , prefixRun, "_", prefixOut, "_", flagDate, "_moduleCorr.pdf")), width=20, height=20)
corrplot::corrplot(corr = mat_modCor, method="number", number.cex = 0.5)
dev.off()
```

<!-- #TODO:  -->

<!-- ## Hierarchically cluster modules -->

<!-- plot the eigengene (kIM embeddings) correlations  -->

<!-- ```{r} -->

<!-- #http://www.sthda.com/english/wiki/beautiful-dendrogram-visualizations-in-r-5-must-known-methods-unsupervised-machine-learning -->

<!-- t(dt_embed_sc[,4:ncol(dt_embed_sc)]) %>% # data -->
<!--   scale %>% # Scale the data -->
<!--   dist %>% # calculate a distance matrix,  -->
<!--   hclust(method = "ward.D2") -> hc # Hierarchical clustering  -->

<!-- dend <- as.dendrogram(hc) # Turn the object into a dendrogram. -->

<!-- plot(dend) -->
<!-- ``` -->

<!-- ```{r} -->
<!-- dendextend::get_leaves_attr(attribute = "label", dend=dend) -->
<!-- ``` -->

<!-- ```{r} -->
<!-- dend %>%  -->
<!--   set(what="labels_col", value=dendextend::get_leaves_attr(attribute = "label", dend=dend)) -> dend # change leaf colors -->
<!--   #set("leaves_pch", rep(c(1:length(unique(dt_geneMod$cell_cluster))),table(dt_geneModDict$cell_cluster)))  # node point type for cell type -->

<!-- #plot(dend) -->
<!-- ``` -->

```{r}

vec_cellCluster_integer <- brewer.pal(n = length(unique(dt_geneMod[[colCellCluster]])), name="Spectral")
names(vec_cellCluster_integer) <- unique(dt_geneMod[[colCellCluster]])

dend %>% 
  set(what="leaves_pch", value=19) %>%
  set(what="leaves_cex", 2) %>% 
  set(what="leaves_col", value=vec_cellCluster_integer[match(dt_geneModDict[[colCellCluster]][match(get_leaves_attr(attribute = "label", dend=dend), dt_geneModDict[[colModule]])], names(vec_cellCluster_integer))]) -> 
  dend

plot(dend)
```

## Module-module correlation matrix 

```{r}
#TODO
```

## Plot the embeddings in UMAP

group plots per celltype 

```{r}
list_vec_modulesSignif <- lapply(unique(dt_metadataAssoc[,cell_cluster,]), function(cell_clust) {
  dt_metadataAssoc[cell_cluster==cell_clust & 
                     p.adjust(p.value, method = pAdjMethod) <= pValThreshold, "module"][[1]]
  })

names(list_vec_modulesSignif) <- unique(dt_metadataAssoc[,cell_cluster,])

list_vec_modulesSignif
```

```{r}

dt_metaTmp <- dt_embed_sc[,-"cell_id",]
rownames(dt_metaTmp) <- dt_embed_sc[,cell_id,]
all.equal(rownames(dt_metaTmp), colnames(seuratObj))

#[1] TRUE

```

```{r}
seuratObj <- AddMetaData(object = seuratObj, 
                         metadata = dt_metaTmp)
```

```{r}

invisible(lapply(names(list_vec_modulesSignif), function(cell_cluster) {
  
  vec_modulesSignif = list_vec_modulesSignif[[cell_cluster]]
  
  list_featPlot <- lapply(vec_modulesSignif, function(module) {
    FeaturePlot(object = seuratObj, 
                features = module, 
                reduction = "umap", pt.size = .1)
    }
    )
  
  p1<- cowplot::plot_grid(plotlist = list_featPlot,
                          ncol= (list_featPlot %>% length %>% sqrt %>% as.integer),
                          labels=NULL)#as.list(vec_modulesSignif))
  
  saveMeta(savefnc = cowplot::save_plot, 
           plot=p1,
            ncol= (list_featPlot %>% length %>% sqrt %>% as.integer),
            filename =  here("plots",paste0(prefixData, "_", prefixRun, "_", prefixOut, "_", cell_cluster, "_wgcnaModuleUMAPembeddings.pdf")),
            base_height = 15, # of a subplot
            #base_width = 5,
            base_aspect_ratio = 0.3,
            limitsize=F)
}))
```

## gene module heatmaps

The gene module embeddings are computed using the SC- normalized counts. 
Hence we need to z-score them to plot on a single heatmap

First order the rows by cell cluster

```{r}
dt_embed_sc[,cell_cluster:=dt_embed_sc_meta[,"cell_cluster",]]
dt_embed_sc <- dt_embed_sc[order(cell_cluster)]
```

```{r}
vec_modulesSignif = colnames(dt_embed_sc)
vec_modulesSignif = vec_modulesSignif[!vec_modulesSignif %in% c("cell_id", "cell_cluster")]

dt_embed_sc[,lapply(.SD, scale),, .SDcols=vec_modulesSignif] %>% as.matrix -> mat_cellModEmbed_scaled 

rownames(mat_cellModEmbed_scaled) <- dt_embed_sc[,c("cell_id")][[1]]
```

Find locations for labels halfway through cell types in rows
```{r}
vec_table <- table(dt_embed_sc[,cell_cluster,])
vec_cumsum <- cumsum(c(1,vec_table)) 
vec_pos <- sapply(2:length(vec_cumsum), function(i) {
  as.integer(vec_cumsum[i-1]+vec_table[i-1]/2)
})
```


```{r}
set.seed(randomSeed)
colorvec = sample(x=unique(gsub("\\d","",colors())), size=length(unique(dt_metadataAssoc[,cell_cluster,])), replace=F)
names(colorvec) =unique(dt_metadataAssoc[,cell_cluster,])
```

```{r}
pdf(file = here("plots", paste0(prefixData, "_" , prefixRun, "_", prefixOut, "_modEmbedHeatmap.pdf")), width=20, height=20)
p <- ComplexHeatmap::Heatmap(mat=mat_cellModEmbed_scaled, 
                             name="Module embeddings",
                             row_title = "cell",
                             column_title = "gene module",
                             cluster_rows = FALSE,
                             #cluster_cols = FALSE,
                             show_row_names=FALSE) 



p <- p + rowAnnotation(cell_cluster = dt_embed_sc$cell_cluster,
    col = list(cell_cluster = colorvec))

# p <- p + columnAnnotation(cell_cluster = dt_geneModDict$cell_cluster[match(vec_modulesSignif,dt_geneModDict$module)],
#     col = list(cell_cluster = colorvec))

# p <- p+ rowAnnotation(foo = anno_link(at = vec_pos, which="row", side="right",
#                     labels = names(vec_table)))
draw(p)
dev.off()
```

## Module - cell cluster specificity

```{r}
#TODO
# NOT PRIORITY
```



### LEFTOVERS FROM LIVER MODULE SC QC  - CHECK ! ###

# QC plotting



The gene module embeddings are computed using the SC- normalized and ComBat corrected counts. 
Hence we need to z-score them to plot on a single heatmap

```{r}
dt_embed_sc_sub[,lapply(.SD, scale),, .SDcols=colnames(dt_embed_sc_sub[,-c("data", "cell_cluster", "cell_id")])] %>% as.matrix -> mat_cellModEmbed_sub_scaled 

colnames(mat_cellModEmbed_sub_scaled) %>% gsub("\\.V1","", .) -> colnames(mat_cellModEmbed_sub_scaled)

rownames(mat_cellModEmbed_sub_scaled) <- dt_embed_sc_sub[,c("cell_id")][[1]]

mat_cellModEmbed_sub_scaled[0:4,0:4]

#                   Cholangiocytes__aliceblue.V1 Cholangiocytes__azure2.V1 Cholangiocytes__black.V1 Cholangiocytes__chocolate.V1
# TGTCCCACAATGAATG_1                    2.4859331              -0.009620634             -0.001129782                 -0.007353158
# AAACGGGGTCTACCTC_1                    0.7263057              -0.009620634             -0.001129782                 -0.007353158
# AAGTCTGGTAATAGCA_1                   -1.6010004              -0.009620634             -0.001129782                 -0.007353158
# AATCGGTCATGATCCA_1                    3.4825015              -0.009620634             -0.001129782                 -0.007353158
```

<!-- Find locations for cell type labels halfway through cell types in rows -->

<!-- ```{r} -->
<!-- vec_cellClustTable <- table(dt_embed_sc_sub[,cell_cluster,]) -->
<!-- vec_cellClustCumsum <- cumsum(c(1,vec_cellClustTable))  -->
<!-- vec_cellClustPos <- sapply(2:length(vec_cellClustCumsum), function(i) { -->
<!--   as.integer(vec_cellClustCumsum[i-1]+vec_cellClustTable[i-1]/2) -->
<!-- }) -->
<!-- ``` -->

make colors to identify cell clusters in row annotation

```{r}
set.seed(randomSeed)
vec_colorCellClust = sample(x=unique(gsub("\\d","",colors())), size=length(unique(dt_embed_sc_sub[,cell_cluster,])), replace=F)
names(vec_colorCellClust) =unique(dt_embed_sc_sub[,cell_cluster,])
vec_colorCellClust
```

make colors to identify sample_ID in row annotation

```{r}
set.seed(randomSeed)
vec_colorSampleID = sample(x=unique(gsub("\\d","",colors())), size=length(unique(seuratObj$sample_ID)), replace=F)
names(vec_colorSampleID) =unique(seuratObj$sample_ID)
vec_colorSampleID

```
<!-- Find locations for labels halfway through sample_ID in rows -->

<!-- ```{r} -->

<!-- vec_cellClustTable <- table(dt_embed_sc_sub[,cell_cluster,]) -->
<!-- vec_cellClustCumsum <- cumsum(c(1,vec_cellClustTable))  -->
<!-- vec_cellClustPos <- sapply(2:length(vec_cellClustCumsum), function(i) { -->
<!--   as.integer(vec_cellClustCumsum[i-1]+vec_cellClustTable[i-1]/2) -->
<!-- }) -->
<!-- ``` -->

```{r}
pdf(file = here("plots", paste0(prefixData, "_" , prefixRun, "_", prefixOut, "_", flagDate, "_modEmbedHeatmap.pdf")), width=20, height=20)
p <- ComplexHeatmap::Heatmap(mat=mat_cellModEmbed_sub_scaled, 
                             name="Module embeddings",
                             row_title = "cell",
                             column_title = "gene module",
                             cluster_rows = FALSE,
                             #cluster_cols = FALSE,
                             show_row_names=FALSE,
                             use_raster = TRUE, 
                             raster_quality = 2) 



p <- p + HeatmapAnnotation(cell_cluster = dt_embed_sc_sub$cell_cluster,
    col = list(cell_cluster = vec_colorCellClust) ,
    which="row")

p <- p + HeatmapAnnotation(df = data.frame("sample_ID" = seuratObj$sample_ID[match(dt_embed_sc_sub$cell_id, rownames(seuratObj@meta.data))]),
    col = list(sample_ID = vec_colorSampleID),
    which="row")

# p <- p + HeatmapAnnotation(df = data.frame("cell_cluster" = mat_cellModEmbed_sub_scaled %>% colnames %>% gsub("__.*$","", .)),
#                            col = list(cell_cluster = vec_colorCellClust), 
#                            which="column")


draw(p)
dev.off()
```







### feature plots

```{r}
genes = c("TREM2")
```

```{r,fig.width=10}
Seurat::FeaturePlot(object=seuratObj, features="TREM2")
```

```{r}
vec_mods = c("aquamarine", "brown", "mediumaquamarine")
```

```{r}
DefaultAssay(object = seuratObj) = "integrated"
```

```{r}
p <- Seurat::FeaturePlot(object=seuratObj, features=vec_mods, blend = F, order = T)
saveMeta(savefnc= ggsave, plot= p, filename = here("plots", paste0(prefixData, "_", prefixRun, "_", prefixOut, "_mod_featplots_",flagDate,".pdf")), width = 12, height=12)
```

### violin plots

```{r}
vec_genes = c("VCAN")
```

```{r}
DefaultAssay(seuratObj) = "SCT_combat"
```

```{r}
#list_seuratObj[[1]]$macs_subclust <- factor(list_seuratObj[[1]]$macs_subclust, levels = 0:6, ordered=T)
for (gene in vec_genes) {
  p <- Seurat::VlnPlot(object = seuratObj, 
                       features = vec_genes,
                       assay = "SCT_combat", 
                       slot = "data", 
                       log=FALSE,
                       sort = F, 
                       group.by = "cluster_perslab")
  
  ggsave(plot = p, filename = here("plots", paste0(prefixData, "_",prefixRun, "_",gene,"_vlnplot_", params$date, ".pdf")), width = 12, height=9)
}
```

## bar plot

sc wgcna modules

```{r}

dt_geneMod_perslab_merged$module_pres %>% unique -> vec_mods_perslab

vec_mods_perslab<- vec_mods_perslab[!is.na(vec_mods_perslab) & nchar(vec_mods_perslab)>0]

mat_embed_perslab <- sapply(vec_mods_perslab, function(mod) {

  # get gene weights
  condition = quote(dt_geneMod_perslab_merged[[colModule]]==mod)
  vec_pkMs <- dt_geneMod_perslab_merged[eval(condition), pkMs,]

  # normalize module gene weights to sum to 1
  vec_pkMs <- vec_pkMs/sum(vec_pkMs)

  # find corresponding rows of the expression matrix
  vec_idxRow <- match(dt_geneMod_perslab_merged[eval(condition), ..colGeneNames, ][[1]], dt_datExpr[,gene])

  # filter out mis-matches
  vec_pkMs <- vec_pkMs[!is.na(vec_idxRow)]
  vec_idxRow <- vec_idxRow[!is.na(vec_idxRow)]

  # compute weighted sum of normalized expression
  dt_datExpr[vec_idxRow,-("gene")] %>% as.matrix -> mat_sub

  vec_pkMs %*% mat_sub

})

rownames(mat_embed_perslab) <- colnames(dt_datExpr)[-1]

mat_embed_perslab[0:4,0:4]
```


## various plots, not for publication

### boxplot of Mac4 and hub genes: calgranulin genes and VCAM 

load Gerhard 2018 bulk expression data 

```{r}
dt_datExpr <- fread(here("data", "liver_gerhard2018_norm.counts.csv.gz"))

mat_datExpr <- as.matrix(dt_datExpr[,-1])
rownames(mat_datExpr) = dt_datExpr$gene
```

```{r}
# df_embed = fread("/projects/jonatan/pub-perslab/18-liver-wgcna/tables/liver_perslab_int_wgcna3_mod_embed_gerhard2018.csv.gz") %>% setDF
df_scaled_embed = fread("/projects/jonatan/pub-perslab/18-liver-wgcna/tables/liver_perslab_int_wgcna3_mat_scaled_embed_gerhard2018.csv.gz") %>% setDF

mat_scaled_embed = df_scaled_embed[,-1] %>% as.matrix
rownames(mat_scaled_embed) = df_scaled_embed$run_accession
mat_scaled_embed[0:4,0:4]
```

```{r}
mod_plot = c("Macrophages_4")#,"Dendritic-cells_1")
vec_genesplot <- c("S100A8","S100A9","S100A12", "VCAN", "S100A6") # top genes in "honeydew3" macrophage module

```

```{r}

dt_boxplot = data.table(dt_metadata[,.(DCL.Patient.ID,Diagnosis)], "Macrophages_4"=mat_scaled_embed[,mod_plot])

dt_boxplot <- data.table(dt_boxplot, t(mat_datExpr)[,vec_genesplot])

#colnames(dt_boxplot) <- gsub("\\.Cholangiocytes_5|\\.Dendritic-cells_1","",colnames(dt_boxplot))
# dt_plot_sc = data.table("sample_ID" = seuratObj$sample_ID, 
#                           "fibrosis_grade"=seuratObj$fibrosis_grade,
#                           "module_activity"=log2(seuratObj@meta.data$`Hepatic-stellate-cells_4`+1))
dt_boxplot <- dt_boxplot[grepl("Fibrosis 3|Fibrosis 4|NORMAL", Diagnosis)]
# dt_boxplot <- dt_boxplot[!grepl("Fibrosis 3/4", Diagnosis)]
#dt_boxplot <- melt(dt_boxplot,id.vars = c("DCL.Patient.ID", "Diagnosis"))

#dt_boxplot$variable = gsub("module\\.","",dt_boxplot$variable)
#colnames(dt_boxplot)[c(3,4)] <- c("module","activity")
dt_boxplot$Diagnosis = factor(dt_boxplot$Diagnosis, levels = c("NORMAL","Fibrosis 3", "Fibrosis 3/4", "Fibrosis 4"), ordered=TRUE)
#dt_boxplot$Diagnosis = factor(dt_boxplot$Diagnosis, levels = c("NORMAL","Fibrosis 3", "Fibrosis 3/4", "Fibrosis 4"), ordered=TRUE)
#dt_plot_box_subsample = dt_plot_box[,lapply(.SD, function(x) {head(x,n=3)}), by=Diagnosis]

```


```{r}
vec_features_plot = c(mod_plot, vec_genesplot)
```

```{r}
vec_cols_plot = c("white", lighten("red", 0.6), lighten("red",0.3), "red")

names(vec_cols_plot) = levels(dt_boxplot$Diagnosis)

```

```{r}

list_p_Mac4 <- lapply(vec_features_plot, function(feat_plot) {
  

  p_box = ggplot(data=dt_boxplot, mapping=aes_string(x="Diagnosis", y=feat_plot, fill="Diagnosis"), ) +
    geom_boxplot()  +
    #facet_grid(rows = module) + 
    scale_fill_manual(values=vec_cols_plot) +
    geom_jitter(width=0.1,alpha=0.5) +
    theme(legend.position = NULL)+
    theme(
        axis.text = element_text(size = fontSize_axisText_sm), 
        axis.line = element_blank(), 
        axis.ticks = element_blank(), 
        axis.title = element_blank(),
        #axis.title = element_text(size = fontSize_axisTitle_sm),#, face = "bold"), 
        panel.background = element_blank(), 
        #legend.position = "none",
        #panel.grid.minor = element_blank(), 
        #panel.grid.major = element_line(colour = "grey80"), 
        plot.background = element_blank(), 
        panel.grid = element_blank(),
        #text = element_text(size = ),
        plot.title = element_text(size = fontSize_plotTitle),
        #legend.title=element_blank(),
        legend.position = "none",
        #legend.text = element_text(size=fontSize_legend_md),
        aspect.ratio = 1.5) + 
    labs(y="module activity") +
    ggtitle(feat_plot)
    

})


p_Mac4 <- patchwork::wrap_plots(... = list_p_Mac4 , 
                            ncol = 3
                            )

p_Mac4
```


```{r}
FeaturePlot(object = seuratObj, features = "VCAN")
```


## plot module-module intersect matrix


```{r}
# dt_geneMod_merged = rbindlist(list(dt_geneMod_perslab,dt_geneMod_gerhard2018, dt_geneMod_moylan2013))
# 
# dt_geneMod_merged[[colModule]] %>% unique -> vec_mods
dt_geneMod_perslab$module_renamed %>% unique -> vec_mods

vec_mods <- vec_mods[!is.na(vec_mods) & nchar(vec_mods)>0]
```

get overlap matrix

```{r}
# dt_geneMod_perslab[[colModule]] %>% unique -> vec_mods
# 
# vec_mods <- vec_mods[!is.na(vec_mods) & nchar(vec_mods)>0]

list_genelists = lapply(vec_mods, function(mod) {
  condition = quote(dt_geneMod_perslab$module_renamed==mod)
  vec_mod_genes = dt_geneMod_perslab[eval(condition), ..colGeneNames, ][[1]] %>% unique
  # filter out moduke genes absent in expression data
  #vec_idxRow <- match(vec_mod_genes, dt_datExpr[,gene])
  #vec_mod_genes <- vec_mod_genes[!is.na(vec_idxRow)]
  return(vec_mod_genes)
})

names(list_genelists) = vec_mods
```

```{r}
  
for (i in 1:4) {
 
names(list_genelists)=   gsub(pattern=c("alpha-", "beta", "gamma-", "delta")[i], 
  replacement=c(alpha_unicode, Beta_unicode, gamma_unicode, delta_unicode)[i],
  x=names(list_genelists))

}

vec_mod_number = gsub("\\D","",names(list_genelists) )
names(list_genelists) = paste0("M-", vec_celltype_abbrev[match(gsub("_\\d$","",names(list_genelists)), names(vec_celltype_abbrev))], "-", vec_mod_number)
```

the intersect matrix tells you how big a proportion of the row module's genes are also in the column module

```{r}

# outer loop is columns, inner is rows
fun = function(genelist) {
    sapply(list_genelists, function(genelistOther) {
      base::intersect(x=genelist, y=genelistOther) %>% length %>% '/'(length(genelist))
    }, simplify=T)
  }
  
mat_moduleGeneIntersect <- sapply(FUN=fun,"X"=list_genelists, simplify = T)

mat_moduleGeneIntersect[row(mat_moduleGeneIntersect)==col(mat_moduleGeneIntersect)] <- 0 
#<- mat_moduleGeneIntersect-diag(1, nrow=nrow(mat_moduleGeneIntersect))
```



```{r}
# pdf(mat_moduleGeneIntersect, file = here("plots", paste0(prefixData, "_", prefixRun, "_Fig3_module_intersect_",flagDate,".pdf")))#, width=max(10,ncol(mat_moduleGeneIntersect) %/% 3),height=max(10,ncol(mat_moduleGeneIntersect) %/% 3))
  
  pheatmap(mat_moduleGeneIntersect, 
           cluster_rows = T, 
           cluster_cols = T,
           )
# corrplot(corr = mat_moduleGeneIntersect, 
#          method = "color",
#          col = colorRampPalette(rev(brewer.pal(n = 11, name = "RdYlBu")), bias = 1)(200),
#          diag = F,
#          is.corr=F,
#          title="prop. of mod j overlapping mod i",
#          order = "original",#hclust",
#          hclust.method = "average",
#          addCoef.col = "black",
#          tl.srt = 45,
#          number.digits = 2L,
#          number.cex = 0.5)

# dev.off()
```
